<!DOCTYPE html>
<!-- START: inst/pkgdown/templates/layout.html --><!-- Generated by pkgdown: do not edit by hand --><html lang="en" data-bs-theme="auto"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><title>Spatial Transcriptomics: Normalization in Spatial Transcriptomics</title><meta name="viewport" content="width=device-width, initial-scale=1"><script src="assets/themetoggle.js"></script><link rel="stylesheet" type="text/css" href="assets/styles.css"><script src="assets/scripts.js" type="text/javascript"></script><!-- mathjax --><script type="text/x-mathjax-config">
    MathJax.Hub.Config({
      config: ["MMLorHTML.js"],
      jax: ["input/TeX","input/MathML","output/HTML-CSS","output/NativeMML", "output/PreviewHTML"],
      extensions: ["tex2jax.js","mml2jax.js","MathMenu.js","MathZoom.js", "fast-preview.js", "AssistiveMML.js", "a11y/accessibility-menu.js"],
      TeX: {
        extensions: ["AMSmath.js","AMSsymbols.js","noErrors.js","noUndefined.js"]
      },
      tex2jax: {
        inlineMath: [['\\(', '\\)']],
        displayMath: [ ['$$','$$'], ['\\[', '\\]'] ],
        processEscapes: true
      }
    });
    </script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><!-- Responsive Favicon for The Carpentries --><link rel="apple-touch-icon" sizes="180x180" href="favicons/incubator/apple-touch-icon.png"><link rel="icon" type="image/png" sizes="32x32" href="favicons/incubator/favicon-32x32.png"><link rel="icon" type="image/png" sizes="16x16" href="favicons/incubator/favicon-16x16.png"><link rel="manifest" href="favicons/incubator/site.webmanifest"><link rel="mask-icon" href="favicons/incubator/safari-pinned-tab.svg" color="#5bbad5"><meta name="msapplication-TileColor" content="#da532c"><meta name="theme-color" media="(prefers-color-scheme: light)" content="white"><meta name="theme-color" media="(prefers-color-scheme: dark)" content="black"></head><body>
    <header id="top" class="navbar navbar-expand-md top-nav incubator"><svg xmlns="http://www.w3.org/2000/svg" class="d-none"><symbol id="check2" viewbox="0 0 16 16"><path d="M13.854 3.646a.5.5 0 0 1 0 .708l-7 7a.5.5 0 0 1-.708 0l-3.5-3.5a.5.5 0 1 1 .708-.708L6.5 10.293l6.646-6.647a.5.5 0 0 1 .708 0z"></path></symbol><symbol id="circle-half" viewbox="0 0 16 16"><path d="M8 15A7 7 0 1 0 8 1v14zm0 1A8 8 0 1 1 8 0a8 8 0 0 1 0 16z"></path></symbol><symbol id="moon-stars-fill" viewbox="0 0 16 16"><path d="M6 .278a.768.768 0 0 1 .08.858 7.208 7.208 0 0 0-.878 3.46c0 4.021 3.278 7.277 7.318 7.277.527 0 1.04-.055 1.533-.16a.787.787 0 0 1 .81.316.733.733 0 0 1-.031.893A8.349 8.349 0 0 1 8.344 16C3.734 16 0 12.286 0 7.71 0 4.266 2.114 1.312 5.124.06A.752.752 0 0 1 6 .278z"></path><path d="M10.794 3.148a.217.217 0 0 1 .412 0l.387 1.162c.173.518.579.924 1.097 1.097l1.162.387a.217.217 0 0 1 0 .412l-1.162.387a1.734 1.734 0 0 0-1.097 1.097l-.387 1.162a.217.217 0 0 1-.412 0l-.387-1.162A1.734 1.734 0 0 0 9.31 6.593l-1.162-.387a.217.217 0 0 1 0-.412l1.162-.387a1.734 1.734 0 0 0 1.097-1.097l.387-1.162zM13.863.099a.145.145 0 0 1 .274 0l.258.774c.115.346.386.617.732.732l.774.258a.145.145 0 0 1 0 .274l-.774.258a1.156 1.156 0 0 0-.732.732l-.258.774a.145.145 0 0 1-.274 0l-.258-.774a1.156 1.156 0 0 0-.732-.732l-.774-.258a.145.145 0 0 1 0-.274l.774-.258c.346-.115.617-.386.732-.732L13.863.1z"></path></symbol><symbol id="sun-fill" viewbox="0 0 16 16"><path d="M8 12a4 4 0 1 0 0-8 4 4 0 0 0 0 8zM8 0a.5.5 0 0 1 .5.5v2a.5.5 0 0 1-1 0v-2A.5.5 0 0 1 8 0zm0 13a.5.5 0 0 1 .5.5v2a.5.5 0 0 1-1 0v-2A.5.5 0 0 1 8 13zm8-5a.5.5 0 0 1-.5.5h-2a.5.5 0 0 1 0-1h2a.5.5 0 0 1 .5.5zM3 8a.5.5 0 0 1-.5.5h-2a.5.5 0 0 1 0-1h2A.5.5 0 0 1 3 8zm10.657-5.657a.5.5 0 0 1 0 .707l-1.414 1.415a.5.5 0 1 1-.707-.708l1.414-1.414a.5.5 0 0 1 .707 0zm-9.193 9.193a.5.5 0 0 1 0 .707L3.05 13.657a.5.5 0 0 1-.707-.707l1.414-1.414a.5.5 0 0 1 .707 0zm9.193 2.121a.5.5 0 0 1-.707 0l-1.414-1.414a.5.5 0 0 1 .707-.707l1.414 1.414a.5.5 0 0 1 0 .707zM4.464 4.465a.5.5 0 0 1-.707 0L2.343 3.05a.5.5 0 1 1 .707-.707l1.414 1.414a.5.5 0 0 1 0 .708z"></path></symbol></svg><a class="visually-hidden-focusable skip-link" href="#main-content">Skip to main content</a>
  <div class="container-fluid top-nav-container">
    <div class="col-md-8">
      <div class="large-logo">
        <img id="incubator-logo" alt="Carpentries Incubator" src="assets/images/incubator-logo.svg"><span class="badge text-bg-info">
          <abbr title="This lesson is in the beta phase, which means that it is ready for teaching by instructors outside of the original author team.">
            <a href="https://cdh.carpentries.org/the-lesson-life-cycle.html#polishing-beta-stage" class="external-link alert-link">
              <i aria-hidden="true" class="icon" data-feather="alert-circle" style="border-radius: 5px"></i>
              Beta
            </a>
            <span class="visually-hidden">This lesson is in the beta phase, which means that it is ready for teaching by instructors outside of the original author team.</span>
          </abbr>
        </span>

      </div>
    </div>
    <div class="selector-container">
      <div id="theme-selector">
        <li class="nav-item dropdown" id="theme-button-list">
          <button class="btn btn-link nav-link px-0 px-lg-2 dropdown-toggle d-flex align-items-center" id="bd-theme" type="button" aria-expanded="false" data-bs-toggle="dropdown" data-bs-display="static" aria-label="Toggle theme (auto)">
            <svg class="bi my-1 theme-icon-active"><use href="#circle-half"></use></svg><i data-feather="chevron-down"></i>
          </button>
          <ul class="dropdown-menu dropdown-menu-end" aria-labelledby="bd-theme-text"><li>
              <button type="button" class="btn dropdown-item d-flex align-items-center" data-bs-theme-value="light" aria-pressed="false">
                <svg class="bi me-2 theme-icon"><use href="#sun-fill"></use></svg>
                Light
                <svg class="bi ms-auto d-none"><use href="#check2"></use></svg></button>
            </li>
            <li>
              <button type="button" class="btn dropdown-item d-flex align-items-center" data-bs-theme-value="dark" aria-pressed="false">
                <svg class="bi me-2 theme-icon"><use href="#moon-stars-fill"></use></svg>
                Dark
                <svg class="bi ms-auto d-none"><use href="#check2"></use></svg></button>
            </li>
            <li>
              <button type="button" class="btn dropdown-item d-flex align-items-center active" data-bs-theme-value="auto" aria-pressed="true">
                <svg class="bi me-2 theme-icon"><use href="#circle-half"></use></svg>
                Auto
                <svg class="bi ms-auto d-none"><use href="#check2"></use></svg></button>
            </li>
          </ul></li>
      </div>

      <div class="dropdown" id="instructor-dropdown">
        <button class="btn btn-secondary dropdown-toggle bordered-button" type="button" id="dropdownMenu1" data-bs-toggle="dropdown" aria-expanded="false">
          <i aria-hidden="true" class="icon" data-feather="eye"></i> Learner View <i data-feather="chevron-down"></i>
        </button>
        <ul class="dropdown-menu" aria-labelledby="dropdownMenu1"><li><button class="dropdown-item" type="button" onclick="window.location.href='instructor/apply-normalization-methods.html';">Instructor View</button></li>
        </ul></div>
    </div>
  </div>
  <hr></header><nav class="navbar navbar-expand-xl bottom-nav incubator" aria-label="Main Navigation"><div class="container-fluid nav-container">
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbarSupportedContent" aria-controls="navbarSupportedContent" aria-expanded="false" aria-label="Toggle Navigation">
      <span class="navbar-toggler-icon"></span>
      <span class="menu-title">Menu</span>
    </button>
    <div class="nav-logo">
      <img class="small-logo" alt="Carpentries Incubator" src="assets/images/incubator-logo-sm.svg"></div>
    <div class="lesson-title-md">
      Spatial Transcriptomics
    </div>
    <div class="search-icon-sm">
      <!-- TODO: do not show until we have search
        <i role="img" aria-label="Search the All In One page" data-feather="search"></i>
      -->
    </div>
    <div class="desktop-nav">
      <ul class="navbar-nav me-auto mb-2 mb-lg-0"><li class="nav-item">
          <span class="lesson-title">
            Spatial Transcriptomics
          </span>
        </li>
        <li class="nav-item">
          <a class="nav-link" href="key-points.html">Key Points</a>
        </li>
        <li class="nav-item">
          <a class="nav-link" href="reference.html#glossary">Glossary</a>
        </li>
        <li class="nav-item">
          <a class="nav-link" href="profiles.html">Learner Profiles</a>
        </li>
        <li class="nav-item dropdown">
          <button class="nav-link dropdown-toggle" id="navbarDropdown" data-bs-toggle="dropdown" aria-expanded="false">
            More <i data-feather="chevron-down"></i>
          </button>
          <ul class="dropdown-menu" aria-labelledby="navbarDropdown"><li><a class="dropdown-item" href="reference.html">Reference</a></li>
          </ul></li>
      </ul></div>
    <!--
    <form class="d-flex col-md-2 search-form">
      <fieldset disabled>
      <input class="form-control me-2 searchbox" type="search" placeholder="" aria-label="">
        <button class="btn btn-outline-success tablet-search-button"  type="submit">
          <i class="search-icon" data-feather="search" role="img" aria-label="Search the All In One page"></i>
        </button>
      </fieldset>
    </form>
    -->
    <a id="search-button" class="btn btn-primary" href="aio.html" role="button" aria-label="Search the All In One page">Search the All In One page</a>
  </div><!--/div.container-fluid -->
</nav><div class="col-md-12 mobile-title">
  Spatial Transcriptomics
</div>

<aside class="col-md-12 lesson-progress"><div style="width: 39%" class="percentage">
    39%
  </div>
  <div class="progress incubator">
    <div class="progress-bar incubator" role="progressbar" style="width: 39%" aria-valuenow="39" aria-label="Lesson Progress" aria-valuemin="0" aria-valuemax="100">
    </div>
  </div>
</aside><div class="container">
      <div class="row">
        <!-- START: inst/pkgdown/templates/navbar.html -->
<div id="sidebar-col" class="col-lg-4">
  <div id="sidebar" class="sidebar">
      <nav aria-labelledby="flush-headingEleven"><button role="button" aria-label="close menu" alt="close menu" aria-expanded="true" aria-controls="sidebar" class="collapse-toggle" data-collapse="Collapse " data-episodes="Episodes ">
          <i class="search-icon" data-feather="x" role="img"></i>
        </button>
        <div class="sidebar-inner">
          <div class="row mobile-row" id="theme-row-mobile">
            <div class="col" id="theme-selector">
              <li class="nav-item dropdown" id="theme-button-list">
                <button class="btn btn-link nav-link px-0 px-lg-2 dropdown-toggle d-flex align-items-center" id="bd-theme" type="button" aria-expanded="false" data-bs-toggle="dropdown" data-bs-display="static" aria-label="Toggle theme (auto)">
                  <svg class="bi my-1 theme-icon-active"><use href="#circle-half"></use></svg><span class="d-lg-none ms-1" id="bd-theme-text">Toggle Theme</span>
                </button>
                <ul class="dropdown-menu dropdown-menu-right" aria-labelledby="bd-theme-text"><li>
                    <button type="button" class="btn dropdown-item d-flex align-items-center" data-bs-theme-value="light" aria-pressed="false">
                      <svg class="bi me-2 theme-icon"><use href="#sun-fill"></use></svg>
                      Light
                      <svg class="bi ms-auto d-none"><use href="#check2"></use></svg></button>
                  </li>
                  <li>
                    <button type="button" class="btn dropdown-item d-flex align-items-center" data-bs-theme-value="dark" aria-pressed="false">
                      <svg class="bi me-2 theme-icon"><use href="#moon-stars-fill"></use></svg>
                      Dark
                      <svg class="bi ms-auto d-none"><use href="#check2"></use></svg></button>
                  </li>
                  <li>
                    <button type="button" class="btn dropdown-item d-flex align-items-center active" data-bs-theme-value="auto" aria-pressed="true">
                      <svg class="bi me-2 theme-icon"><use href="#circle-half"></use></svg>
                      Auto
                      <svg class="bi ms-auto d-none"><use href="#check2"></use></svg></button>
                  </li>
                </ul></li>
            </div>
          </div>
          <div class="row mobile-row">
            <div class="col">
              <div class="sidenav-view-selector">
                <div class="accordion accordion-flush" id="accordionFlush9">
                  <div class="accordion-item">
                    <h2 class="accordion-header" id="flush-headingNine">
                      <button class="accordion-button collapsed" id="instructor" type="button" data-bs-toggle="collapse" data-bs-target="#flush-collapseNine" aria-expanded="false" aria-controls="flush-collapseNine">
                        <i id="eye" aria-hidden="true" class="icon" data-feather="eye"></i> Learner View
                      </button>
                    </h2>
                    <div id="flush-collapseNine" class="accordion-collapse collapse" aria-labelledby="flush-headingNine" data-bs-parent="#accordionFlush2">
                      <div class="accordion-body">
                        <a href="instructor/apply-normalization-methods.html">Instructor View</a>
                      </div>
                    </div>
                  </div><!--/div.accordion-item-->
                </div><!--/div.accordion-flush-->
              </div><!--div.sidenav-view-selector -->
            </div><!--/div.col -->

            <hr></div><!--/div.mobile-row -->

          <div class="accordion accordion-flush" id="accordionFlush11">
            <div class="accordion-item">

              <button id="chapters" class="accordion-button show" type="button" data-bs-toggle="collapse" data-bs-target="#flush-collapseEleven" aria-expanded="false" aria-controls="flush-collapseEleven">
                <h2 class="accordion-header chapters" id="flush-headingEleven">
                  EPISODES
                </h2>
              </button>
              <div id="flush-collapseEleven" class="accordion-collapse show collapse" aria-labelledby="flush-headingEleven" data-bs-parent="#accordionFlush11">

                <div class="accordion-body">
                  <div class="accordion accordion-flush" id="accordionFlush1">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading1">
        <a href="index.html">Summary and Setup</a>
    </div><!--/div.accordion-header-->

  </div><!--/div.accordion-item-->
</div><!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush2">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading2">
        <a href="introduction.html">1. Spatially Resolved Transcriptomics in Life Sciences Research</a>
    </div><!--/div.accordion-header-->

  </div><!--/div.accordion-item-->
</div><!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush3">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading3">
        <a href="data-and-study-design.html">2. Data and Study Design</a>
    </div><!--/div.accordion-header-->

  </div><!--/div.accordion-item-->
</div><!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush4">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading4">
        <a href="data-preprocessing.html">3. Data Preprocessing</a>
    </div><!--/div.accordion-header-->

  </div><!--/div.accordion-item-->
</div><!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush5">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading5">
        <a href="remove-low-quality-spots.html">4. Remove Low-quality Spots</a>
    </div><!--/div.accordion-header-->

  </div><!--/div.accordion-item-->
</div><!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlushcurrent">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-headingcurrent">
      <button class="accordion-button" type="button" data-bs-toggle="collapse" data-bs-target="#flush-collapsecurrent" aria-expanded="true" aria-controls="flush-collapsecurrent">
        <span class="visually-hidden">Current Chapter</span>
        <span class="current-chapter">
        5. Normalization in Spatial Transcriptomics
        </span>
      </button>
    </div><!--/div.accordion-header-->

    <div id="flush-collapsecurrent" class="accordion-collapse collapse show" aria-labelledby="flush-headingcurrent" data-bs-parent="#accordionFlushcurrent">
      <div class="accordion-body">
        <ul><li><a href="#understanding-normalization-in-spatial-transcriptomics">Understanding Normalization in Spatial Transcriptomics</a></li>
        </ul></div><!--/div.accordion-body-->
    </div><!--/div.accordion-collapse-->

  </div><!--/div.accordion-item-->
</div><!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush7">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading7">
        <a href="feature-selection-dimensionality-reduction-clustering.html">6. Feature Selection, Dimensionality Reduction, and Spot Clustering</a>
    </div><!--/div.accordion-header-->

  </div><!--/div.accordion-item-->
</div><!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush8">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading8">
        <a href="deconvolve-cell-types-in-a-spot.html">7. Deconvolution in Spatial Transcriptomics</a>
    </div><!--/div.accordion-header-->

  </div><!--/div.accordion-item-->
</div><!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush9">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading9">
        <a href="differential-expression-testing.html">8. Differential Expression Testing</a>
    </div><!--/div.accordion-header-->

  </div><!--/div.accordion-item-->
</div><!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush10">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading10">
        <a href="final-exercise.html">9. Putting it all Together</a>
    </div><!--/div.accordion-header-->

  </div><!--/div.accordion-item-->
</div><!--/div.accordion-flush-->

                </div>
              </div>
            </div>

            <hr class="half-width"><div class="accordion accordion-flush lesson-resources" id="accordionFlush12">
              <div class="accordion-item">
                <h2 class="accordion-header" id="flush-headingTwelve">
                  <button class="accordion-button collapsed" id="lesson-resources" type="button" data-bs-toggle="collapse" data-bs-target="#flush-collapseTwelve" aria-expanded="false" aria-controls="flush-collapseTwelve">
                    RESOURCES
                  </button>
                </h2>
                <div id="flush-collapseTwelve" class="accordion-collapse collapse" aria-labelledby="flush-headingTwelve" data-bs-parent="#accordionFlush12">
                  <div class="accordion-body">
                    <ul><li>
                        <a href="key-points.html">Key Points</a>
                      </li>
                      <li>
                        <a href="reference.html#glossary">Glossary</a>
                      </li>
                      <li>
                        <a href="profiles.html">Learner Profiles</a>
                      </li>
                      <li><a href="reference.html">Reference</a></li>
                    </ul></div>
                </div>
              </div>
            </div>
            <hr class="half-width lesson-resources"><a href="aio.html">See all in one page</a>


            <hr class="d-none d-sm-block d-md-none"><div class="d-grid gap-1">

            </div>
          </div><!-- /div.accordion -->
        </div><!-- /div.sidebar-inner -->
      </nav></div><!-- /div.sidebar -->
  </div><!-- /div.sidebar-col -->
<!-- END:   inst/pkgdown/templates/navbar.html-->

        <!-- START: inst/pkgdown/templates/content-instructor.html -->
  <div class="col-xl-8 col-lg-12 primary-content">
    <nav class="lesson-content mx-md-4" aria-label="Previous and Next Chapter"><!-- content for small screens --><div class="d-block d-sm-block d-md-none">
        <a class="chapter-link" href="remove-low-quality-spots.html"><i aria-hidden="true" class="small-arrow" data-feather="arrow-left"></i>Previous</a>
        <a class="chapter-link float-end" href="feature-selection-dimensionality-reduction-clustering.html">Next<i aria-hidden="true" class="small-arrow" data-feather="arrow-right"></i></a>
      </div>
      <!-- content for large screens -->
      <div class="d-none d-sm-none d-md-block">
        <a class="chapter-link" href="remove-low-quality-spots.html" rel="prev">
          <i aria-hidden="true" class="small-arrow" data-feather="arrow-left"></i>
          Previous: Remove Low-quality
        </a>
        <a class="chapter-link float-end" href="feature-selection-dimensionality-reduction-clustering.html" rel="next">
          Next: Feature Selection,...
          <i aria-hidden="true" class="small-arrow" data-feather="arrow-right"></i>
        </a>
      </div>
      <hr></nav><main id="main-content" class="main-content"><div class="container lesson-content">
        <h1>Normalization in Spatial Transcriptomics</h1>
        <p>Last updated on 2025-04-29 |

        <a href="https://github.com/smcclatchy/spatial-transcriptomics/edit/main/episodes/apply-normalization-methods.Rmd" class="external-link">Edit this page <i aria-hidden="true" data-feather="edit"></i></a></p>



        <div class="text-end">
          <button role="button" aria-pressed="false" tabindex="0" id="expand-code" class="pull-right" data-expand="Expand All Solutions " data-collapse="Collapse All Solutions "> Expand All Solutions <i aria-hidden="true" data-feather="plus"></i></button>
        </div>

        

<div class="overview card">
<h2 class="card-header">Overview</h2>
<div class="row g-0">
<div class="col-md-4">
<div class="card-body">
<div class="inner">
<h3 class="card-title">Questions</h3>
<ul><li>What technical and biological factors impact spatial transcriptomics
data?</li>
<li>How do these factors motivate the need for normalization?</li>
<li>What are popular normalization methods?</li>
<li>How do we assess the impact of normalization?</li>
</ul></div>
</div>
</div>
<div class="col-md-8">
<div class="card-body">
<div class="inner bordered">
<h3 class="card-title">Objectives</h3>
<ul><li>Understand and learn to apply popular normalization techniques, such
as CPM normalization, log normalization, and SCTransform.</li>
<li>Diagnose the impact of normalization.</li>
</ul></div>
</div>
</div>
</div>
</div>
<section><h2 class="section-heading" id="understanding-normalization-in-spatial-transcriptomics">Understanding Normalization in Spatial Transcriptomics<a class="anchor" aria-label="anchor" href="#understanding-normalization-in-spatial-transcriptomics"></a></h2>
<hr class="half-width"><p>As with scRNA-seq data, normalization is necessary to overcome two
technical artifacts in spatial transcriptomics:</p>
<ol style="list-style-type: decimal"><li>the difference in total counts across spots, and</li>
<li>the dependence of a gene’s expression variance on its expression
level.</li>
</ol><p>The number of total counts in a spot is termed its <em>library
size</em>. Since library sizes differ across spots, it will be difficult
to compare gene expression values between them in a meaningful way
because the denominator (total spot counts) is different in each spot.
On the other hand, different spots may contain different types of cells,
which may express differing numbers of transcripts. So there is a
balance between normalizing all spots to have the same total counts and
leaving some variation in total counts which may be due to the biology
of the tissue. In general, we want to be cautious that removing the
above technical artifacts may also obscure true biological
differences.</p>
<p>Regarding the second artifact, we will see below that the variance in
a gene’s expression scales with its expression. If we do not correct for
this effect, differentially expressed genes will be skewed towards the
high end of the expression spectrum. Hence, we seek to <em>stabilize the
variance</em> – <em>i.e.</em>, transform the data such that expression
variance is independent of mean expression.</p>
<p>In this lesson we will:</p>
<ul><li>Observe that total spots per spot are variable</li>
<li>Explore biological factors that contribute to that variability</li>
<li>See that gene expression variance is correlated with mean
expression</li>
<li>Apply three methods aimed at mitigating one or both of these
technical observations: counts per million (CPM) normalization, log
normalization, and Seurat’s SCTransform</li>
</ul><div class="section level3">
<h3 id="total-counts-per-spot-are-variable">Total Counts per Spot are Variable<a class="anchor" aria-label="anchor" href="#total-counts-per-spot-are-variable"></a></h3>
<p>Let’s first assess the variability in the total counts per spot.</p>
<p>The spots are arranged in columns in the data matrix. We will look at
the distribution of total counts per spot by summing the counts in each
column and making a histogram.</p>
<div class="codewrapper sourceCode" id="cb1">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Extract the raw counts (gene by spot matrix) from the Seurat object</span></span>
<span><span class="va">counts</span> <span class="op">&lt;-</span> <span class="fu">LayerData</span><span class="op">(</span><span class="va">filter_st</span>, layer <span class="op">=</span> <span class="st">'counts'</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Plot a histogram of the total counts (library size or sum across genes).</span></span>
<span><span class="co"># Note that this column sum is also encoded in the nCount_Spatial metadata</span></span>
<span><span class="co"># variable. We could have simply made a histogram of that variable.</span></span>
<span><span class="fu">hist</span><span class="op">(</span><span class="fu">colSums2</span><span class="op">(</span><span class="va">counts</span><span class="op">)</span>, breaks <span class="op">=</span> <span class="fl">100</span>, </span>
<span>     main <span class="op">=</span> <span class="st">"Histogram of Counts per Spot"</span><span class="op">)</span></span></code></pre>
</div>
<figure><img src="fig/apply-normalization-methods-rendered-unnamed-chunk-2-1.png" style="display: block; margin: auto;" class="figure mx-auto d-block"></figure><p>As you can see, the total counts per spot ranges cross four orders of
magnitude. Some of this may be due to the biology of the tissue,
<em>i.e.</em> some cells may express more transcripts. But some of this
may be due to technical issues. Let’s explore each of these two
considerations further.</p>
</div>
<div class="section level3">
<h3 id="sources-of-biological-variation-in-total-counts">Sources of Biological Variation in Total Counts<a class="anchor" aria-label="anchor" href="#sources-of-biological-variation-in-total-counts"></a></h3>
<p>Hematoxylin and Eosin (H&amp;E) staining is routinely performed on
tissue sections in the clinic for diagnosis. Hematoxylin stains nuclei
purple or blue and eosin stains cytoplasm and extracellular matrix pink.
Collectively, they elucidate tissue and cellular morphology, which can
guide interpretation of transcriptomic data. For example, observing high
RNA counts in a necrotic region, which should instead have fewer cells,
might suggest technical artifacts and, thus, indicate a need for
normalization.</p>
<p>Maynard and colleagues used the information encoded in the H&amp;E,
in particular cellular organization, morphology, and density, in
conjunction with expression data to annotate the six layers and the
white matter of the neocortex. Additionally, they applied standard image
processing techniques to the H&amp;E image to segment and count nuclei
under each spot. They provide this as metadata. Let’s load that layer
annotation and cell count metadata and add it to our Seurat object.</p>
<div class="codewrapper sourceCode" id="cb2">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Load the metadata provided by Maynard et al.</span></span>
<span><span class="va">spot_metadata</span> <span class="op">&lt;-</span> <span class="fu">read.table</span><span class="op">(</span><span class="st">"./data/spot-meta.tsv"</span>, sep<span class="op">=</span><span class="st">"\t"</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Subset the metadata (across all samples) to our sample</span></span>
<span><span class="va">spot_metadata</span> <span class="op">&lt;-</span> <span class="fu">subset</span><span class="op">(</span><span class="va">spot_metadata</span>, <span class="va">sample_name</span> <span class="op">==</span> <span class="fl">151673</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Format the metadata by setting rowname to the barcode (id) of each spot,</span></span>
<span><span class="co"># by ensuring that each spot in our data is represented in the metadata,</span></span>
<span><span class="co"># and by ordering the spots within the metadata consistently with the data.</span></span>
<span><span class="fu">rownames</span><span class="op">(</span><span class="va">spot_metadata</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="va">spot_metadata</span><span class="op">$</span><span class="va">barcode</span></span>
<span><span class="fu">stopifnot</span><span class="op">(</span><span class="fu">all</span><span class="op">(</span><span class="fu">Cells</span><span class="op">(</span><span class="va">filter_st</span><span class="op">)</span> <span class="op">%in%</span> <span class="fu">rownames</span><span class="op">(</span><span class="va">spot_metadata</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">spot_metadata</span> <span class="op">&lt;-</span> <span class="va">spot_metadata</span><span class="op">[</span><span class="fu">Cells</span><span class="op">(</span><span class="va">filter_st</span><span class="op">)</span>,<span class="op">]</span></span>
<span></span>
<span><span class="co"># Add the layer annotation (layer_guess) and cell count as </span></span>
<span><span class="co"># metadata to the Seurat object using AddMetaData.</span></span>
<span><span class="va">filter_st</span> <span class="op">&lt;-</span> <span class="fu">AddMetaData</span><span class="op">(</span>object <span class="op">=</span> <span class="va">filter_st</span>,</span>
<span>                         metadata <span class="op">=</span> <span class="va">spot_metadata</span><span class="op">[</span>, <span class="fu">c</span><span class="op">(</span><span class="st">"layer_guess"</span>, <span class="st">"cell_count"</span><span class="op">)</span>, drop<span class="op">=</span><span class="cn">FALSE</span><span class="op">]</span><span class="op">)</span></span></code></pre>
</div>
<p>Now, we can plot the layer annotations to understand the structure of
the tissue. We will use a simple wrapper,
<code>SpatialDimPlotColorSafe</code>, around the Seurat function
<code>SpatialDimPlot</code>. This is defined in
<code>code/spatial_utils.R</code> and uses a color-blind safe
palette.</p>
<div class="codewrapper sourceCode" id="cb3">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Plot the layer annotations on the tissue, omitting any spots</span></span>
<span><span class="co"># that do not have annotations (*i.e.*, having NA values)</span></span>
<span><span class="fu">SpatialDimPlotColorSafe</span><span class="op">(</span><span class="va">filter_st</span><span class="op">[</span>, <span class="op">!</span><span class="fu">is.na</span><span class="op">(</span><span class="va">filter_st</span><span class="op">[[</span><span class="op">]</span><span class="op">]</span><span class="op">$</span><span class="va">layer_guess</span><span class="op">)</span><span class="op">]</span>,</span>
<span>                        <span class="st">"layer_guess"</span><span class="op">)</span> <span class="op">+</span> <span class="fu">labs</span><span class="op">(</span>fill<span class="op">=</span><span class="st">"Layer"</span><span class="op">)</span></span></code></pre>
</div>
<figure><img src="fig/apply-normalization-methods-rendered-unnamed-chunk-4-1.png" style="display: block; margin: auto;" class="figure mx-auto d-block"></figure><p>We noted that the authors used cellular density to aid in discerning
layers. Let’s see how those H&amp;E-derived cell counts vary across
layers.</p>
<div class="codewrapper sourceCode" id="cb4">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Make a boxplot of spot-level cell counts, faceted by layer annotation.</span></span>
<span><span class="co"># As above, remove any spots without annotations (*i.e.*, having NA values).</span></span>
<span><span class="va">g</span> <span class="op">&lt;-</span> <span class="fu">ggplot</span><span class="op">(</span><span class="fu">na.omit</span><span class="op">(</span><span class="va">filter_st</span><span class="op">[[</span><span class="op">]</span><span class="op">]</span><span class="op">[</span>, <span class="fu">c</span><span class="op">(</span><span class="st">"layer_guess"</span>, <span class="st">"cell_count"</span><span class="op">)</span><span class="op">]</span><span class="op">)</span>, </span>
<span>            <span class="fu">aes</span><span class="op">(</span>x <span class="op">=</span> <span class="va">layer_guess</span>, y <span class="op">=</span> <span class="va">cell_count</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">g</span> <span class="op">&lt;-</span> <span class="va">g</span> <span class="op">+</span> <span class="fu">geom_boxplot</span><span class="op">(</span><span class="op">)</span> <span class="op">+</span> <span class="fu">xlab</span><span class="op">(</span><span class="st">"Layer"</span><span class="op">)</span> <span class="op">+</span> <span class="fu">ylab</span><span class="op">(</span><span class="st">"Cell Count"</span><span class="op">)</span></span>
<span><span class="va">g</span></span></code></pre>
</div>
<figure><img src="fig/apply-normalization-methods-rendered-unnamed-chunk-5-1.png" style="display: block; margin: auto;" class="figure mx-auto d-block"></figure><p>We see that the white matter (WM) has increased cells per spot,
whereas Layer 1 has fewer cells per spot.</p>
<p>We can also plot these cell counts spatially.</p>
<div class="codewrapper sourceCode" id="cb5">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">SpatialFeaturePlot</span><span class="op">(</span><span class="va">filter_st</span>, <span class="st">"cell_count"</span><span class="op">)</span></span></code></pre>
</div>
<figure><img src="fig/apply-normalization-methods-rendered-unnamed-chunk-6-1.png" style="display: block; margin: auto;" class="figure mx-auto d-block"></figure><p>The cell counts partially reflect the banding of the layers.</p>
<p>As a potential surrogate for cell count, let’s plot the total counts
(number of UMIs or library size) per spot as a function of layer.</p>
<div class="codewrapper sourceCode" id="cb6">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Make a boxplot of spot-level total read counts (library size), faceted by layer annotation.</span></span>
<span><span class="co"># Remove any spots without annotations (*i.e.*, having NA values).</span></span>
<span><span class="va">g</span> <span class="op">&lt;-</span> <span class="fu">ggplot</span><span class="op">(</span><span class="fu">na.omit</span><span class="op">(</span><span class="va">filter_st</span><span class="op">[[</span><span class="op">]</span><span class="op">]</span><span class="op">[</span>, <span class="fu">c</span><span class="op">(</span><span class="st">"layer_guess"</span>, <span class="st">"nCount_Spatial"</span><span class="op">)</span><span class="op">]</span><span class="op">)</span>, </span>
<span>            <span class="fu">aes</span><span class="op">(</span>x <span class="op">=</span> <span class="va">layer_guess</span>, y <span class="op">=</span> <span class="va">nCount_Spatial</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">g</span> <span class="op">&lt;-</span> <span class="va">g</span> <span class="op">+</span> <span class="fu">geom_boxplot</span><span class="op">(</span><span class="op">)</span></span>
<span><span class="va">g</span></span></code></pre>
</div>
<figure><img src="fig/apply-normalization-methods-rendered-unnamed-chunk-7-1.png" style="display: block; margin: auto;" class="figure mx-auto d-block"></figure><p>Layer 1 has fewer total read counts, consistent with its lower cell
count. An increase in total read counts consistent with that in cell
count is not observed for the white matter, however. Regardless, there
are clear differences in total read counts across brain layers.</p>
<p>In summary, we have observed that both total read counts (library
size) and feature counts (number of detected genes) can encode
biological information. As such, we strongly recommend visualizing raw
gene and features counts prior to normalization, which would remove
differences in library size across spots.</p>
</div>
<div class="section level3">
<h3 id="normalization-techniques-to-mitigate-sources-of-technical-variation-in-total-counts">Normalization Techniques to Mitigate Sources of Technical Variation
in Total Counts<a class="anchor" aria-label="anchor" href="#normalization-techniques-to-mitigate-sources-of-technical-variation-in-total-counts"></a></h3>
<div class="section level4">
<h4 id="counts-per-million-library-size-normalization">“Counts Per Million” Library Size Normalization<a class="anchor" aria-label="anchor" href="#counts-per-million-library-size-normalization"></a></h4>
<p>The first technical issue we noted above was a difference in total
counts or library size across spots. A straightfoward means of
addressing this is simply to divide all gene counts within the spot by
the total counts in that spot. Conventionally, we then multiply by a
million, which yields “counts per million” (CPM). Adopting this
particular factor consistently establishes a standard scale across
studies.</p>
<p>You should be aware, however, that the CPM approach is susceptible to
“compositional bias” – if a small number of genes make a large
contribution to the total count, any significant fluctuation in their
expression across samples will impact the quantification of <em>all
other</em> genes. To overcome this, more robust measures of library size
that are more resilient to compositional bias are sometimes used,
including the 75th percentile of counts within a sample (or here, spot).
For simplicity, here we will use CPM.</p>
<p>In Seurat, we can apply this transformation via the
<code>NormalizeData</code> function, parameterized by the relative
counts (or “RC”) normalization method. We scale the results to a million
cells through the <code>scale.factor</code> parameter.</p>
<div class="codewrapper sourceCode" id="cb7">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Apply CPM normalization using NormalizeData, by specifying the relative</span></span>
<span><span class="co"># counts (RC) normalization.method and a scale factor of one million.</span></span>
<span><span class="va">cpm_st</span> <span class="op">&lt;-</span> <span class="fu">NormalizeData</span><span class="op">(</span><span class="va">filter_st</span>, </span>
<span>                        assay                <span class="op">=</span> <span class="st">"Spatial"</span>, </span>
<span>                        normalization.method <span class="op">=</span> <span class="st">"RC"</span>, </span>
<span>                        scale.factor         <span class="op">=</span> <span class="fl">1e6</span><span class="op">)</span></span></code></pre>
</div>
<p><code>NormalizeData</code> adds a <code>data</code> object to the
Seurat object.</p>
<div class="codewrapper sourceCode" id="cb8">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Access the layers of the Seurat object</span></span>
<span><span class="fu">Layers</span><span class="op">(</span><span class="va">cpm_st</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>[1] "counts" "data"  </code></pre>
</div>
<p>We can confirm that we have indeed normalized away differences in
total counts – all spots now have one million reads:</p>
<div class="codewrapper sourceCode" id="cb10">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Examine the total counts in each spot, as the sum of the columns.</span></span>
<span><span class="co"># As above, we could have also used nCount_Spatial in the metadata.</span></span>
<span><span class="fu">head</span><span class="op">(</span><span class="fu">colSums</span><span class="op">(</span><span class="fu">LayerData</span><span class="op">(</span><span class="va">cpm_st</span>, <span class="st">"data"</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>AAACAAGTATCTCCCA-1 AAACAATCTACTAGCA-1 AAACACCAATAACTGC-1 AAACAGAGCGACTCCT-1
             1e+06              1e+06              1e+06              1e+06
AAACAGCTTTCAGAAG-1 AAACAGGGTCTATATT-1
             1e+06              1e+06 </code></pre>
</div>
<p>Our second concern was that variance might differ across genes in an
expression dependent manner. To diagnose this, we will make a so-called
mean-variance plot, with each gene’s mean expression across spots on the
x axis and its variance across spots on the y axis. This shows any
potential trends between each gene’s mean expression and the variance of
that expression.</p>
<div class="codewrapper sourceCode" id="cb12">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Extract the CPM data computed above</span></span>
<span><span class="va">cpms</span>      <span class="op">&lt;-</span> <span class="fu">LayerData</span><span class="op">(</span><span class="va">cpm_st</span>, <span class="st">"data"</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Calculate the mean and variance of the CPMs</span></span>
<span><span class="va">means</span>     <span class="op">&lt;-</span> <span class="fu">apply</span><span class="op">(</span><span class="va">cpms</span>, <span class="fl">1</span>, <span class="va">mean</span><span class="op">)</span></span>
<span><span class="va">vars</span>      <span class="op">&lt;-</span> <span class="fu">apply</span><span class="op">(</span><span class="va">cpms</span>, <span class="fl">1</span>, <span class="va">var</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Assemble the mean and variance into a data.frame</span></span>
<span><span class="va">gene.info</span> <span class="op">&lt;-</span> <span class="fu">data.frame</span><span class="op">(</span>mean <span class="op">=</span> <span class="va">means</span>, variance <span class="op">=</span> <span class="va">vars</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Plot the mean expression on the x axis and the variance in expression on </span></span>
<span><span class="co"># the y axis</span></span>
<span><span class="va">g</span> <span class="op">&lt;-</span> <span class="fu">ggplot</span><span class="op">(</span><span class="op">)</span> <span class="op">+</span> <span class="fu">geom_point</span><span class="op">(</span>data <span class="op">=</span> <span class="va">gene.info</span>, <span class="fu">aes</span><span class="op">(</span>x <span class="op">=</span> <span class="va">mean</span>, y <span class="op">=</span> <span class="va">variance</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co"># Log transform both axes</span></span>
<span><span class="va">g</span> <span class="op">&lt;-</span> <span class="va">g</span> <span class="op">+</span> <span class="fu">scale_x_continuous</span><span class="op">(</span>trans<span class="op">=</span><span class="st">'log2'</span><span class="op">)</span> <span class="op">+</span> <span class="fu">scale_y_continuous</span><span class="op">(</span>trans<span class="op">=</span><span class="st">'log2'</span><span class="op">)</span></span>
<span><span class="va">g</span> <span class="op">&lt;-</span> <span class="va">g</span> <span class="op">+</span> <span class="fu">xlab</span><span class="op">(</span><span class="st">"Log Mean Expression"</span><span class="op">)</span> <span class="op">+</span> <span class="fu">ylab</span><span class="op">(</span><span class="st">"Log Mean Variance"</span><span class="op">)</span></span>
<span><span class="va">g</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">WARNING<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="warning" tabindex="0"><code>Warning in scale_x_continuous(trans = "log2"): log-2 transformation introduced
infinite values.</code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">WARNING<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="warning" tabindex="0"><code>Warning in scale_y_continuous(trans = "log2"): log-2 transformation introduced
infinite values.</code></pre>
</div>
<figure><img src="fig/apply-normalization-methods-rendered-unnamed-chunk-11-1.png" style="display: block; margin: auto;" class="figure mx-auto d-block"></figure><p>There is a clear relationship between the mean and variance of gene
expression. Our goal was instead that the variance be independent of the
mean. One way of achieving this is to <em>detrend</em> the data by
fitting a smooth curve to the mean-variance plot. This fit will capture
the general behavior of most genes. And, since we expect most genes to
exhibit technical variability only and not biological variability
additionally, this trend will reflect technical variance. Let’s start by
characterizing the trend in the data by fitting them with a smooth line.
We will use LOESS (locally estimated scatterplot smoothing) regression,
an approach originally developed to fit a smooth line through data
points in a scatterplot.</p>
<div class="codewrapper sourceCode" id="cb15">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># This is the default LOESS span used by Seurat in FindVariableFeatures</span></span>
<span><span class="va">loess.span</span> <span class="op">&lt;-</span> <span class="fl">0.3</span></span>
<span></span>
<span><span class="co"># Exclude genes with constant variance from our fit.</span></span>
<span><span class="va">not.const</span>  <span class="op">&lt;-</span> <span class="va">gene.info</span><span class="op">$</span><span class="va">variance</span> <span class="op">&gt;</span> <span class="fl">0</span></span>
<span></span>
<span><span class="co"># Fit a LOESS trend line relating the (log10) gene expression variance</span></span>
<span><span class="co"># to the (log10) gene expression mean, but only for the non-constant</span></span>
<span><span class="co"># variance genes.</span></span>
<span><span class="va">fit</span> <span class="op">&lt;-</span> <span class="fu">loess</span><span class="op">(</span>formula <span class="op">=</span> <span class="fu">log10</span><span class="op">(</span>x <span class="op">=</span> <span class="va">variance</span><span class="op">)</span> <span class="op">~</span> <span class="fu">log10</span><span class="op">(</span>x <span class="op">=</span> <span class="va">mean</span><span class="op">)</span>,</span>
<span>             data <span class="op">=</span> <span class="va">gene.info</span><span class="op">[</span><span class="va">not.const</span>, <span class="op">]</span>, span <span class="op">=</span> <span class="va">loess.span</span><span class="op">)</span></span></code></pre>
</div>
<p>Let’s now plot the fitted/expected variances as a function of the
observed means.</p>
<div class="codewrapper sourceCode" id="cb16">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># The expected variance computed from the model are in fit$fitted.</span></span>
<span><span class="co"># Exponentiate because the original model was fit to log10-transformed means and variances.</span></span>
<span><span class="va">gene.info</span><span class="op">$</span><span class="va">variance.expected</span>               <span class="op">&lt;-</span> <span class="cn">NA</span></span>
<span><span class="va">gene.info</span><span class="op">[</span><span class="va">not.const</span>, <span class="st">"variance.expected"</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="fl">10</span><span class="op">^</span><span class="va">fit</span><span class="op">$</span><span class="va">fitted</span></span>
<span></span>
<span><span class="co"># Plot the expected variance as a function of the observed means for only</span></span>
<span><span class="co"># the non-constant variance genes.</span></span>
<span><span class="va">g</span> <span class="op">&lt;-</span> <span class="va">g</span> <span class="op">+</span> <span class="fu">geom_line</span><span class="op">(</span>data <span class="op">=</span> <span class="fu">na.omit</span><span class="op">(</span><span class="va">gene.info</span><span class="op">[</span><span class="va">not.const</span>,<span class="op">]</span><span class="op">)</span>,</span>
<span>                   <span class="fu">aes</span><span class="op">(</span>x <span class="op">=</span> <span class="va">mean</span>, y <span class="op">=</span> <span class="va">variance.expected</span><span class="op">)</span>, linewidth <span class="op">=</span> <span class="fl">3</span>, color <span class="op">=</span> <span class="st">"blue"</span><span class="op">)</span>            </span>
<span><span class="va">g</span>             </span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">WARNING<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="warning" tabindex="0"><code>Warning in scale_x_continuous(trans = "log2"): log-2 transformation introduced
infinite values.</code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">WARNING<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="warning" tabindex="0"><code>Warning in scale_y_continuous(trans = "log2"): log-2 transformation introduced
infinite values.</code></pre>
</div>
<figure><img src="fig/apply-normalization-methods-rendered-unnamed-chunk-13-1.png" style="display: block; margin: auto;" class="figure mx-auto d-block"></figure><p>We see that the trend line nicely fits the data – <em>i.e.</em>, it
characterizes the observed variance as a function of the mean for the
vast majority of genes.</p>
<p>We now follow the logic applied in Seurat, as described by <a href="https://genomebiology.biomedcentral.com/articles/10.1186/s13059-019-1874-1" class="external-link">Hafemeister
and Satija</a>. We define the “standardized variance” as the variance in
the expression values after those expression values have been
standardized by the trend line (<em>i.e.</em>, they have been mean
centered and divided by the predicted variance).</p>
<p>As we did above, we can use a mean-variance plot to diagnose whether
our transformation, here the standardized variance, has indeed
stabilized the variance across different mean expression ranges. In
particular, since we expect <em>most</em> genes to have only technical
(and not also biological) variance, the trend line should be dominated
by those genes and will capture the technical variance. As such, the
standardized variance should be near one for most genes – those with
only technical variance. Genes with additional biological variance will
deviate from the trend line and also from the near-one standardized
variance. Is this what we observe?</p>
<div class="codewrapper sourceCode" id="cb19">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">gene.info</span><span class="op">$</span><span class="va">variance.standardized</span> <span class="op">&lt;-</span> <span class="cn">NA</span></span>
<span></span>
<span><span class="co"># Compute standardized CPM = ( CPM - mean_CPM ) / sqrt(expected_variance)</span></span>
<span><span class="va">standardized.cpms</span> <span class="op">&lt;-</span> </span>
<span>   <span class="op">(</span><span class="va">cpms</span><span class="op">[</span><span class="va">not.const</span>,<span class="op">]</span> <span class="op">-</span> <span class="va">means</span><span class="op">[</span><span class="va">not.const</span><span class="op">]</span><span class="op">)</span> <span class="op">/</span> <span class="fu">sqrt</span><span class="op">(</span><span class="va">gene.info</span><span class="op">[</span><span class="va">not.const</span>, <span class="st">"variance.expected"</span><span class="op">]</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Calculate the "standardized variance" -- *i.e.*, the variance of the standardized CPMs</span></span>
<span><span class="va">gene.info</span><span class="op">[</span><span class="va">not.const</span>, <span class="st">"variance.standardized"</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="fu">apply</span><span class="op">(</span><span class="va">standardized.cpms</span>, <span class="fl">1</span>, <span class="va">var</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Plot the standardized variance for the non-constant variance genes</span></span>
<span><span class="va">g</span> <span class="op">&lt;-</span> <span class="fu">ggplot</span><span class="op">(</span><span class="op">)</span> <span class="op">+</span> <span class="fu">geom_point</span><span class="op">(</span>data <span class="op">=</span> <span class="va">gene.info</span><span class="op">[</span><span class="va">not.const</span>,<span class="op">]</span>, <span class="fu">aes</span><span class="op">(</span>x <span class="op">=</span> <span class="va">mean</span>, y <span class="op">=</span> <span class="va">variance.standardized</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">g</span> <span class="op">&lt;-</span> <span class="va">g</span> <span class="op">+</span> <span class="fu">geom_hline</span><span class="op">(</span>yintercept <span class="op">=</span> <span class="fl">1</span>, colour <span class="op">=</span> <span class="st">"yellow"</span><span class="op">)</span></span>
<span><span class="co"># Log10-transform the x axis</span></span>
<span><span class="va">g</span> <span class="op">&lt;-</span> <span class="va">g</span> <span class="op">+</span> <span class="fu">scale_x_log10</span><span class="op">(</span><span class="op">)</span></span>
<span><span class="va">g</span> <span class="op">&lt;-</span> <span class="va">g</span> <span class="op">+</span> <span class="fu">xlab</span><span class="op">(</span><span class="st">"Log Mean CPM"</span><span class="op">)</span> <span class="op">+</span> <span class="fu">ylab</span><span class="op">(</span><span class="st">"Variance of\nTrend-Standardized CPMs"</span><span class="op">)</span></span>
<span><span class="va">g</span></span></code></pre>
</div>
<figure><img src="fig/apply-normalization-methods-rendered-unnamed-chunk-14-1.png" style="display: block; margin: auto;" class="figure mx-auto d-block"></figure><p>Indeed, many genes do have a variance near one.</p>
<p>Seurat provides more efficient means for calculating expression
means, variances, variances predicted from the trend, and the variance
of means standardized by the predicted variances, namely <a href="https://satijalab.org/seurat/reference/findvariablefeatures" class="external-link"><code>FindVariableFeatures</code></a>.
The computed results can then be formatted into a data frame using <a href="https://satijalab.org/seurat/reference/hvfinfo.sctassay" class="external-link"><code>HVFInfo</code></a>.
Be advised that these functions can calculate expected variances using
models that are dependent on input parameters and, sometimes implicitly,
based on different data formats (e.g., raw counts or normalized
expression values). In this case, the values we computed above
manually:</p>
<div class="codewrapper sourceCode" id="cb20">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">head</span><span class="op">(</span><span class="va">gene.info</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>                  mean    variance variance.expected variance.standardized
MIR1302-2HG 0.04896898    8.711791          7.998485             1.0891801
AL627309.1  0.71174425  141.672131        212.427026             0.6669214
AL669831.5  5.45042760 1452.565233       1698.941124             0.8549827
FAM87B      0.09675030   34.007136         21.967802             1.5480446
LINC00115   0.81579878  188.767639        245.068894             0.7702636
FAM41C      1.36714334  338.634811        418.031204             0.8100707</code></pre>
</div>
<p>are the same, with few exceptions, to those computed by the
equivalent Seurat functions:</p>
<div class="codewrapper sourceCode" id="cb22">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Compute expression means and variances using FindVariableFeatures.</span></span>
<span><span class="co"># By passing the total number of genes as nfeatures, we force FindVariableFeatures to</span></span>
<span><span class="co"># compute these metrics for all genes, not just the most variable ones.</span></span>
<span><span class="co"># The metrics will be computed on the specified layer ("data") of the active assay -- here, the</span></span>
<span><span class="co"># CPMs we computed above.</span></span>
<span><span class="co"># Finally, we will use the "vst" (variance-stabilizing transformation) method for selecting highly</span></span>
<span><span class="co"># variable genes.</span></span>
<span><span class="va">cpm_st</span> <span class="op">&lt;-</span> <span class="fu">FindVariableFeatures</span><span class="op">(</span><span class="va">cpm_st</span>, nfeatures <span class="op">=</span> <span class="fu">dim</span><span class="op">(</span><span class="va">cpm_st</span><span class="op">)</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span>, layer<span class="op">=</span><span class="st">"data"</span>, selection.method <span class="op">=</span> <span class="st">"vst"</span><span class="op">)</span></span>
<span><span class="va">cinfo</span> <span class="op">&lt;-</span> <span class="fu">HVFInfo</span><span class="op">(</span><span class="va">cpm_st</span><span class="op">)</span></span>
<span><span class="fu">head</span><span class="op">(</span><span class="va">cinfo</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>                  mean    variance variance.expected variance.standardized
MIR1302-2HG 0.04896898    8.711791          7.998485             1.0005751
AL627309.1  0.71174425  141.672131        212.427026             0.6669214
AL669831.5  5.45042760 1452.565233       1698.941124             0.8549827
FAM87B      0.09675030   34.007136         21.967802             1.0007014
LINC00115   0.81579878  188.767639        245.068894             0.7702636
FAM41C      1.36714334  338.634811        418.031204             0.8100707</code></pre>
</div>
<p>The above code applies the “vst” or variance-stabilizing
transformation to detect highly variable genes. This is essentially the
loess fitting and standardized variance computation that we performed
above. More details are available in the FindVariableFeatures
documentation. We can verify that this FindVariableFeatures functions
gives us similar results to those we manually computed above by using <a href="https://satijalab.org/seurat/reference/variablefeatureplot" class="external-link"><code>VariableFeaturePlot</code></a>
to plot the relationship between the mean expression and the variance of
the trend-standardized expression.</p>
<div class="codewrapper sourceCode" id="cb24">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">VariableFeaturePlot</span><span class="op">(</span><span class="va">cpm_st</span><span class="op">)</span> <span class="op">+</span> <span class="fu">NoLegend</span><span class="op">(</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">WARNING<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="warning" tabindex="0"><code>Warning in scale_x_log10(): log-10 transformation introduced infinite values.</code></pre>
</div>
<figure><img src="fig/apply-normalization-methods-rendered-unnamed-chunk-17-1.png" style="display: block; margin: auto;" class="figure mx-auto d-block"></figure><p>Notice that this plot is similar to that we created manually.</p>
<p>In both our manual plot and the one created by
FindVariableFeatures/VariableFeaturePlot the “standardized” variance of
most genes is near one. Achieving such a flat trend line is our
objective and a good indication that we have stabilized the variance
across a wide range of expression values. Yet, in both, there remains an
evident, if subtle, trend in the plot between mean and variance. Let’s
next look at two common transformations aimed at further stabilizing the
variance across genes – <em>i.e.</em>, for further mitigating the
relationship between gene expression and variance – log-normalization
and SCTransform.</p>
</div>
<div class="section level4">
<h4 id="lognormalize">LogNormalize<a class="anchor" aria-label="anchor" href="#lognormalize"></a></h4>
<p>One common approach that attempts to meet our two objectives above –
normalizing for different total spot counts and for expression-dependent
variance – is log-transformation of normalized counts. The resulting
values are often ambiguously referred to as log-normalized counts, which
elides stating that the raw counts are first normalized or scaled and
then log transformed. Scaling accounts for the differences in
spot-specific RNA counts, as with CPMs above. The log transformation
reduces skewness caused by highly expressed genes and stabilizes the
variance, at least for certain mean-variance relationships. In practice,
the log transformation is applied to <i>1+x</i>, where <i>x</i> is the
scaled expression value – the so-called <em>log1p</em> transformation.
This avoids taking the log of zero.</p>
<p>In Seurat, we can apply this transformation by specifying
<code>LogNormalize</code> as the <code>normalization.method</code>
parameter to the same <code>NormalizeData</code> function we applied
above to compute CPMs.</p>
<div class="codewrapper sourceCode" id="cb26">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Apply log normalization using NormalizeData, by specifying the</span></span>
<span><span class="co"># LogNormalize normalization.method and a scale factor of one million.</span></span>
<span><span class="va">lognorm_st</span> <span class="op">&lt;-</span> <span class="fu">NormalizeData</span><span class="op">(</span><span class="va">filter_st</span>, </span>
<span>                           assay                <span class="op">=</span> <span class="st">"Spatial"</span>, </span>
<span>                           normalization.method <span class="op">=</span> <span class="st">"LogNormalize"</span>, </span>
<span>                           scale.factor         <span class="op">=</span> <span class="fl">1e6</span><span class="op">)</span></span></code></pre>
</div>
<p>This function first normalizes the raw counts by
<code>scale.factor</code> before applying the log1p transformation. As
above, we set the <code>scale.factor</code> parameter such that we first
compute CPMs and then apply log1p to these CPMs.</p>
<p>As above, log normalization adds a <code>data</code> object to the
Seurat object.</p>
<div class="codewrapper sourceCode" id="cb27">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Access the layers of the Seurat object</span></span>
<span><span class="fu">Layers</span><span class="op">(</span><span class="va">lognorm_st</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>[1] "counts" "data"  </code></pre>
</div>
<p>Let’s again plot the relationship between the expression mean and its
standardized variance. Additionally, we will highlight those genes that
have standardized variance significantly larger than one. These are
likely to exhibit biological variation.</p>
<div class="codewrapper sourceCode" id="cb29">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Compute highly variable genes and their expression means and variances using FindVariableFeatures.</span></span>
<span><span class="co"># Apply this to the log-transformed data, by specifying the "data" layer of the active assay.</span></span>
<span><span class="co"># As above, use the vst method for computing variable genes.</span></span>
<span><span class="va">lognorm_st</span> <span class="op">&lt;-</span> <span class="fu">FindVariableFeatures</span><span class="op">(</span><span class="va">lognorm_st</span>, layer<span class="op">=</span><span class="st">"data"</span>, selection.method <span class="op">=</span> <span class="st">"vst"</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Extract the top 15 most variable genes</span></span>
<span><span class="va">top15</span>        <span class="op">&lt;-</span> <span class="fu">head</span><span class="op">(</span><span class="fu">VariableFeatures</span><span class="op">(</span><span class="va">lognorm_st</span>, layer<span class="op">=</span><span class="st">"data"</span>, method<span class="op">=</span><span class="st">"vst"</span><span class="op">)</span>, <span class="fl">15</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Make a mean-variance plot of the log-transformed data</span></span>
<span><span class="va">plot_lognorm</span> <span class="op">&lt;-</span> <span class="fu">VariableFeaturePlot</span><span class="op">(</span><span class="va">lognorm_st</span><span class="op">)</span> <span class="op">+</span></span>
<span>                  <span class="fu">ggtitle</span><span class="op">(</span><span class="st">"Variable Features - Log normalization"</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Add the names of the top 15 most variable genes</span></span>
<span><span class="va">plot_lognorm</span> <span class="op">&lt;-</span> <span class="fu">LabelPoints</span><span class="op">(</span>plot <span class="op">=</span> <span class="va">plot_lognorm</span>, points <span class="op">=</span> <span class="va">top15</span>, repel <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="va">plot_lognorm</span></span></code></pre>
</div>
<figure><img src="fig/apply-normalization-methods-rendered-unnamed-chunk-20-1.png" style="display: block; margin: auto;" class="figure mx-auto d-block"></figure><p>As with the standardized CPM values above, we have significantly
reduced the dependence of variance in gene expression on its mean – at
least at all but the highest end of the expression. By default, Seurat
selects a set of 2,000 variable genes which are colored in red. Note
that a majority of the highly variable genes, and all of the top 15 most
variable that we labeled, are peaked in a narrow range of high
expression values. We would not expect this <em>a priori</em>, and it
signals inadequate normalization.</p>
<p>Putting this concern aside for a moment, let’s look at the spatial,
normalized expression of the two most variable genes.</p>
<div class="codewrapper sourceCode" id="cb30">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Seurat's nomenclature is inconsistent. Here slot is the same as layer.</span></span>
<span><span class="fu">SpatialFeaturePlot</span><span class="op">(</span><span class="va">lognorm_st</span>, slot<span class="op">=</span><span class="st">"data"</span>, <span class="fu">c</span><span class="op">(</span><span class="st">"SCGB2A2"</span>, <span class="st">"PLP1"</span><span class="op">)</span><span class="op">)</span></span></code></pre>
</div>
<figure><img src="fig/apply-normalization-methods-rendered-unnamed-chunk-21-1.png" style="display: block; margin: auto;" class="figure mx-auto d-block"></figure><p>Both clearly show significant spatial variability – despite the fact
that we did not explicitly query for genes whose expression varied
spatially. We will look at approaches for doing that later, such as the
Moran’s I statistic. <i>PLP1</i> shows coherent spatial variability
closely linked to brain morphology – this is a marker for the white
matter. <i>SCGB2A2</i> also shows strong spatial variability, though it
does not appear to reflect the underlying neocortical layers. It should
not surprise us that there is spatially variable gene expression beyond
that caused by the (known) layered architecture of the brain. This is
one such example.</p>
<p>As a further sanity check that the normalization is doing something
sensible, let’s look at the expression of two, known layer-restricted
marker genes – <i>MOBP</i> and <i>PCP4</i>. <i>MOBP</i> is known to be
restricted to the white matter, while <i>PCP4</i> is expressed in Layer
5.</p>
<div class="codewrapper sourceCode" id="cb31">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">SpatialFeaturePlot</span><span class="op">(</span><span class="va">lognorm_st</span>, slot<span class="op">=</span><span class="st">"data"</span>, <span class="fu">c</span><span class="op">(</span><span class="st">"MOBP"</span>, <span class="st">"PCP4"</span><span class="op">)</span><span class="op">)</span></span></code></pre>
</div>
<figure><img src="fig/apply-normalization-methods-rendered-unnamed-chunk-22-1.png" style="display: block; margin: auto;" class="figure mx-auto d-block"></figure><p>Indeed, this is what we observe.</p>
</div>
<div class="section level4">
<h4 id="sctransform">SCTransform<a class="anchor" aria-label="anchor" href="#sctransform"></a></h4>
<p>Let’s see if we can improve on two aspects of the above
log-normalization: 1) the non-uniform standardized variance of highly
expressed genes and 2) the bias of highly variable genes towards highly
expressed genes. We will apply <a href="https://satijalab.org/seurat/articles/sctransform_vignette.html" class="external-link"><code>SCTransform</code></a>,
a normalization approach that uses a regularized negative binomial
regression to stabilize variance across expression levels (<a href="https://link.springer.com/article/10.1186/s13059-021-02584-9" class="external-link">Choudhary
et al., Genome Biol 23, 27 (2022)</a>).</p>
<div class="codewrapper sourceCode" id="cb32">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Apply SCTransform to the raw count data in the Spatial assay</span></span>
<span><span class="va">sct_st</span> <span class="op">&lt;-</span> <span class="fu">SCTransform</span><span class="op">(</span><span class="va">filter_st</span>, assay <span class="op">=</span> <span class="st">"Spatial"</span><span class="op">)</span></span></code></pre>
</div>
<p>The <code>SCTransform</code> method added a new assay called
<code>SCT</code>, as we can see below by assessing the assays of the
Seurat object:</p>
<div class="codewrapper sourceCode" id="cb33">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Access assays of the Seurat object</span></span>
<span><span class="fu">Assays</span><span class="op">(</span><span class="va">sct_st</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>[1] "Spatial" "SCT"    </code></pre>
</div>
<p>It made this new assay the default. Be aware that Seurat functions
often operate on the <code>DefaultAssay</code>.</p>
<div class="codewrapper sourceCode" id="cb35">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Access the default assay of the Seurat object</span></span>
<span><span class="fu">DefaultAssay</span><span class="op">(</span><span class="va">sct_st</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>[1] "SCT"</code></pre>
</div>
<p>Within this new <code>SCT</code> Assay, <code>SCTransform</code> has
created three <code>Layers</code> to store data, not to be confused with
the neocortical layers of the brain.</p>
<div class="codewrapper sourceCode" id="cb37">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Assess the layers of the default assay of the Seurat object</span></span>
<span><span class="fu">Layers</span><span class="op">(</span><span class="va">sct_st</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>[1] "counts"     "data"       "scale.data"</code></pre>
</div>
<p>As you can see by reading the <code>SCTransform</code> documentation
with</p>
<div class="codewrapper sourceCode" id="cb39">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="op">?</span><span class="va">SCTransform</span></span></code></pre>
</div>
<p>these new <code>Layers</code> are <code>counts</code> (counts
corrected for differences in sequencing depth between cells),
<code>data</code> (<code>log1p</code> transformation of the corrected
counts), and <code>scale.data</code> (scaled Pearson residuals,
<em>i.e.</em>, the difference between an observed count and its expected
value under the model used by <code>SCTransform</code>, divided by the
standard deviation in that count under the model).</p>
<figure><img src="./fig/seurat_object_assays_layers.png" alt="A diagram of the sttructure of a Seurat object showing the assays (Spatial &amp; SCT) and Layers (counts, data &amp; scale.data." class="figure mx-auto d-block"><div class="figcaption">Seurat object structure</div>
</figure><p>Notice, in particular, that the <code>counts</code> Layers in the
<code>Spatial</code> and <code>SCT</code> Assays are different. As
mentioned above, the latter have been corrected for differences in
sequencing depth between cells. As such, the distribution in total
counts per cell is much more uniform in the latter case.</p>
<div class="codewrapper sourceCode" id="cb40">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Create a one-column plot with two panels</span></span>
<span><span class="fu">layout</span><span class="op">(</span><span class="fu">matrix</span><span class="op">(</span><span class="fl">1</span><span class="op">:</span><span class="fl">2</span>, ncol <span class="op">=</span> <span class="fl">1</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Extract the raw count data from the counts layer of the Spatial assay</span></span>
<span><span class="va">raw_counts_spatial</span> <span class="op">&lt;-</span> <span class="fu">LayerData</span><span class="op">(</span><span class="va">sct_st</span>, layer <span class="op">=</span> <span class="st">"counts"</span>, assay <span class="op">=</span> <span class="st">"Spatial"</span><span class="op">)</span></span>
<span><span class="co"># Plot a histogram of the total counts (column sum) of the raw counts</span></span>
<span><span class="fu">hist</span><span class="op">(</span><span class="fu">colSums2</span><span class="op">(</span><span class="va">raw_counts_spatial</span><span class="op">)</span>, main <span class="op">=</span> <span class="st">"Raw counts (Spatial)"</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Extract the corrected counts from the data layer of the SCT assay</span></span>
<span><span class="va">corrected_counts_sct</span> <span class="op">&lt;-</span> <span class="fu">LayerData</span><span class="op">(</span><span class="va">sct_st</span>, layer <span class="op">=</span> <span class="st">"counts"</span>, assay <span class="op">=</span> <span class="st">"SCT"</span><span class="op">)</span></span>
<span><span class="co"># Plot a histogram of the total corrected counts (column sum)</span></span>
<span><span class="fu">hist</span><span class="op">(</span><span class="fu">colSums2</span><span class="op">(</span><span class="va">corrected_counts_sct</span><span class="op">)</span>, main <span class="op">=</span> <span class="st">"Corrected counts (SCT)"</span><span class="op">)</span></span></code></pre>
</div>
<figure><img src="fig/apply-normalization-methods-rendered-unnamed-chunk-28-1.png" style="display: block; margin: auto;" class="figure mx-auto d-block"></figure><p>Let’s plot the mean-variance relationship. Notice that SCTransform
implicitly computes variable features, so we do not need to explicitly
call FindVariableFeatures, as we did above.</p>
<div class="codewrapper sourceCode" id="cb41">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Extract the top 15 most variable genes, as computed from the SCTransformed data.</span></span>
<span><span class="va">top15SCT</span>    <span class="op">&lt;-</span> <span class="fu">head</span><span class="op">(</span><span class="fu">VariableFeatures</span><span class="op">(</span><span class="va">sct_st</span><span class="op">)</span>, <span class="fl">15</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Make a mean-variance plot of the SCTransformed data.</span></span>
<span><span class="va">plot_sct</span>    <span class="op">&lt;-</span> <span class="fu">VariableFeaturePlot</span><span class="op">(</span><span class="va">sct_st</span><span class="op">)</span> <span class="op">+</span> </span>
<span>                 <span class="fu">ggtitle</span><span class="op">(</span><span class="st">"Variable Features - SCT"</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Add the names of the top 15 most variable genes.</span></span>
<span><span class="va">plot_sct</span>    <span class="op">&lt;-</span> <span class="fu">LabelPoints</span><span class="op">(</span>plot <span class="op">=</span> <span class="va">plot_sct</span>, points <span class="op">=</span> <span class="va">top15SCT</span>, repel <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="va">plot_sct</span></span></code></pre>
</div>
<figure><img src="fig/apply-normalization-methods-rendered-unnamed-chunk-29-1.png" style="display: block; margin: auto;" class="figure mx-auto d-block"></figure><p>Notice that VariableFeaturePlot outputs slightly different axes,
depending on the transformed data to display. In this case, for
SCTransformed data, it plots the geometric mean (mean of the log counts)
on the x axis and the “residual variance” on the y axis. “Residual
variance” is a confusing term; it refers to the variance in the Pearson
residuals, similar to the standardized variance we encountered
above.</p>
<p>We will compare this mean-variance plot to the one derived using
log-normalization below. Before we do, let’s look at the spatial
expression of the two most variable genes following SCTransform. One of
these, <i>PLP1</i>, we have already seen above.</p>
<div class="codewrapper sourceCode" id="cb42">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Plot the SCTransformed values (specifically, the log-transformed correct counts)</span></span>
<span><span class="co"># of the genes MBP and PLP1 spatially, by accessing the "data" slot (or layer). </span></span>
<span><span class="fu">SpatialFeaturePlot</span><span class="op">(</span><span class="va">sct_st</span>, slot<span class="op">=</span><span class="st">"data"</span>, <span class="fu">c</span><span class="op">(</span><span class="st">"MBP"</span>, <span class="st">"PLP1"</span><span class="op">)</span><span class="op">)</span></span></code></pre>
</div>
<figure><img src="fig/apply-normalization-methods-rendered-unnamed-chunk-30-1.png" style="display: block; margin: auto;" class="figure mx-auto d-block"></figure><p>Both genes have stunning spatially variable expresion. And, here,
that variability happens to reflect the underlying layered architecture
of the brain. As we noted above, it needn’t have been the case that
spatially variable genes are also layer restricted. That the top
variable genes following SCTransform are layer restricted, whereas only
one but not the other is following log normalization, should not in and
of itself imply that the former transformation is “better” than the
latter.</p>
<p>Let’s again check that the two marker genes show the appropriate
layer-restricted expression following normalization with
SCTransform.</p>
<div class="codewrapper sourceCode" id="cb43">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Plot the SCTransformed values (specifically, the log-transformed correct counts)</span></span>
<span><span class="co"># of the known marker genes MOBP and PCP4 spatially, by accessing the "data" slot (or layer). </span></span>
<span><span class="fu">SpatialFeaturePlot</span><span class="op">(</span><span class="va">sct_st</span>, slot<span class="op">=</span><span class="st">"data"</span>, <span class="fu">c</span><span class="op">(</span><span class="st">"MOBP"</span>, <span class="st">"PCP4"</span><span class="op">)</span><span class="op">)</span></span></code></pre>
</div>
<figure><img src="fig/apply-normalization-methods-rendered-unnamed-chunk-31-1.png" style="display: block; margin: auto;" class="figure mx-auto d-block"></figure><p>Indeed, they do. We conclude that SCTransform normalization looks
reasonable.</p>
<div id="challenge-1-compare-mean-variance-plots" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<div id="challenge-1-compare-mean-variance-plots" class="callout-inner">
<h3 class="callout-title">Challenge 1: Compare Mean-Variance Plots</h3>
<div class="callout-content">
<p>Above, we created mean-variance plots for the SCTransform and log
normalizations. Which method does a better job of stabilizing the
variance across genes? Turn to the person next to you and put the
log-normalized plot on one of your screens and the SCTransform plot on
the other person’s screen. Discuss the mean-variance relationship in
each plot and decide which one you think stabilizes the variance across
genes better.</p>
</div>
</div>
</div>
<div id="accordionSolution1" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution1" aria-expanded="false" aria-controls="collapseSolution1">
  <h4 class="accordion-header" id="headingSolution1"> Solution 1 </h4>
</button>
<div id="collapseSolution1" class="accordion-collapse collapse" data-bs-parent="#accordionSolution1" aria-labelledby="headingSolution1">
<div class="accordion-body">
<p>We will compare the two plots in the next section.</p>
</div>
</div>
</div>
</div>
</div>
</div>
<div class="section level3">
<h3 id="comparing-normalizations">Comparing Normalizations<a class="anchor" aria-label="anchor" href="#comparing-normalizations"></a></h3>

<p>First, let’s look at the sum of the normalized expression values per
spot for each method. Recall from above that <code>NormalizeData</code>
put the log-transformed values in the <code>data</code> layer.
Similarly, <code>SCTransform</code> puts the log1p transformed values of
the corrected counts in the <code>data</code> layer. So, we access that
layer below.</p>
<div class="codewrapper sourceCode" id="cb44">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Create a one-column plot with two panels</span></span>
<span><span class="fu">layout</span><span class="op">(</span><span class="fu">matrix</span><span class="op">(</span><span class="fl">1</span><span class="op">:</span><span class="fl">2</span>, ncol <span class="op">=</span> <span class="fl">1</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Extract the log-normalized data from the data layer of the lognorm_st Seurat object</span></span>
<span><span class="va">sum_log</span> <span class="op">&lt;-</span> <span class="fu">LayerData</span><span class="op">(</span><span class="va">lognorm_st</span>, layer <span class="op">=</span> <span class="st">"data"</span><span class="op">)</span></span>
<span><span class="co"># Plot a histogram of the column sum of the log-normalized data</span></span>
<span><span class="fu">hist</span><span class="op">(</span><span class="fu">colSums2</span><span class="op">(</span><span class="va">sum_log</span><span class="op">)</span>, main <span class="op">=</span> <span class="st">"Log-norm"</span>, xlab <span class="op">=</span> <span class="st">"Summed, Normalized Expression Values"</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Extract the SCTransformed (log-transformed corrected count) data from the data layer of the sct_st Seurat object</span></span>
<span><span class="va">sum_sct</span> <span class="op">&lt;-</span> <span class="fu">LayerData</span><span class="op">(</span><span class="va">sct_st</span>, layer <span class="op">=</span> <span class="st">"data"</span><span class="op">)</span></span>
<span><span class="co"># Plot a histogram of the column sum of the SCTransformed (log-transformed corrected count) data </span></span>
<span><span class="fu">hist</span><span class="op">(</span><span class="fu">colSums2</span><span class="op">(</span><span class="va">sum_sct</span><span class="op">)</span>, main <span class="op">=</span> <span class="st">"SCT"</span>, xlab <span class="op">=</span> <span class="st">"Summed, Normalized Expression Values"</span><span class="op">)</span></span></code></pre>
</div>
<figure><img src="fig/apply-normalization-methods-rendered-unnamed-chunk-32-1.png" style="display: block; margin: auto;" class="figure mx-auto d-block"></figure><p>Notice that the log-normalization has a range of summed expression
values per spot spanning several orders of magnitude. The SCTransform
has a more uniform distribution of summed normalized expression values
that spans a factor of three, from ~1000 to ~2700.</p>
<p>Next, we will compare the mean-variance plots between the two
methods.</p>
<div class="codewrapper sourceCode" id="cb45">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">plot_lognorm</span></span></code></pre>
</div>
<figure><img src="fig/apply-normalization-methods-rendered-unnamed-chunk-33-1.png" style="display: block; margin: auto;" class="figure mx-auto d-block"></figure><div class="codewrapper sourceCode" id="cb46">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">plot_sct</span></span></code></pre>
</div>
<figure><img src="fig/apply-normalization-methods-rendered-unnamed-chunk-33-2.png" style="display: block; margin: auto;" class="figure mx-auto d-block"></figure><p>Following SCTransform, the variance is largely stable across a range
of mean expression values. Certainly, we do not observe the wild ripple
in the mean-variance trend at the high expression end that we saw with
log-normalization. Secondly, the highly variable genes are spread across
a larger region and not overly biased towards the high end of the
expression spectrum.</p>
<p>We conclude that SCTransform is more appropriate for these data than
log normalization for those two reasons:</p>
<ul><li>The variance is stable throughout the entire range of mean
expression values, <em>i.e.</em>, the standardized variance is near one
throughout this range.</li>
<li>The highly variable genes, <em>i.e.</em>, those whose standardized
variance deviates from one, are spread throughout the expression
spectrum and not localized to highly expressed genes.</li>
</ul></div>
<div class="section level3">
<h3 id="no-one-size-fits-all-approach">No One-Size-Fits-All Approach<a class="anchor" aria-label="anchor" href="#no-one-size-fits-all-approach"></a></h3>
<p>Raw read counts provide essential insights into absolute cell
densities within a sample. In some cases, this may reflect
morphologically distinct regions and / or regions enriched for
particular cell types. We advise assessing raw gene and count spatial
distributions prior to normalization for this reason.</p>
<p>In this particular example, we found that SCTransform better
ameliorated the mean-variance trend than log-normalization. We make no
claim that this will hold universally across samples. Indeed, rather
than advocating for one particular normalization approach, our goal was
to introduce you to means of diagnosing the impact of normalization
methods and of comparing them.</p>
<div id="keypoints1" class="callout keypoints">
<div class="callout-square">
<i class="callout-icon" data-feather="key"></i>
</div>
<div class="callout-inner">
<h3 class="callout-title">Key Points</h3>
<div class="callout-content">
<ul><li>Normalization is necessary to deal with technical artifacts, in
particular, varying total counts (library size) across spots and the
dependence of a gene’s variance on its mean expression.</li>
<li>Nevertheless, raw (unnormalized) data can provide biologically
meaningful insights such as region-specific differences in cell type or
density that impact total reads.</li>
<li>Popular methods aim to address these artifacts and include, but are
not limited to, CPM normalization, log normalization, and
SCTransform.</li>
<li>The ability of a normalization method to stabilize expression
variance across its mean (<em>i.e.</em>, to remove the dependence of the
former on the latter) can be assessed visually with a mean-variance
plot.</li>
</ul></div>
</div>
</div>
</div>
<div class="section level3">
<h3 id="where-we-are-headed">Where We Are Headed<a class="anchor" aria-label="anchor" href="#where-we-are-headed"></a></h3>
<p>Now that we have normalized away technical differences across spots,
we are poised for downstream analyses that will implicitly make
comparisons across them. These include dimensionality reduction and
clustering, which we will investigate next.</p>
<!--
Place links that you need to refer to multiple times across pages here. Delete
any links that you are not going to use.
 -->
</div>
</section></div> <!-- / div.lesson-content -->
    </main><!-- / main#main-content.main-content --><nav class="bottom-pagination mx-md-4" aria-label="Previous and Next Chapter"><div class="d-block d-sm-block d-md-none">
        <a class="chapter-link" href="remove-low-quality-spots.html"><i aria-hidden="true" class="small-arrow" data-feather="arrow-left"></i>Previous</a>
        <a class="chapter-link float-end" href="feature-selection-dimensionality-reduction-clustering.html">Next<i aria-hidden="true" class="small-arrow" data-feather="arrow-right"></i></a>
      </div>
      <!-- content for large screens -->
      <div class="d-none d-sm-none d-md-block">
        <a class="chapter-link" href="remove-low-quality-spots.html" rel="prev">
          <i aria-hidden="true" class="small-arrow" data-feather="arrow-left"></i>
          Previous: Remove Low-quality
        </a>
        <a class="chapter-link float-end" href="feature-selection-dimensionality-reduction-clustering.html" rel="next">
          Next: Feature Selection,...
          <i aria-hidden="true" class="small-arrow" data-feather="arrow-right"></i>
        </a>
      </div>
    </nav></div> <!-- / div.primary-content.col-xs-12 -->
<!-- END:   inst/pkgdown/templates/content-instructor.html-->

      </div><!--/div.row-->
      		<footer class="row footer mx-md-3"><hr><div class="col-md-6">
        <p>This lesson is subject to the <a href="CODE_OF_CONDUCT.html">Code of Conduct</a></p>
        <p>

        <a href="https://github.com/smcclatchy/spatial-transcriptomics/edit/main/episodes/apply-normalization-methods.Rmd" class="external-link">Edit on GitHub</a>

	
        | <a href="https://github.com/smcclatchy/spatial-transcriptomics/blob/main/CONTRIBUTING.md" class="external-link">Contributing</a>
        | <a href="https://github.com/smcclatchy/spatial-transcriptomics/" class="external-link">Source</a></p>
				<p><a href="https://github.com/smcclatchy/spatial-transcriptomics/blob/main/CITATION.cff" class="external-link">Cite</a> | <a href="mailto:susan.mcclatchy@jax.org">Contact</a> | <a href="https://carpentries.org/about/" class="external-link">About</a></p>
			</div>
			<div class="col-md-6">

        <p>Materials licensed under <a href="LICENSE.html">CC-BY 4.0</a> by the authors</p>

        <p>Template licensed under <a href="https://creativecommons.org/licenses/by-sa/4.0/" class="external-link">CC-BY 4.0</a> by <a href="https://carpentries.org/" class="external-link">The Carpentries</a></p>
        <p>Built with <a href="https://github.com/epiverse-trace/sandpaper/tree/patch-renv-github-bug" class="external-link">sandpaper (0.16.11)</a>, <a href="https://github.com/carpentries/pegboard/tree/5e0d711121a581777d83cf270e298c933f112d52" class="external-link">pegboard (0.7.9)</a>, and <a href="https://github.com/carpentries/varnish/tree/1.0.5" class="external-link">varnish (1.0.5)</a></p>
			</div>
		</footer></div> <!-- / div.container -->
	<div id="to-top">
		<a href="#top">
      <i class="search-icon" data-feather="arrow-up" role="img" aria-label="Back To Top"></i><br><!-- <span class="d-none d-sm-none d-md-none d-lg-none d-xl-block">Back</span> To Top --><span class="d-none d-sm-none d-md-none d-lg-none d-xl-block">Back</span> To Top
		</a>
	</div>
  <script type="application/ld+json">
    {
  "@context": "https://schema.org",
  "@type": "TrainingMaterial",
  "@id": "https://smcclatchy.github.io/spatial-transcriptomics/apply-normalization-methods.html",
  "inLanguage": "en",
  "dct:conformsTo": "https://bioschemas.org/profiles/TrainingMaterial/1.0-RELEASE",
  "description": "A Carpentries Lesson teaching foundational data and coding skills to researchers worldwide",
  "keywords": "software, data, lesson, The Carpentries, transcriptome, Visium, cancer",
  "name": "Normalization in Spatial Transcriptomics",
  "creativeWorkStatus": "active",
  "url": "https://smcclatchy.github.io/spatial-transcriptomics/apply-normalization-methods.html",
  "identifier": "https://smcclatchy.github.io/spatial-transcriptomics/apply-normalization-methods.html",
  "dateCreated": "2024-03-01",
  "dateModified": "2025-04-29",
  "datePublished": "2025-04-29"
}

  </script><script>
		feather.replace();
	</script></body></html><!-- END:   inst/pkgdown/templates/layout.html-->

