<!DOCTYPE html>
<!-- START: inst/pkgdown/templates/layout.html --><!-- Generated by pkgdown: do not edit by hand --><html lang="en" data-bs-theme="auto"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><title>Spatial Transcriptomics: Differential Expression Testing</title><meta name="viewport" content="width=device-width, initial-scale=1"><script src="assets/themetoggle.js"></script><link rel="stylesheet" type="text/css" href="assets/styles.css"><script src="assets/scripts.js" type="text/javascript"></script><!-- mathjax --><script type="text/x-mathjax-config">
    MathJax.Hub.Config({
      config: ["MMLorHTML.js"],
      jax: ["input/TeX","input/MathML","output/HTML-CSS","output/NativeMML", "output/PreviewHTML"],
      extensions: ["tex2jax.js","mml2jax.js","MathMenu.js","MathZoom.js", "fast-preview.js", "AssistiveMML.js", "a11y/accessibility-menu.js"],
      TeX: {
        extensions: ["AMSmath.js","AMSsymbols.js","noErrors.js","noUndefined.js"]
      },
      tex2jax: {
        inlineMath: [['\\(', '\\)']],
        displayMath: [ ['$$','$$'], ['\\[', '\\]'] ],
        processEscapes: true
      }
    });
    </script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><!-- Responsive Favicon for The Carpentries --><link rel="apple-touch-icon" sizes="180x180" href="favicons/incubator/apple-touch-icon.png"><link rel="icon" type="image/png" sizes="32x32" href="favicons/incubator/favicon-32x32.png"><link rel="icon" type="image/png" sizes="16x16" href="favicons/incubator/favicon-16x16.png"><link rel="manifest" href="favicons/incubator/site.webmanifest"><link rel="mask-icon" href="favicons/incubator/safari-pinned-tab.svg" color="#5bbad5"><meta name="msapplication-TileColor" content="#da532c"><meta name="theme-color" media="(prefers-color-scheme: light)" content="white"><meta name="theme-color" media="(prefers-color-scheme: dark)" content="black"></head><body>
    <header id="top" class="navbar navbar-expand-md top-nav incubator"><svg xmlns="http://www.w3.org/2000/svg" class="d-none"><symbol id="check2" viewbox="0 0 16 16"><path d="M13.854 3.646a.5.5 0 0 1 0 .708l-7 7a.5.5 0 0 1-.708 0l-3.5-3.5a.5.5 0 1 1 .708-.708L6.5 10.293l6.646-6.647a.5.5 0 0 1 .708 0z"></path></symbol><symbol id="circle-half" viewbox="0 0 16 16"><path d="M8 15A7 7 0 1 0 8 1v14zm0 1A8 8 0 1 1 8 0a8 8 0 0 1 0 16z"></path></symbol><symbol id="moon-stars-fill" viewbox="0 0 16 16"><path d="M6 .278a.768.768 0 0 1 .08.858 7.208 7.208 0 0 0-.878 3.46c0 4.021 3.278 7.277 7.318 7.277.527 0 1.04-.055 1.533-.16a.787.787 0 0 1 .81.316.733.733 0 0 1-.031.893A8.349 8.349 0 0 1 8.344 16C3.734 16 0 12.286 0 7.71 0 4.266 2.114 1.312 5.124.06A.752.752 0 0 1 6 .278z"></path><path d="M10.794 3.148a.217.217 0 0 1 .412 0l.387 1.162c.173.518.579.924 1.097 1.097l1.162.387a.217.217 0 0 1 0 .412l-1.162.387a1.734 1.734 0 0 0-1.097 1.097l-.387 1.162a.217.217 0 0 1-.412 0l-.387-1.162A1.734 1.734 0 0 0 9.31 6.593l-1.162-.387a.217.217 0 0 1 0-.412l1.162-.387a1.734 1.734 0 0 0 1.097-1.097l.387-1.162zM13.863.099a.145.145 0 0 1 .274 0l.258.774c.115.346.386.617.732.732l.774.258a.145.145 0 0 1 0 .274l-.774.258a1.156 1.156 0 0 0-.732.732l-.258.774a.145.145 0 0 1-.274 0l-.258-.774a1.156 1.156 0 0 0-.732-.732l-.774-.258a.145.145 0 0 1 0-.274l.774-.258c.346-.115.617-.386.732-.732L13.863.1z"></path></symbol><symbol id="sun-fill" viewbox="0 0 16 16"><path d="M8 12a4 4 0 1 0 0-8 4 4 0 0 0 0 8zM8 0a.5.5 0 0 1 .5.5v2a.5.5 0 0 1-1 0v-2A.5.5 0 0 1 8 0zm0 13a.5.5 0 0 1 .5.5v2a.5.5 0 0 1-1 0v-2A.5.5 0 0 1 8 13zm8-5a.5.5 0 0 1-.5.5h-2a.5.5 0 0 1 0-1h2a.5.5 0 0 1 .5.5zM3 8a.5.5 0 0 1-.5.5h-2a.5.5 0 0 1 0-1h2A.5.5 0 0 1 3 8zm10.657-5.657a.5.5 0 0 1 0 .707l-1.414 1.415a.5.5 0 1 1-.707-.708l1.414-1.414a.5.5 0 0 1 .707 0zm-9.193 9.193a.5.5 0 0 1 0 .707L3.05 13.657a.5.5 0 0 1-.707-.707l1.414-1.414a.5.5 0 0 1 .707 0zm9.193 2.121a.5.5 0 0 1-.707 0l-1.414-1.414a.5.5 0 0 1 .707-.707l1.414 1.414a.5.5 0 0 1 0 .707zM4.464 4.465a.5.5 0 0 1-.707 0L2.343 3.05a.5.5 0 1 1 .707-.707l1.414 1.414a.5.5 0 0 1 0 .708z"></path></symbol></svg><a class="visually-hidden-focusable skip-link" href="#main-content">Skip to main content</a>
  <div class="container-fluid top-nav-container">
    <div class="col-md-8">
      <div class="large-logo">
        <img id="incubator-logo" alt="Carpentries Incubator" src="assets/images/incubator-logo.svg"><span class="badge text-bg-info">
          <abbr title="This lesson is in the beta phase, which means that it is ready for teaching by instructors outside of the original author team.">
            <a href="https://cdh.carpentries.org/the-lesson-life-cycle.html#polishing-beta-stage" class="external-link alert-link">
              <i aria-hidden="true" class="icon" data-feather="alert-circle" style="border-radius: 5px"></i>
              Beta
            </a>
            <span class="visually-hidden">This lesson is in the beta phase, which means that it is ready for teaching by instructors outside of the original author team.</span>
          </abbr>
        </span>

      </div>
    </div>
    <div class="selector-container">
      <div id="theme-selector">
        <li class="nav-item dropdown" id="theme-button-list">
          <button class="btn btn-link nav-link px-0 px-lg-2 dropdown-toggle d-flex align-items-center" id="bd-theme" type="button" aria-expanded="false" data-bs-toggle="dropdown" data-bs-display="static" aria-label="Toggle theme (auto)">
            <svg class="bi my-1 theme-icon-active"><use href="#circle-half"></use></svg><i data-feather="chevron-down"></i>
          </button>
          <ul class="dropdown-menu dropdown-menu-end" aria-labelledby="bd-theme-text"><li>
              <button type="button" class="btn dropdown-item d-flex align-items-center" data-bs-theme-value="light" aria-pressed="false">
                <svg class="bi me-2 theme-icon"><use href="#sun-fill"></use></svg>
                Light
                <svg class="bi ms-auto d-none"><use href="#check2"></use></svg></button>
            </li>
            <li>
              <button type="button" class="btn dropdown-item d-flex align-items-center" data-bs-theme-value="dark" aria-pressed="false">
                <svg class="bi me-2 theme-icon"><use href="#moon-stars-fill"></use></svg>
                Dark
                <svg class="bi ms-auto d-none"><use href="#check2"></use></svg></button>
            </li>
            <li>
              <button type="button" class="btn dropdown-item d-flex align-items-center active" data-bs-theme-value="auto" aria-pressed="true">
                <svg class="bi me-2 theme-icon"><use href="#circle-half"></use></svg>
                Auto
                <svg class="bi ms-auto d-none"><use href="#check2"></use></svg></button>
            </li>
          </ul></li>
      </div>

      <div class="dropdown" id="instructor-dropdown">
        <button class="btn btn-secondary dropdown-toggle bordered-button" type="button" id="dropdownMenu1" data-bs-toggle="dropdown" aria-expanded="false">
          <i aria-hidden="true" class="icon" data-feather="eye"></i> Learner View <i data-feather="chevron-down"></i>
        </button>
        <ul class="dropdown-menu" aria-labelledby="dropdownMenu1"><li><button class="dropdown-item" type="button" onclick="window.location.href='instructor/differential-expression-testing.html';">Instructor View</button></li>
        </ul></div>
    </div>
  </div>
  <hr></header><nav class="navbar navbar-expand-xl bottom-nav incubator" aria-label="Main Navigation"><div class="container-fluid nav-container">
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbarSupportedContent" aria-controls="navbarSupportedContent" aria-expanded="false" aria-label="Toggle Navigation">
      <span class="navbar-toggler-icon"></span>
      <span class="menu-title">Menu</span>
    </button>
    <div class="nav-logo">
      <img class="small-logo" alt="Carpentries Incubator" src="assets/images/incubator-logo-sm.svg"></div>
    <div class="lesson-title-md">
      Spatial Transcriptomics
    </div>
    <div class="search-icon-sm">
      <!-- TODO: do not show until we have search
        <i role="img" aria-label="Search the All In One page" data-feather="search"></i>
      -->
    </div>
    <div class="desktop-nav">
      <ul class="navbar-nav me-auto mb-2 mb-lg-0"><li class="nav-item">
          <span class="lesson-title">
            Spatial Transcriptomics
          </span>
        </li>
        <li class="nav-item">
          <a class="nav-link" href="key-points.html">Key Points</a>
        </li>
        <li class="nav-item">
          <a class="nav-link" href="reference.html#glossary">Glossary</a>
        </li>
        <li class="nav-item">
          <a class="nav-link" href="profiles.html">Learner Profiles</a>
        </li>
        <li class="nav-item dropdown">
          <button class="nav-link dropdown-toggle" id="navbarDropdown" data-bs-toggle="dropdown" aria-expanded="false">
            More <i data-feather="chevron-down"></i>
          </button>
          <ul class="dropdown-menu" aria-labelledby="navbarDropdown"><li><a class="dropdown-item" href="reference.html">Reference</a></li>
          </ul></li>
      </ul></div>
    <!--
    <form class="d-flex col-md-2 search-form">
      <fieldset disabled>
      <input class="form-control me-2 searchbox" type="search" placeholder="" aria-label="">
        <button class="btn btn-outline-success tablet-search-button"  type="submit">
          <i class="search-icon" data-feather="search" role="img" aria-label="Search the All In One page"></i>
        </button>
      </fieldset>
    </form>
    -->
    <a id="search-button" class="btn btn-primary" href="aio.html" role="button" aria-label="Search the All In One page">Search the All In One page</a>
  </div><!--/div.container-fluid -->
</nav><div class="col-md-12 mobile-title">
  Spatial Transcriptomics
</div>

<aside class="col-md-12 lesson-progress"><div style="width: 76%" class="percentage">
    76%
  </div>
  <div class="progress incubator">
    <div class="progress-bar incubator" role="progressbar" style="width: 76%" aria-valuenow="76" aria-label="Lesson Progress" aria-valuemin="0" aria-valuemax="100">
    </div>
  </div>
</aside><div class="container">
      <div class="row">
        <!-- START: inst/pkgdown/templates/navbar.html -->
<div id="sidebar-col" class="col-lg-4">
  <div id="sidebar" class="sidebar">
      <nav aria-labelledby="flush-headingEleven"><button role="button" aria-label="close menu" alt="close menu" aria-expanded="true" aria-controls="sidebar" class="collapse-toggle" data-collapse="Collapse " data-episodes="Episodes ">
          <i class="search-icon" data-feather="x" role="img"></i>
        </button>
        <div class="sidebar-inner">
          <div class="row mobile-row" id="theme-row-mobile">
            <div class="col" id="theme-selector">
              <li class="nav-item dropdown" id="theme-button-list">
                <button class="btn btn-link nav-link px-0 px-lg-2 dropdown-toggle d-flex align-items-center" id="bd-theme" type="button" aria-expanded="false" data-bs-toggle="dropdown" data-bs-display="static" aria-label="Toggle theme (auto)">
                  <svg class="bi my-1 theme-icon-active"><use href="#circle-half"></use></svg><span class="d-lg-none ms-1" id="bd-theme-text">Toggle Theme</span>
                </button>
                <ul class="dropdown-menu dropdown-menu-right" aria-labelledby="bd-theme-text"><li>
                    <button type="button" class="btn dropdown-item d-flex align-items-center" data-bs-theme-value="light" aria-pressed="false">
                      <svg class="bi me-2 theme-icon"><use href="#sun-fill"></use></svg>
                      Light
                      <svg class="bi ms-auto d-none"><use href="#check2"></use></svg></button>
                  </li>
                  <li>
                    <button type="button" class="btn dropdown-item d-flex align-items-center" data-bs-theme-value="dark" aria-pressed="false">
                      <svg class="bi me-2 theme-icon"><use href="#moon-stars-fill"></use></svg>
                      Dark
                      <svg class="bi ms-auto d-none"><use href="#check2"></use></svg></button>
                  </li>
                  <li>
                    <button type="button" class="btn dropdown-item d-flex align-items-center active" data-bs-theme-value="auto" aria-pressed="true">
                      <svg class="bi me-2 theme-icon"><use href="#circle-half"></use></svg>
                      Auto
                      <svg class="bi ms-auto d-none"><use href="#check2"></use></svg></button>
                  </li>
                </ul></li>
            </div>
          </div>
          <div class="row mobile-row">
            <div class="col">
              <div class="sidenav-view-selector">
                <div class="accordion accordion-flush" id="accordionFlush9">
                  <div class="accordion-item">
                    <h2 class="accordion-header" id="flush-headingNine">
                      <button class="accordion-button collapsed" id="instructor" type="button" data-bs-toggle="collapse" data-bs-target="#flush-collapseNine" aria-expanded="false" aria-controls="flush-collapseNine">
                        <i id="eye" aria-hidden="true" class="icon" data-feather="eye"></i> Learner View
                      </button>
                    </h2>
                    <div id="flush-collapseNine" class="accordion-collapse collapse" aria-labelledby="flush-headingNine" data-bs-parent="#accordionFlush2">
                      <div class="accordion-body">
                        <a href="instructor/differential-expression-testing.html">Instructor View</a>
                      </div>
                    </div>
                  </div><!--/div.accordion-item-->
                </div><!--/div.accordion-flush-->
              </div><!--div.sidenav-view-selector -->
            </div><!--/div.col -->

            <hr></div><!--/div.mobile-row -->

          <div class="accordion accordion-flush" id="accordionFlush11">
            <div class="accordion-item">

              <button id="chapters" class="accordion-button show" type="button" data-bs-toggle="collapse" data-bs-target="#flush-collapseEleven" aria-expanded="false" aria-controls="flush-collapseEleven">
                <h2 class="accordion-header chapters" id="flush-headingEleven">
                  EPISODES
                </h2>
              </button>
              <div id="flush-collapseEleven" class="accordion-collapse show collapse" aria-labelledby="flush-headingEleven" data-bs-parent="#accordionFlush11">

                <div class="accordion-body">
                  <div class="accordion accordion-flush" id="accordionFlush1">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading1">
        <a href="index.html">Summary and Setup</a>
    </div><!--/div.accordion-header-->

  </div><!--/div.accordion-item-->
</div><!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush2">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading2">
        <a href="introduction.html">1. Spatially Resolved Transcriptomics in Life Sciences Research</a>
    </div><!--/div.accordion-header-->

  </div><!--/div.accordion-item-->
</div><!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush3">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading3">
        <a href="data-and-study-design.html">2. Data and Study Design</a>
    </div><!--/div.accordion-header-->

  </div><!--/div.accordion-item-->
</div><!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush4">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading4">
        <a href="data-preprocessing.html">3. Data Preprocessing</a>
    </div><!--/div.accordion-header-->

  </div><!--/div.accordion-item-->
</div><!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush5">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading5">
        <a href="remove-low-quality-spots.html">4. Remove Low-quality Spots</a>
    </div><!--/div.accordion-header-->

  </div><!--/div.accordion-item-->
</div><!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush6">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading6">
        <a href="apply-normalization-methods.html">5. Normalization in Spatial Transcriptomics</a>
    </div><!--/div.accordion-header-->

  </div><!--/div.accordion-item-->
</div><!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush7">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading7">
        <a href="feature-selection-dimensionality-reduction-clustering.html">6. Feature Selection, Dimensionality Reduction, and Spot Clustering</a>
    </div><!--/div.accordion-header-->

  </div><!--/div.accordion-item-->
</div><!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush8">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading8">
        <a href="deconvolve-cell-types-in-a-spot.html">7. Deconvolution in Spatial Transcriptomics</a>
    </div><!--/div.accordion-header-->

  </div><!--/div.accordion-item-->
</div><!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlushcurrent">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-headingcurrent">
      <button class="accordion-button" type="button" data-bs-toggle="collapse" data-bs-target="#flush-collapsecurrent" aria-expanded="true" aria-controls="flush-collapsecurrent">
        <span class="visually-hidden">Current Chapter</span>
        <span class="current-chapter">
        8. Differential Expression Testing
        </span>
      </button>
    </div><!--/div.accordion-header-->

    <div id="flush-collapsecurrent" class="accordion-collapse collapse show" aria-labelledby="flush-headingcurrent" data-bs-parent="#accordionFlushcurrent">
      <div class="accordion-body">
        <ul><li><a href="#spot-level-differential-expression-across-annotated-regions">Spot-level Differential Expression Across Annotated Regions</a></li>
<li><a href="#spot-level-differential-expression-across-clusters">Spot-level Differential Expression Across Clusters</a></li>
<li><a href="#morans-i-statistic">Moran’s I Statistic</a></li>
<li><a href="#correlation-of-region-specific-differentially-expressed-genes-and-spatially-variable-genes">Correlation of Region-specific Differentially Expressed Genes and
Spatially Variable Genes</a></li>
<li><a href="#differential-expression-analysis-for-multiple-samples">Differential Expression Analysis for Multiple Samples</a></li>
<li><a href="#other-considerations">Other Considerations</a></li>
        </ul></div><!--/div.accordion-body-->
    </div><!--/div.accordion-collapse-->

  </div><!--/div.accordion-item-->
</div><!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush10">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading10">
        <a href="final-exercise.html">9. Putting it all Together</a>
    </div><!--/div.accordion-header-->

  </div><!--/div.accordion-item-->
</div><!--/div.accordion-flush-->

                </div>
              </div>
            </div>

            <hr class="half-width"><div class="accordion accordion-flush lesson-resources" id="accordionFlush12">
              <div class="accordion-item">
                <h2 class="accordion-header" id="flush-headingTwelve">
                  <button class="accordion-button collapsed" id="lesson-resources" type="button" data-bs-toggle="collapse" data-bs-target="#flush-collapseTwelve" aria-expanded="false" aria-controls="flush-collapseTwelve">
                    RESOURCES
                  </button>
                </h2>
                <div id="flush-collapseTwelve" class="accordion-collapse collapse" aria-labelledby="flush-headingTwelve" data-bs-parent="#accordionFlush12">
                  <div class="accordion-body">
                    <ul><li>
                        <a href="key-points.html">Key Points</a>
                      </li>
                      <li>
                        <a href="reference.html#glossary">Glossary</a>
                      </li>
                      <li>
                        <a href="profiles.html">Learner Profiles</a>
                      </li>
                      <li><a href="reference.html">Reference</a></li>
                    </ul></div>
                </div>
              </div>
            </div>
            <hr class="half-width lesson-resources"><a href="aio.html">See all in one page</a>


            <hr class="d-none d-sm-block d-md-none"><div class="d-grid gap-1">

            </div>
          </div><!-- /div.accordion -->
        </div><!-- /div.sidebar-inner -->
      </nav></div><!-- /div.sidebar -->
  </div><!-- /div.sidebar-col -->
<!-- END:   inst/pkgdown/templates/navbar.html-->

        <!-- START: inst/pkgdown/templates/content-instructor.html -->
  <div class="col-xl-8 col-lg-12 primary-content">
    <nav class="lesson-content mx-md-4" aria-label="Previous and Next Chapter"><!-- content for small screens --><div class="d-block d-sm-block d-md-none">
        <a class="chapter-link" href="deconvolve-cell-types-in-a-spot.html"><i aria-hidden="true" class="small-arrow" data-feather="arrow-left"></i>Previous</a>
        <a class="chapter-link float-end" href="final-exercise.html">Next<i aria-hidden="true" class="small-arrow" data-feather="arrow-right"></i></a>
      </div>
      <!-- content for large screens -->
      <div class="d-none d-sm-none d-md-block">
        <a class="chapter-link" href="deconvolve-cell-types-in-a-spot.html" rel="prev">
          <i aria-hidden="true" class="small-arrow" data-feather="arrow-left"></i>
          Previous: Deconvolution in
        </a>
        <a class="chapter-link float-end" href="final-exercise.html" rel="next">
          Next: Putting it all...
          <i aria-hidden="true" class="small-arrow" data-feather="arrow-right"></i>
        </a>
      </div>
      <hr></nav><main id="main-content" class="main-content"><div class="container lesson-content">
        <h1>Differential Expression Testing</h1>
        <p>Last updated on 2025-03-18 |

        <a href="https://github.com/smcclatchy/spatial-transcriptomics/edit/main/episodes/differential-expression-testing.Rmd" class="external-link">Edit this page <i aria-hidden="true" data-feather="edit"></i></a></p>



        <div class="text-end">
          <button role="button" aria-pressed="false" tabindex="0" id="expand-code" class="pull-right" data-expand="Expand All Solutions " data-collapse="Collapse All Solutions "> Expand All Solutions <i aria-hidden="true" data-feather="plus"></i></button>
        </div>

        

<div class="overview card">
<h2 class="card-header">Overview</h2>
<div class="row g-0">
<div class="col-md-4">
<div class="card-body">
<div class="inner">
<h3 class="card-title">Questions</h3>
<ul><li>How can we assess region-specific gene expression using differential
expression?</li>
<li>How can we assess spatially varying gene expression using spatial
statistics?</li>
</ul></div>
</div>
</div>
<div class="col-md-8">
<div class="card-body">
<div class="inner bordered">
<h3 class="card-title">Objectives</h3>
<ul><li>Identify differentially expressed genes across different layers
defined by expert annotations.</li>
<li>Utilize the Moran’s I statistic to find spatially variable
genes.</li>
<li>Explore the relationship between layer-specific genes identified by
differential expression and spatially varying genes identified by
Moran’s I.</li>
</ul></div>
</div>
</div>
</div>
</div>
<section><h2 class="section-heading" id="spot-level-differential-expression-across-annotated-regions">Spot-level Differential Expression Across Annotated Regions<a class="anchor" aria-label="anchor" href="#spot-level-differential-expression-across-annotated-regions"></a></h2>
<hr class="half-width"><p>We will begin by performing differential expression (DE) across the
annotated layers. We will use the source publication’s spot annotation,
which is stored in the <code>layer_guess</code> column of the metadata.
As a reminder, those look like:</p>
<div class="codewrapper sourceCode" id="cb1">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">SpatialDimPlotColorSafe</span><span class="op">(</span><span class="va">sct_st</span><span class="op">[</span>, <span class="op">!</span><span class="fu">is.na</span><span class="op">(</span><span class="va">sct_st</span><span class="op">[[</span><span class="op">]</span><span class="op">]</span><span class="op">$</span><span class="va">layer_guess</span><span class="op">)</span><span class="op">]</span>, </span>
<span>                        <span class="st">"layer_guess"</span><span class="op">)</span> <span class="op">+</span> </span>
<span>  <span class="fu">labs</span><span class="op">(</span>fill <span class="op">=</span> <span class="st">"Layer"</span><span class="op">)</span> </span></code></pre>
</div>
<figure><img src="fig/differential-expression-testing-rendered-layers-1.png" style="display: block; margin: auto;" class="figure mx-auto d-block"></figure><p>We identify genes that are upregulated in each annotated brain region
using the <a href="https://satijalab.org/seurat/reference/findallmarkers" class="external-link"><code>FindAllMarkers</code></a>
function in Seurat. This performs a “one versus rest” comparison of a
gene’s expression in one region relative to that gene’s expression in
all other regions. The default test used here, the Wilcoxon Rank Sum
test, will use an efficient implementation within the <a href="https://github.com/immunogenomics/presto" class="external-link"><code>presto</code></a>
library, if installed. The speedup over the default implementation is
substantial, and we highly recommend installing <code>presto</code> and
using Seurat v5, which leverages it.</p>
<div class="codewrapper sourceCode" id="cb2">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">Idents</span><span class="op">(</span><span class="va">sct_st</span><span class="op">)</span>  <span class="op">&lt;-</span> <span class="st">"layer_guess"</span></span>
<span><span class="va">de_genes</span>           <span class="op">&lt;-</span> <span class="fu">FindAllMarkers</span><span class="op">(</span><span class="va">sct_st</span>, </span>
<span>                                     assay    <span class="op">=</span> <span class="st">"SCT"</span>,</span>
<span>                                     verbose  <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>                                     only.pos <span class="op">=</span> <span class="cn">TRUE</span>, </span>
<span>                                     min.pct  <span class="op">=</span> <span class="fl">0.25</span>, </span>
<span>                                     logfc.threshold <span class="op">=</span> <span class="fl">0.25</span><span class="op">)</span></span></code></pre>
</div>
<p>The resulting table indicates DE p-values and adjusted p-values for
each gene, along with the percentage of spots in which the gene was
detected (pct.1) in the corresponding cluster and the percentage of
spots in which it was detected in all <em>other</em> clusters
(pct.2):</p>
<div class="codewrapper sourceCode" id="cb3">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">head</span><span class="op">(</span><span class="va">de_genes</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>                p_val avg_log2FC pct.1 pct.2     p_val_adj cluster    gene
MT-CO1  7.490920e-163  0.3732349 1.000  1.00 1.348066e-158  Layer3  MT-CO1
ENC1    1.306480e-142  0.8693104 0.998  0.93 2.351142e-138  Layer3    ENC1
MT-ATP6 7.395810e-127  0.3333148 1.000  1.00 1.330950e-122  Layer3 MT-ATP6
MT-CYB  1.306736e-117  0.3288096 1.000  1.00 2.351602e-113  Layer3  MT-CYB
MT-CO2  3.611734e-116  0.2869348 1.000  1.00 6.499677e-112  Layer3  MT-CO2
MT-CO3  6.536277e-105  0.2765344 1.000  1.00 1.176268e-100  Layer3  MT-CO3</code></pre>
</div>
<p>It is not uncommon to have pathologist annotations of regions. The
Visium assay is performed on a tissue that is also stained for
hematoxylin and eosin (H&amp;E) – a routine practice in pathology for
diagnosing cancer, for example. A pathologist could manually annotate
this H&amp;E image using a histology viewer, such as QuPath.</p>
</section><section><h2 class="section-heading" id="spot-level-differential-expression-across-clusters">Spot-level Differential Expression Across Clusters<a class="anchor" aria-label="anchor" href="#spot-level-differential-expression-across-clusters"></a></h2>
<hr class="half-width"><p>In cases where we do not have expert annotations, we could perform
the analysis above across clusters. Indeed, this is often the first step
to interpreting a cluster – comparing its marker genes to those of
regions expected within the tissue. It is important to remember,
however, that these markers are for clusters of spots – aggregations of
cells – not individual cells, as would be the case in scRNA-seq. Those
aggregations may be of cells with similar types, in which case analysis
may be similar to that of scRNA-seq, or of different cell types, in
which case the interpretation would be quite different than with
scRNA-seq. Let’s try this.</p>
<p>First, let’s remind ourselves what the clusters look like:</p>
<div class="codewrapper sourceCode" id="cb5">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">SpatialDimPlotColorSafe</span><span class="op">(</span><span class="va">sct_st</span>, <span class="st">"seurat_clusters"</span><span class="op">)</span> <span class="op">+</span> </span>
<span>  <span class="fu">labs</span><span class="op">(</span>fill <span class="op">=</span> <span class="st">"Cluster"</span><span class="op">)</span> </span></code></pre>
</div>
<figure><img src="fig/differential-expression-testing-rendered-clusters-1.png" style="display: block; margin: auto;" class="figure mx-auto d-block"></figure><p>Now, we will use nearly identical code as above to perform the
differential expression analysis. We only change each cell’s identity
class with the <a href="https://satijalab.github.io/seurat-object/reference/Idents.html" class="external-link"><code>Idents</code></a>
function, to associate each cell with its cluster.</p>
<div class="codewrapper sourceCode" id="cb6">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">Idents</span><span class="op">(</span><span class="va">sct_st</span><span class="op">)</span>  <span class="op">&lt;-</span> <span class="st">"seurat_clusters"</span></span>
<span><span class="va">de_genes_cluster</span>   <span class="op">&lt;-</span> <span class="fu">FindAllMarkers</span><span class="op">(</span><span class="va">sct_st</span>, </span>
<span>                                     assay    <span class="op">=</span> <span class="st">"SCT"</span>,</span>
<span>                                     verbose  <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>                                     only.pos <span class="op">=</span> <span class="cn">TRUE</span>, </span>
<span>                                     min.pct  <span class="op">=</span> <span class="fl">0.25</span>, </span>
<span>                                     logfc.threshold <span class="op">=</span> <span class="fl">0.25</span><span class="op">)</span></span></code></pre>
</div>
<p>Let’s visualize the overlap between genes differentially expressed
across annotated layers and those differentially expressed across
clusters. We will do so using a confusion matrix.</p>
<div class="codewrapper sourceCode" id="cb7">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">de_genes_all</span> <span class="op">&lt;-</span> <span class="fu">merge</span><span class="op">(</span><span class="fu">subset</span><span class="op">(</span><span class="va">de_genes</span>, <span class="va">p_val_adj</span> <span class="op">&lt;</span> <span class="fl">0.05</span><span class="op">)</span>,</span>
<span>                      <span class="fu">subset</span><span class="op">(</span><span class="va">de_genes_cluster</span>, <span class="va">p_val_adj</span> <span class="op">&lt;</span> <span class="fl">0.05</span><span class="op">)</span>,</span>
<span>                      by<span class="op">=</span><span class="fu">c</span><span class="op">(</span><span class="st">"gene"</span><span class="op">)</span>, all<span class="op">=</span><span class="cn">TRUE</span>, </span>
<span>                      suffixes <span class="op">=</span> <span class="fu">c</span><span class="op">(</span><span class="st">".anno"</span>, <span class="st">".cluster"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">layer.order</span> <span class="op">&lt;-</span> <span class="fu">c</span><span class="op">(</span><span class="st">"WM"</span>, <span class="st">"Layer1"</span>, <span class="st">"Layer2"</span>, <span class="st">"Layer3"</span>, <span class="st">"Layer4"</span>, </span>
<span>                 <span class="st">"Layer5"</span>, <span class="st">"Layer6"</span><span class="op">)</span></span>
<span><span class="va">df</span>            <span class="op">&lt;-</span> <span class="fu">as.data.frame</span><span class="op">(</span><span class="fu">table</span><span class="op">(</span><span class="va">de_genes_all</span><span class="op">$</span><span class="va">cluster.anno</span>, </span>
<span>                                     <span class="va">de_genes_all</span><span class="op">$</span><span class="va">cluster.cluster</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu">colnames</span><span class="op">(</span><span class="va">df</span><span class="op">)</span>  <span class="op">&lt;-</span> <span class="fu">c</span><span class="op">(</span><span class="st">"de.anno"</span>, <span class="st">"de.cluster"</span>, <span class="st">"Freq"</span><span class="op">)</span></span>
<span><span class="va">df</span><span class="op">$</span><span class="va">de.anno</span> <span class="op">&lt;-</span> <span class="fu">factor</span><span class="op">(</span><span class="va">df</span><span class="op">$</span><span class="va">de.anno</span>, levels <span class="op">=</span> <span class="fu">c</span><span class="op">(</span><span class="st">"NA"</span>, <span class="va">layer.order</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">df</span><span class="op">$</span><span class="va">de.cluster</span> <span class="op">&lt;-</span> <span class="fu">factor</span><span class="op">(</span><span class="va">df</span><span class="op">$</span><span class="va">de.cluster</span><span class="op">)</span></span>
<span></span>
<span><span class="fu">ggplot</span><span class="op">(</span>data <span class="op">=</span> <span class="va">df</span>, <span class="fu">aes</span><span class="op">(</span>x <span class="op">=</span> <span class="va">de.anno</span>, y <span class="op">=</span> <span class="va">de.cluster</span>, fill <span class="op">=</span> <span class="va">Freq</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu">geom_tile</span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu">theme</span><span class="op">(</span>text <span class="op">=</span> <span class="fu">element_text</span><span class="op">(</span>size <span class="op">=</span> <span class="fl">20</span><span class="op">)</span>, </span>
<span>        axis.text.x <span class="op">=</span> <span class="fu">element_text</span><span class="op">(</span>angle <span class="op">=</span> <span class="fl">45</span>, vjust <span class="op">=</span> <span class="fl">1</span>, hjust<span class="op">=</span><span class="fl">1</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu">xlab</span><span class="op">(</span><span class="st">"Annotation-based DE Genes"</span><span class="op">)</span> <span class="op">+</span> <span class="fu">ylab</span><span class="op">(</span><span class="st">"Cluster-based DE Genes"</span><span class="op">)</span></span></code></pre>
</div>
<figure><img src="fig/differential-expression-testing-rendered-de-comparison-confusion-1.png" style="display: block; margin: auto;" class="figure mx-auto d-block"></figure><p>The above plot shows only marginal overlap between the
annotation-derived DE genes on the x axis and the cluster-derived DE
genes on the y axis. This might indicate that the two approaches are
complementary. In fact, they could be used in conjunction. We could have
performed clustering <em>within</em> the annotated regions to assess
intra-layer heterogeneity.</p>
</section><section><h2 class="section-heading" id="morans-i-statistic">Moran’s I Statistic<a class="anchor" aria-label="anchor" href="#morans-i-statistic"></a></h2>
<hr class="half-width"><p>We next consider an alternative to the above approaches that both
rely on defined regions – be they defined from expert annotation or via
clustering. Instead, we will apply the Moran’s I statistic. Moran’s I is
a measure used to assess spatial autocorrelation in data, indicating
whether similar values of a feature (<em>e.g.</em>, expression levels of
a gene) are clustered, dispersed, or random (<a href="https://ij-healthgeographics.biomedcentral.com/articles/10.1186/1476-072X-9-33" class="external-link">Jackson,
et al. 2010</a>). These correspond to Moran’s I values that are
positive, negative, or near zero, respectively.</p>
<figure><img src="https://upload.wikimedia.org/wikipedia/commons/f/f0/Moran%27s_I_example.png" alt="Moran's I statistic quantifies spatial correlation." class="figure mx-auto d-block"><div class="figcaption">Moran’s I statistic quantifies spatial
correlation. <strong>Top Left:</strong> Checkerboard pattern results in
negative Moran’s I, indicating anti-correlation. <strong>Top
Right:</strong> Linear gradient shows a high positive Moran’s I,
reflecting a strong spatial gradient. <strong>Bottom Left:</strong>
Random pattern leads to a Moran’s I near zero, suggesting no significant
spatial autocorrelation. <strong>Bottom Right:</strong> ‘Ink blot’
pattern demonstrates positive autocorrelation, indicative of a clustered
or spreading pattern. Relationships are calculated using direct, equally
weighted neighbors, normalized for each cell.</div>
</figure><p>Image by
<a href="https://commons.wikimedia.org/wiki/File:Moran%27s_I_example.png" class="external-link">WikiNukalito</a>,
<a href="https://creativecommons.org/licenses/by-sa/4.0" class="external-link">CC BY-SA
4.0</a>, via Wikimedia Commons.</p>
<p>Here, we can apply it to detect genes whose expression patterns
exhibit spatial structure, which may reflect region-specific, biological
function. That is, we anticipate that spatially variable genes will
exhibit region-specific expression. Let’s check that hypothesis by first
computing spatially variable genes and then assessing whether they are
differentially expressed across regions.</p>
<div class="section level3">
<h3 id="spatial-differential-expression-using-morans-i">Spatial Differential Expression Using Moran’s I<a class="anchor" aria-label="anchor" href="#spatial-differential-expression-using-morans-i"></a></h3>
<p>We identify the genes whose expression patterns exhibit clear spatial
structure using Moran’s I algorithm, as implemented in <a href="https://satijalab.org/seurat/reference/findspatiallyvariablefeatures" class="external-link"><code>FindSpatiallyVariableFeatures</code></a>.
We have selected the top 1,000 genes, which should be sufficient to
identify brain regions. The following will take several minutes to
run.</p>
<div class="codewrapper sourceCode" id="cb8">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">svg</span> <span class="op">&lt;-</span> </span>
<span>  <span class="fu">FindSpatiallyVariableFeatures</span><span class="op">(</span><span class="va">sct_st</span>, </span>
<span>                                assay            <span class="op">=</span> <span class="st">"SCT"</span>, </span>
<span>                                features         <span class="op">=</span> <span class="fu">VariableFeatures</span><span class="op">(</span><span class="va">sct_st</span><span class="op">)</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">1000</span><span class="op">]</span>, </span>
<span>                                selection.method <span class="op">=</span> <span class="st">"moransi"</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">ERROR<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="error" tabindex="0"><code>Error: 'RunMoransI' requires either Rfast2 or ape to be installed</code></pre>
</div>
<p><code>FindSpatiallyVariableFeatures</code> returns a Seurat object,
populated with Moran’s I-derived results. Normally, we would use the <a href="https://satijalab.github.io/seurat-object/reference/VariableFeatures.html" class="external-link"><code>SpatiallyVariableFeatures</code></a>
function to query those results. But, there is a bug in that function as
described <a href="https://github.com/satijalab/seurat/issues/7422" class="external-link">in
Seurat’s issue pages</a>. So, instead we will manually extract the top
100 ranked spatially variable genes, along with their Moran’s I values
and associated p-values:</p>
<div class="codewrapper sourceCode" id="cb10">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Get and sort 'MoransI_observed' values</span></span>
<span><span class="va">morans_i_genes</span> <span class="op">&lt;-</span> <span class="va">svg</span><span class="op">@</span><span class="va">assays</span><span class="op">[[</span><span class="st">"SCT"</span><span class="op">]</span><span class="op">]</span><span class="op">@</span><span class="va">meta.features</span> <span class="op">%&gt;%</span></span>
<span>                    <span class="fu">rownames_to_column</span><span class="op">(</span><span class="st">"gene"</span><span class="op">)</span> <span class="op">%&gt;%</span></span>
<span>                    <span class="fu">arrange</span><span class="op">(</span><span class="fu">desc</span><span class="op">(</span><span class="va">MoransI_observed</span><span class="op">)</span><span class="op">)</span> <span class="op">%&gt;%</span></span>
<span>                    <span class="fu">slice_head</span><span class="op">(</span>n <span class="op">=</span> <span class="fl">100</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">ERROR<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="error" tabindex="0"><code>Error in svg@assays: no applicable method for `@` applied to an object of class "function"</code></pre>
</div>
<div class="codewrapper sourceCode" id="cb12">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">head</span><span class="op">(</span><span class="va">morans_i_genes</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">ERROR<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="error" tabindex="0"><code>Error in h(simpleError(msg, call)): error in evaluating the argument 'x' in selecting a method for function 'head': object 'morans_i_genes' not found</code></pre>
</div>
</div>
</section><section><h2 class="section-heading" id="correlation-of-region-specific-differentially-expressed-genes-and-spatially-variable-genes">Correlation of Region-specific Differentially Expressed Genes and
Spatially Variable Genes<a class="anchor" aria-label="anchor" href="#correlation-of-region-specific-differentially-expressed-genes-and-spatially-variable-genes"></a></h2>
<hr class="half-width"><div class="section level3">
<h3 id="heatmap-of-differential-expression">Heatmap of Differential Expression<a class="anchor" aria-label="anchor" href="#heatmap-of-differential-expression"></a></h3>
<p>As a sanity check of both the region-specific differential expression
and the Moran’s I approaches, let’s check our hypothesis that spatially
variable genes are likely to show regional differential expression. To
do this, we will plot a heatmap of the <em>differential expression</em>
p-values for the top 100 spatially variable genes, organized by brain
region.</p>
<div class="codewrapper sourceCode" id="cb14">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Merge the Moran's I values with the DE genes</span></span>
<span><span class="va">df</span> <span class="op">&lt;-</span> <span class="fu">merge</span><span class="op">(</span><span class="va">morans_i_genes</span>, <span class="va">de_genes</span>, all.x <span class="op">=</span> <span class="cn">TRUE</span>, by <span class="op">=</span> <span class="st">"gene"</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">ERROR<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="error" tabindex="0"><code>Error in h(simpleError(msg, call)): error in evaluating the argument 'x' in selecting a method for function 'merge': object 'morans_i_genes' not found</code></pre>
</div>
<div class="codewrapper sourceCode" id="cb16">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">df</span> <span class="op">&lt;-</span> <span class="fu">subset</span><span class="op">(</span><span class="va">df</span>, <span class="op">!</span><span class="fu">is.na</span><span class="op">(</span><span class="va">cluster</span><span class="op">)</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">ERROR<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="error" tabindex="0"><code>Error in eval(e, x, parent.frame()): object 'cluster' not found</code></pre>
</div>
<div class="codewrapper sourceCode" id="cb18">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># We will plot the -log2 pvalues. Compute this and adjust for taking </span></span>
<span><span class="co"># log of 0.</span></span>
<span><span class="va">df</span><span class="op">$</span><span class="va">log_p_val_adj</span> <span class="op">&lt;-</span> <span class="op">-</span><span class="fu">log2</span><span class="op">(</span><span class="va">df</span><span class="op">$</span><span class="va">p_val_adj</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">ERROR<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="error" tabindex="0"><code>Error in log2(df$p_val_adj): non-numeric argument to mathematical function</code></pre>
</div>
<div class="codewrapper sourceCode" id="cb20">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">df</span><span class="op">$</span><span class="va">log_p_val_adj</span><span class="op">[</span><span class="fu">is.infinite</span><span class="op">(</span><span class="va">df</span><span class="op">$</span><span class="va">log_p_val_adj</span><span class="op">)</span><span class="op">]</span> <span class="op">&lt;-</span> </span>
<span>  <span class="fu">max</span><span class="op">(</span><span class="va">df</span><span class="op">$</span><span class="va">log_p_val_adj</span><span class="op">[</span><span class="op">!</span><span class="fu">is.infinite</span><span class="op">(</span><span class="va">df</span><span class="op">$</span><span class="va">log_p_val_adj</span><span class="op">)</span><span class="op">]</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">WARNING<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="warning" tabindex="0"><code>Warning in max(df$log_p_val_adj[!is.infinite(df$log_p_val_adj)]): no
non-missing arguments to max; returning -Inf</code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">ERROR<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="error" tabindex="0"><code>Error in `$&lt;-.data.frame`(`*tmp*`, log_p_val_adj, value = numeric(0)): replacement has 0 rows, data has 56</code></pre>
</div>
<div class="codewrapper sourceCode" id="cb23">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Create a matrix whose rows are the spatially variable genes </span></span>
<span><span class="co"># (indicated by Moran's I), whose columns are the clusters, and whose </span></span>
<span><span class="co"># entries are the adjusted DE pvalue for the corresponding gene and cluster.</span></span>
<span><span class="va">p_val_adj_matrix</span> <span class="op">&lt;-</span> <span class="va">df</span> <span class="op">%&gt;%</span></span>
<span>                       <span class="fu">select</span><span class="op">(</span><span class="va">gene</span>, <span class="va">cluster</span>,<span class="va">log_p_val_adj</span><span class="op">)</span> <span class="op">%&gt;%</span></span>
<span>                       <span class="fu">pivot_wider</span><span class="op">(</span>names_from <span class="op">=</span> <span class="va">cluster</span>, </span>
<span>                                   values_from <span class="op">=</span> <span class="va">log_p_val_adj</span>, </span>
<span>                                   values_fill <span class="op">=</span> <span class="fl">0</span><span class="op">)</span> <span class="op">%&gt;%</span></span>
<span>                       <span class="fu">column_to_rownames</span><span class="op">(</span><span class="st">"gene"</span><span class="op">)</span> <span class="op">%&gt;%</span></span>
<span>                       <span class="fu">as.matrix</span><span class="op">(</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">ERROR<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="error" tabindex="0"><code>Error in (function (cond) : error in evaluating the argument 'x' in selecting a method for function 'as.matrix': Can't select columns that don't exist.
✖ Column `gene` doesn't exist.</code></pre>
</div>
<div class="codewrapper sourceCode" id="cb25">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Order the regions according to their spatial organization from </span></span>
<span><span class="co"># inner to outer layers</span></span>
<span><span class="va">p_val_adj_matrix</span> <span class="op">&lt;-</span> <span class="va">p_val_adj_matrix</span><span class="op">[</span>, <span class="va">layer.order</span><span class="op">]</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">ERROR<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="error" tabindex="0"><code>Error: object 'p_val_adj_matrix' not found</code></pre>
</div>
<div class="codewrapper sourceCode" id="cb27">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Create a heatmap of the DE p-values of spatially variable genes</span></span>
<span><span class="fu">Heatmap</span><span class="op">(</span><span class="va">p_val_adj_matrix</span>,</span>
<span>        column_title      <span class="op">=</span> <span class="st">"Heatmap of DE p-values of spatially variable genes"</span>,</span>
<span>        name              <span class="op">=</span> <span class="st">"DE -log2(p-values)"</span>, <span class="co"># Title for the heatmap legend</span></span>
<span>        row_title         <span class="op">=</span> <span class="st">"Spatially variable genes"</span>,</span>
<span>        cluster_rows      <span class="op">=</span> <span class="cn">TRUE</span>, </span>
<span>        cluster_columns   <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>        show_row_names    <span class="op">=</span> <span class="cn">FALSE</span>, </span>
<span>        show_column_names <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>        show_row_dend     <span class="op">=</span> <span class="cn">FALSE</span>, </span>
<span>        show_column_dend  <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">ERROR<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="error" tabindex="0"><code>Error: object 'p_val_adj_matrix' not found</code></pre>
</div>
<p>The heatmap visualization reveals a key finding of our analysis:
genes displaying the highest Moran’s I values show distinct expression
patterns that align with specific brain regions identified through
expert annotations. This observation underscores the spatial correlation
of gene expression, highlighting its potential relevance in
understanding regional brain functions and pathologies.</p>
</div>
</section><section><h2 class="section-heading" id="differential-expression-analysis-for-multiple-samples">Differential Expression Analysis for Multiple Samples<a class="anchor" aria-label="anchor" href="#differential-expression-analysis-for-multiple-samples"></a></h2>
<hr class="half-width"><p>Till now we have explored spot-level analysis within one sample.
However, most studies (including the published one that furnished the
data we used here) include multiple samples. This has some similarities
to single cell-level DE analysis in scRNA-seq. In the scRNA-seq setting,
several studies, including one by <a href="https://www.nature.com/articles/s41467-021-25960-2" class="external-link">Squair, et
al.</a>, have emphasized drawbacks of performing DE on individual cells.
Single-cell DE may be caused by technical variability across single
cells, biological variability that may not be of interest across
conditions studied (<em>e.g.</em>, transcriptional bursting), or simply
by the large number of individual cells that inflate false positives.
One solution is to perform <em>pseudobulking</em> of the scRNA-seq and
then to use conventional bulk RNA-seq DE tools (<em>e.g.</em>, DESeq2)
to compare the pseudobulks. A Seurat <a href="https://satijalab.org/seurat/articles/de_vignette" class="external-link">vignette</a>
describing the approach for scRNA-seq should likewise be applicable for
spatial transcriptomics. However, we note the caveat that spots are not
cells – the former may have considerably more intra- and inter-spot
biological heterogeneity. Hence, pseudobulking should be done with
caution.</p>
<p>We selected one sample from each subject (<code>151508</code>,
<code>151669</code>, and <code>151673</code>), as detailed in the
referenced study (<a href="https://doi.org/10.1038/s41593-020-00787-0" class="external-link">Maynard et al, Nat
Neurosci 24, 425–436 (2021)</a>). This selection allows us to analyze
the consistency of differential expression across individuals while
accounting for subject-specific batch effects. We start by loading our
samples:</p>
<div class="codewrapper sourceCode" id="cb29">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Define the list of selected sample IDs</span></span>
<span><span class="va">selected_samples</span> <span class="op">&lt;-</span> <span class="fu">c</span><span class="op">(</span><span class="st">"151508"</span>, <span class="st">"151669"</span>,<span class="st">"151673"</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Initialize an empty list to store Seurat objects</span></span>
<span><span class="va">st_objects</span> <span class="op">&lt;-</span> <span class="fu">list</span><span class="op">(</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Loop over each sample ID to process the corresponding data</span></span>
<span><span class="kw">for</span> <span class="op">(</span><span class="va">sample_id</span> <span class="kw">in</span> <span class="va">selected_samples</span><span class="op">)</span> <span class="op">{</span></span>
<span>  <span class="co"># Construct the directory and filename for data loading</span></span>
<span>  <span class="va">data_dir</span> <span class="op">&lt;-</span> <span class="fu">paste0</span><span class="op">(</span><span class="st">"./data/"</span>, <span class="va">sample_id</span><span class="op">)</span></span>
<span>  <span class="va">filename</span> <span class="op">&lt;-</span> <span class="fu">paste0</span><span class="op">(</span><span class="va">sample_id</span>, <span class="st">"_filtered_feature_bc_matrix.h5"</span><span class="op">)</span></span>
<span>  </span>
<span>  <span class="co"># Load the spatial transcriptomics data for the current sample</span></span>
<span>  <span class="va">st_obj</span> <span class="op">&lt;-</span> <span class="fu">Load10X_Spatial</span><span class="op">(</span>data.dir <span class="op">=</span> <span class="va">data_dir</span>, filename <span class="op">=</span> <span class="va">filename</span><span class="op">)</span></span>
<span>  </span>
<span>  <span class="co"># Read in the tissue spot positions from the CSV file</span></span>
<span>  <span class="va">tissue_positions_file</span> <span class="op">&lt;-</span> <span class="fu">paste0</span><span class="op">(</span><span class="va">data_dir</span>, <span class="st">"/spatial/tissue_positions_list.csv"</span><span class="op">)</span></span>
<span>  <span class="va">tissue_position</span> <span class="op">&lt;-</span> <span class="fu">read_csv</span><span class="op">(</span><span class="va">tissue_positions_file</span>, </span>
<span>                              col_names <span class="op">=</span> <span class="cn">FALSE</span>, show_col_types <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span> <span class="op">%&gt;%</span></span>
<span>                     <span class="fu">column_to_rownames</span><span class="op">(</span><span class="st">'X1'</span><span class="op">)</span></span>
<span>  </span>
<span>  <span class="co"># Define column names for the tissue position data</span></span>
<span>  <span class="fu">colnames</span><span class="op">(</span><span class="va">tissue_position</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu">c</span><span class="op">(</span><span class="st">"in_tissue"</span>, <span class="st">"array_row"</span>, <span class="st">"array_col"</span>, </span>
<span>                                 <span class="st">"pxl_row_in_fullres"</span>, <span class="st">"pxl_col_in_fullres"</span><span class="op">)</span></span>
<span>  </span>
<span>  <span class="co"># Align spot barcodes between the Seurat object and the tissue position metadata</span></span>
<span>  <span class="va">tissue_position</span> <span class="op">&lt;-</span> <span class="va">tissue_position</span><span class="op">[</span><span class="fu">Cells</span><span class="op">(</span><span class="va">st_obj</span><span class="op">)</span>,<span class="op">]</span></span>
<span>  <span class="fu">stopifnot</span><span class="op">(</span><span class="fu">rownames</span><span class="op">(</span><span class="va">tissue_position</span><span class="op">)</span> <span class="op">==</span> <span class="fu">Cells</span><span class="op">(</span><span class="va">st_obj</span><span class="op">)</span><span class="op">)</span></span>
<span>  </span>
<span>  <span class="co"># Add the aligned tissue position metadata to the Seurat object</span></span>
<span>  <span class="va">st_obj</span> <span class="op">&lt;-</span> <span class="fu">AddMetaData</span><span class="op">(</span>object <span class="op">=</span> <span class="va">st_obj</span>, metadata <span class="op">=</span> <span class="va">tissue_position</span><span class="op">)</span></span>
<span></span>
<span>  <span class="co"># Additionally we add the "layer_guess" and "cell-count" metadata </span></span>
<span>  <span class="va">spot_metadata</span> <span class="op">&lt;-</span> <span class="fu">read.table</span><span class="op">(</span><span class="st">"./data/spot-meta.tsv"</span>, sep<span class="op">=</span><span class="st">"\t"</span><span class="op">)</span></span>
<span>  <span class="co"># Subset to our sample</span></span>
<span>  <span class="va">spot_metadata</span> <span class="op">&lt;-</span> <span class="fu">subset</span><span class="op">(</span><span class="va">spot_metadata</span>, <span class="va">sample_name</span> <span class="op">==</span> <span class="va">sample_id</span><span class="op">)</span></span>
<span>  <span class="fu">rownames</span><span class="op">(</span><span class="va">spot_metadata</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="va">spot_metadata</span><span class="op">$</span><span class="va">barcode</span></span>
<span>  <span class="fu">stopifnot</span><span class="op">(</span><span class="fu">all</span><span class="op">(</span><span class="fu">Cells</span><span class="op">(</span><span class="va">st_obj</span><span class="op">)</span> <span class="op">%in%</span> <span class="fu">rownames</span><span class="op">(</span><span class="va">spot_metadata</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span>
<span>  <span class="va">spot_metadata</span> <span class="op">&lt;-</span> <span class="va">spot_metadata</span><span class="op">[</span><span class="fu">Cells</span><span class="op">(</span><span class="va">st_obj</span><span class="op">)</span>,<span class="op">]</span></span>
<span>  </span>
<span>  <span class="va">st_obj</span> <span class="op">&lt;-</span> <span class="fu">AddMetaData</span><span class="op">(</span>object <span class="op">=</span> <span class="va">st_obj</span>,</span>
<span>                           metadata <span class="op">=</span> <span class="va">spot_metadata</span><span class="op">[</span>, <span class="fu">c</span><span class="op">(</span><span class="st">"layer_guess"</span>, <span class="st">"cell_count"</span><span class="op">)</span>, drop<span class="op">=</span><span class="cn">FALSE</span><span class="op">]</span><span class="op">)</span></span>
<span></span>
<span>  <span class="co"># Store the updated Seurat object in the list, using the sample ID as the key</span></span>
<span>  <span class="va">st_objects</span><span class="op">[[</span><span class="va">sample_id</span><span class="op">]</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="va">st_obj</span></span>
<span><span class="op">}</span></span></code></pre>
</div>
<p>Now, st_objects variable contains all the Seurat objects for each
selected sample, indexed by their respective sample IDs.</p>
<div class="section level3">
<h3 id="batch-correction-analysis">Batch correction Analysis<a class="anchor" aria-label="anchor" href="#batch-correction-analysis"></a></h3>
<p>When working with multiple samples, batch effects can introduce
technical variability that may confound biological signals. Batch
correction methods can help mitigate these effects. In this section, we
perform pseudobulk analysis and apply PCA to visualize any potential
batch effects across layers.</p>
<div class="codewrapper sourceCode" id="cb30">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Initialize an empty list to store pseudobulked Seurat objects</span></span>
<span><span class="va">pseudobulked_objs</span> <span class="op">&lt;-</span> <span class="fu">list</span><span class="op">(</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Pseudobulk each selected sample by aggregating expression within annotated layers</span></span>
<span><span class="kw">for</span> <span class="op">(</span><span class="va">i</span> <span class="kw">in</span> <span class="va">selected_samples</span><span class="op">)</span> <span class="op">{</span></span>
<span>  <span class="va">pseudobulk_seurat</span> <span class="op">&lt;-</span> <span class="fu">AggregateExpression</span><span class="op">(</span><span class="va">st_objects</span><span class="op">[[</span><span class="va">i</span><span class="op">]</span><span class="op">]</span>, </span>
<span>                                           assays <span class="op">=</span> <span class="st">'Spatial'</span>, </span>
<span>                                           return.seurat <span class="op">=</span> <span class="cn">TRUE</span>, </span>
<span>                                           group.by <span class="op">=</span> <span class="st">'layer_guess'</span><span class="op">)</span></span>
<span>  <span class="va">pseudobulk_seurat</span><span class="op">@</span><span class="va">meta.data</span><span class="op">$</span><span class="va">sample</span> <span class="op">&lt;-</span> <span class="fu">names</span><span class="op">(</span><span class="va">st_objects</span><span class="op">[</span><span class="va">i</span><span class="op">]</span><span class="op">)</span></span>
<span>  <span class="va">pseudobulk_seurat</span><span class="op">@</span><span class="va">meta.data</span><span class="op">$</span><span class="va">layer</span> <span class="op">&lt;-</span> <span class="fu">colnames</span><span class="op">(</span><span class="va">pseudobulk_seurat</span><span class="op">@</span><span class="va">assays</span><span class="op">[[</span><span class="st">"Spatial"</span><span class="op">]</span><span class="op">]</span><span class="op">)</span></span>
<span>  </span>
<span>  <span class="co"># Store the pseudobulked Seurat object in the list</span></span>
<span>  <span class="va">pseudobulked_objs</span><span class="op">[[</span><span class="va">i</span><span class="op">]</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="va">pseudobulk_seurat</span></span>
<span><span class="op">}</span></span>
<span></span>
<span><span class="co"># Merge all pseudobulked Seurat objects into a single Seurat object</span></span>
<span><span class="va">merged_pseudobulk</span> <span class="op">&lt;-</span> <span class="fu">Reduce</span><span class="op">(</span><span class="kw">function</span><span class="op">(</span><span class="va">x</span>, <span class="va">y</span><span class="op">)</span> <span class="fu">merge</span><span class="op">(</span><span class="va">x</span>, <span class="va">y</span><span class="op">)</span>, <span class="va">pseudobulked_objs</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Normalize, find variable features, and scale the data</span></span>
<span><span class="va">merged_pseudobulk</span> <span class="op">&lt;-</span> <span class="fu">NormalizeData</span><span class="op">(</span><span class="va">merged_pseudobulk</span>, normalization.method <span class="op">=</span> <span class="st">"LogNormalize"</span>, scale.factor <span class="op">=</span> <span class="fl">1e6</span><span class="op">)</span></span>
<span><span class="va">merged_pseudobulk</span> <span class="op">&lt;-</span> <span class="fu">FindVariableFeatures</span><span class="op">(</span><span class="va">merged_pseudobulk</span><span class="op">)</span></span>
<span><span class="va">merged_pseudobulk</span> <span class="op">&lt;-</span> <span class="fu">ScaleData</span><span class="op">(</span><span class="va">merged_pseudobulk</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Run PCA on the merged pseudobulked data</span></span>
<span><span class="va">merged_pseudobulk</span> <span class="op">&lt;-</span> <span class="fu">RunPCA</span><span class="op">(</span><span class="va">merged_pseudobulk</span>, features <span class="op">=</span> <span class="fu">VariableFeatures</span><span class="op">(</span>object <span class="op">=</span> <span class="va">merged_pseudobulk</span><span class="op">)</span>, npcs <span class="op">=</span> <span class="fl">2</span><span class="op">)</span></span></code></pre>
</div>
<p>To assess the presence of batch effects, we visualize the PCA
results. Ideally, the samples should cluster by layer rather than by
sample, indicating that batch effects are not a significant concern.</p>
<div class="codewrapper sourceCode" id="cb31">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Visualize PCA grouped by layer with increased point size and different shapes</span></span>
<span><span class="va">PCAPlot_layer_symbol</span> <span class="op">&lt;-</span> <span class="fu">DimPlot</span><span class="op">(</span><span class="va">merged_pseudobulk</span>, reduction <span class="op">=</span> <span class="st">"pca"</span>, shape.by <span class="op">=</span> <span class="st">"sample"</span>, pt.size <span class="op">=</span> <span class="fl">3</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu">theme</span><span class="op">(</span>legend.position <span class="op">=</span> <span class="st">"right"</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu">ggtitle</span><span class="op">(</span><span class="st">"PCA Plot Annotated by Layer (Shapes)"</span><span class="op">)</span></span>
<span><span class="va">PCAPlot_layer_symbol</span></span></code></pre>
</div>
<p><img src="fig/differential-expression-testing-rendered-pca_results-1.png" style="display: block; margin: auto;" class="figure"><img src="fig/de_pca.png" alt="" class="figure"></p>
<p>The PCA plot indeed shows that the dots are clustered by layer rather
than by sample, validating that the layers are the primary source of
variation, rather than differences between samples from different
subjects.</p>
<p>If the dots were instead clustered by sample rather than by layer, it
would indicate the presence of batch effects. In bulk or single-cell
RNA-seq studies, batch effects would typically be mitigated by applying
regression-based tools such as <a href="https://portals.broadinstitute.org/harmony/articles/quickstart.html" class="external-link"><code>Harmony</code></a>
or by incorporating a batch factor in a linear modeling framework
(<em>e.g.</em>, of DESeq2). Those approaches could be applied to spatial
transcriptomics data, as well.</p>
</div>
<div class="section level3">
<h3 id="pseudobulk-analysis-with-deseq2">Pseudobulk Analysis with DESeq2<a class="anchor" aria-label="anchor" href="#pseudobulk-analysis-with-deseq2"></a></h3>
<p>Pseudobulk analysis allows for the aggregation of single-cell or
spot-level data into larger groups, which can reduce noise and enhance
the detection of biologically meaningful signals. In this section, we
perform differential expression analysis using DESeq2 on pseudobulked
data. DESeq2 is a tool for differential expression analysis,
particularly well-suited for count data, which is the type of data we
have after aggregating the spatial transcriptomics spots into
pseudobulks. It employs a statistical model that accounts for
variability in read counts while controlling for factors such as sample
size and sequencing depth. This makes it particularly powerful in
situations where we expect relatively small changes in gene expression
but need to ensure robustness against noise, making it an ideal choice
for our pseudobulk analysis.</p>
<p>In this case, we compare White Matter (aggregated expression from
spots labeled as White Matter) versus the rest of the tissue (aggregated
expression from spots belonging to all other cotrical layers). This
binary comparison is straightforward yet powerful, as the White Matter
in the brain is distinct from the cortical layers in terms of both
structure and function. By comparing the gene expression in White Matter
to that in all other layers, we aim to identify genes that are uniquely
or predominantly expressed in White Matter and associated with its
functions or pathologies</p>
<div class="codewrapper sourceCode" id="cb32">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">all_columns</span> <span class="op">&lt;-</span> <span class="fu">list</span><span class="op">(</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Function to process each sample and create a new matrix with WM vs. others</span></span>
<span><span class="va">process_sample</span> <span class="op">&lt;-</span> <span class="kw">function</span><span class="op">(</span><span class="va">sample_name</span>, <span class="va">pseudobulked_objs</span><span class="op">)</span> <span class="op">{</span></span>
<span>  <span class="va">counts_matrix</span> <span class="op">&lt;-</span> <span class="va">pseudobulked_objs</span><span class="op">[[</span><span class="va">sample_name</span><span class="op">]</span><span class="op">]</span><span class="op">@</span><span class="va">assays</span><span class="op">[[</span><span class="st">"Spatial"</span><span class="op">]</span><span class="op">]</span><span class="op">@</span><span class="va">layers</span><span class="op">[[</span><span class="st">"counts"</span><span class="op">]</span><span class="op">]</span></span>
<span>  <span class="va">counts_matrix</span> <span class="op">&lt;-</span> <span class="fu">as</span><span class="op">(</span><span class="va">counts_matrix</span>, <span class="st">"dgCMatrix"</span><span class="op">)</span></span>
<span>  <span class="va">layer_info</span> <span class="op">&lt;-</span> <span class="va">pseudobulked_objs</span><span class="op">[[</span><span class="va">sample_name</span><span class="op">]</span><span class="op">]</span><span class="op">@</span><span class="va">meta.data</span><span class="op">[[</span><span class="st">"layer_guess"</span><span class="op">]</span><span class="op">]</span></span>
<span>  <span class="va">white_matter_col</span> <span class="op">&lt;-</span> <span class="fu">which</span><span class="op">(</span><span class="va">layer_info</span> <span class="op">==</span> <span class="st">"WM"</span><span class="op">)</span></span>
<span>  <span class="va">other_counts</span> <span class="op">&lt;-</span> <span class="fu">rowSums</span><span class="op">(</span><span class="va">counts_matrix</span><span class="op">[</span>, <span class="op">-</span><span class="va">white_matter_col</span><span class="op">]</span><span class="op">)</span></span>
<span>  <span class="va">white_matter_counts</span> <span class="op">&lt;-</span> <span class="va">counts_matrix</span><span class="op">[</span>, <span class="va">white_matter_col</span><span class="op">]</span></span>
<span>  <span class="va">new_matrix</span> <span class="op">&lt;-</span> <span class="fu">cbind</span><span class="op">(</span><span class="va">white_matter_counts</span>, <span class="va">other_counts</span><span class="op">)</span></span>
<span>  <span class="fu">colnames</span><span class="op">(</span><span class="va">new_matrix</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu">c</span><span class="op">(</span><span class="fu">paste0</span><span class="op">(</span><span class="st">"white_matter_"</span>, <span class="va">sample_name</span><span class="op">)</span>, <span class="fu">paste0</span><span class="op">(</span><span class="st">"other_"</span>, <span class="va">sample_name</span><span class="op">)</span><span class="op">)</span></span>
<span>  <span class="kw">return</span><span class="op">(</span><span class="va">new_matrix</span><span class="op">)</span></span>
<span><span class="op">}</span></span>
<span></span>
<span><span class="co"># Process each sample and store the resulting matrices</span></span>
<span><span class="kw">for</span> <span class="op">(</span><span class="va">sample</span> <span class="kw">in</span> <span class="va">selected_samples</span><span class="op">)</span> <span class="op">{</span></span>
<span>  <span class="va">new_matrix</span> <span class="op">&lt;-</span> <span class="fu">process_sample</span><span class="op">(</span><span class="va">sample</span>, <span class="va">pseudobulked_objs</span><span class="op">)</span></span>
<span>  <span class="va">all_columns</span><span class="op">[[</span><span class="va">sample</span><span class="op">]</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="va">new_matrix</span></span>
<span><span class="op">}</span></span>
<span></span>
<span><span class="co"># Combine all processed matrices into a single matrix</span></span>
<span><span class="va">combined_matrix</span> <span class="op">&lt;-</span> <span class="fu">do.call</span><span class="op">(</span><span class="va">cbind</span>, <span class="va">all_columns</span><span class="op">)</span></span>
<span><span class="fu">rownames</span><span class="op">(</span><span class="va">combined_matrix</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu">rownames</span><span class="op">(</span><span class="va">pseudobulked_objs</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Create a dataframe for DESeq2 with sample and layer information</span></span>
<span><span class="va">sample</span> <span class="op">&lt;-</span> <span class="fu">rep</span><span class="op">(</span><span class="va">selected_samples</span>, each <span class="op">=</span> <span class="fl">2</span><span class="op">)</span></span>
<span><span class="va">layer</span> <span class="op">&lt;-</span> <span class="fu">rep</span><span class="op">(</span><span class="fu">c</span><span class="op">(</span><span class="st">'WM'</span>, <span class="st">'others'</span><span class="op">)</span>, times <span class="op">=</span> <span class="fu">length</span><span class="op">(</span><span class="va">selected_samples</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">new_dataframe</span> <span class="op">&lt;-</span> <span class="fu">data.frame</span><span class="op">(</span>sample <span class="op">=</span> <span class="va">sample</span>, layer <span class="op">=</span> <span class="va">layer</span><span class="op">)</span></span>
<span><span class="fu">rownames</span><span class="op">(</span><span class="va">new_dataframe</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu">colnames</span><span class="op">(</span><span class="va">combined_matrix</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Perform differential expression analysis using DESeq2</span></span>
<span><span class="co"># Unfortunately, there are still dependency issues between DESeq2 and Seurat,</span></span>
<span><span class="co"># so we recommend not running this part. To prevent execution, we wrap the code in an `if (FALSE)` statement.</span></span>
<span><span class="co">#if (FALSE) {</span></span>
<span><span class="co">#  BiocManager::install("DESeq2")</span></span>
<span><span class="co">#  library(DESeq2)</span></span>
<span><span class="va">dds</span> <span class="op">&lt;-</span> <span class="fu">DESeqDataSetFromMatrix</span><span class="op">(</span>countData <span class="op">=</span> <span class="va">combined_matrix</span>,</span>
<span>                              colData <span class="op">=</span> <span class="va">new_dataframe</span>,</span>
<span>                              design <span class="op">=</span> <span class="op">~</span> <span class="va">layer</span> <span class="op">+</span> <span class="va">sample</span><span class="op">)</span></span>
<span><span class="va">dds</span> <span class="op">&lt;-</span> <span class="fu">DESeq</span><span class="op">(</span><span class="va">dds</span><span class="op">)</span></span>
<span><span class="fu">resultsNames</span><span class="op">(</span><span class="va">dds</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>[1] "Intercept"               "layer_WM_vs_others"
[3] "sample_151669_vs_151508" "sample_151673_vs_151508"</code></pre>
</div>
<div class="codewrapper sourceCode" id="cb34">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">res</span> <span class="op">&lt;-</span> <span class="fu">results</span><span class="op">(</span><span class="va">dds</span>, name <span class="op">=</span> <span class="st">"layer_WM_vs_others"</span><span class="op">)</span></span>
<span><span class="fu">head</span><span class="op">(</span><span class="va">res</span><span class="op">[</span><span class="fu">order</span><span class="op">(</span><span class="va">res</span><span class="op">$</span><span class="va">padj</span><span class="op">)</span>,<span class="op">]</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>log2 fold change (MLE): layer WM vs others
Wald test p-value: layer WM vs others
DataFrame with 6 rows and 6 columns
        baseMean log2FoldChange     lfcSE      stat      pvalue        padj
       &lt;numeric&gt;      &lt;numeric&gt; &lt;numeric&gt; &lt;numeric&gt;   &lt;numeric&gt;   &lt;numeric&gt;
MOBP    1364.273        3.04042  0.334509   9.08921 9.97603e-20 1.26825e-15
AQP1     160.269        3.22379  0.359589   8.96523 3.09639e-19 1.31215e-15
MBP    16027.210        2.72729  0.303229   8.99418 2.37995e-19 1.31215e-15
MOG      422.622        2.94064  0.357036   8.23624 1.77704e-16 5.64789e-13
MYRF     318.623        2.72961  0.338574   8.06208 7.50046e-16 1.90707e-12
NKX6-2   325.542        2.91717  0.368060   7.92579 2.26698e-15 4.80335e-12</code></pre>
</div>
<div class="codewrapper sourceCode" id="cb36">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">#}</span></span></code></pre>
</div>
<figure><img src="fig/de_genes.PNG" alt="Genes with the lowest adjusted p-values from differential expression analysis with DESeq2" class="figure mx-auto d-block"><div class="figcaption">Visium spatial gene expression slide</div>
</figure><p>We have selected to plot spatially the differentially expressed genes
with the highest mean values, namely “MBP” and “MOBP”</p>
<div class="codewrapper sourceCode" id="cb37">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">plot_sample_expression</span> <span class="op">&lt;-</span> <span class="kw">function</span><span class="op">(</span><span class="va">sample</span>, <span class="va">gene</span>, <span class="va">sample_id</span><span class="op">)</span> <span class="op">{</span></span>
<span>  <span class="va">sample</span><span class="op">@</span><span class="va">meta.data</span><span class="op">$</span><span class="va">temp_feat</span> <span class="op">&lt;-</span> <span class="va">sample</span><span class="op">@</span><span class="va">assays</span><span class="op">[[</span><span class="st">"Spatial"</span><span class="op">]</span><span class="op">]</span><span class="op">$</span><span class="va">counts</span><span class="op">[</span><span class="va">gene</span>,<span class="op">]</span></span>
<span>  <span class="va">p1</span> <span class="op">&lt;-</span> <span class="fu">SpatialFeaturePlot</span><span class="op">(</span><span class="va">sample</span>, features <span class="op">=</span> <span class="st">"temp_feat"</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu">ggtitle</span><span class="op">(</span><span class="fu">paste</span><span class="op">(</span><span class="st">"Spatial Expression of"</span>, <span class="va">gene</span>, <span class="st">"in Sample"</span>, <span class="va">sample_id</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu">theme</span><span class="op">(</span>plot.title <span class="op">=</span> <span class="fu">element_text</span><span class="op">(</span>hjust <span class="op">=</span> <span class="fl">0.5</span><span class="op">)</span><span class="op">)</span></span>
<span>  <span class="kw">return</span><span class="op">(</span><span class="va">p1</span><span class="op">)</span></span>
<span><span class="op">}</span></span>
<span></span>
<span><span class="va">plot_layer_expression</span> <span class="op">&lt;-</span> <span class="kw">function</span><span class="op">(</span><span class="va">sample</span>, <span class="va">sample_id</span><span class="op">)</span> <span class="op">{</span></span>
<span>  <span class="va">p1</span> <span class="op">&lt;-</span> <span class="fu">SpatialDimPlot</span><span class="op">(</span><span class="va">sample</span>, group.by <span class="op">=</span> <span class="st">"layer_guess"</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu">ggtitle</span><span class="op">(</span><span class="fu">paste</span><span class="op">(</span><span class="st">"Layer guess of Sample"</span>, <span class="va">sample_id</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu">theme</span><span class="op">(</span>plot.title <span class="op">=</span> <span class="fu">element_text</span><span class="op">(</span>hjust <span class="op">=</span> <span class="fl">0.5</span><span class="op">)</span>,</span>
<span>          legend.text <span class="op">=</span> <span class="fu">element_text</span><span class="op">(</span>size <span class="op">=</span> <span class="fl">12</span><span class="op">)</span><span class="op">)</span></span>
<span>  <span class="kw">return</span><span class="op">(</span><span class="va">p1</span><span class="op">)</span></span>
<span><span class="op">}</span></span>
<span></span>
<span><span class="co"># Plot and save the top DE genes across samples</span></span>
<span><span class="va">plots</span> <span class="op">&lt;-</span> <span class="fu">list</span><span class="op">(</span><span class="op">)</span></span>
<span><span class="kw">for</span> <span class="op">(</span><span class="va">sample_id</span> <span class="kw">in</span> <span class="va">selected_samples</span><span class="op">)</span> <span class="op">{</span></span>
<span>  <span class="va">p1</span> <span class="op">&lt;-</span> <span class="fu">plot_sample_expression</span><span class="op">(</span><span class="va">st_objects</span><span class="op">[[</span><span class="va">sample_id</span><span class="op">]</span><span class="op">]</span>, <span class="st">"MOBP"</span>, <span class="va">sample_id</span><span class="op">)</span></span>
<span>  <span class="va">p2</span> <span class="op">&lt;-</span> <span class="fu">plot_sample_expression</span><span class="op">(</span><span class="va">st_objects</span><span class="op">[[</span><span class="va">sample_id</span><span class="op">]</span><span class="op">]</span>, <span class="st">"MBP"</span> , <span class="va">sample_id</span><span class="op">)</span></span>
<span>  <span class="va">p3</span> <span class="op">&lt;-</span> <span class="fu">plot_layer_expression</span><span class="op">(</span><span class="va">st_objects</span><span class="op">[[</span><span class="va">sample_id</span><span class="op">]</span><span class="op">]</span>, <span class="va">sample_id</span><span class="op">)</span></span>
<span>  <span class="va">combined</span> <span class="op">&lt;-</span> <span class="va">p1</span> <span class="op">+</span> <span class="va">p2</span> <span class="op">+</span> <span class="va">p3</span></span>
<span>  <span class="va">plots</span><span class="op">[[</span><span class="va">sample_id</span><span class="op">]</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="va">combined</span></span>
<span><span class="op">}</span></span>
<span></span>
<span><span class="co"># Combine and save plots</span></span>
<span><span class="va">combined_plot</span> <span class="op">&lt;-</span> <span class="fu">wrap_plots</span><span class="op">(</span><span class="va">plots</span>, ncol <span class="op">=</span> <span class="fl">1</span><span class="op">)</span></span></code></pre>
</div>
<figure><img src="fig/de_genes_plot.png" class="figure mx-auto d-block"><div class="figcaption">Spatial distribution of the most DE genes for
all samples</div>
</figure><p>As it is evident from the spatila distribution of the genes, they are
mostly expressed specifically in the White Matter for all samples.</p>
</div>
</section><section><h2 class="section-heading" id="other-considerations">Other Considerations<a class="anchor" aria-label="anchor" href="#other-considerations"></a></h2>
<hr class="half-width"><p>We here explored the very basics of differential expression analysis
and differential expression analysis with multiple samples. For batch
correction procedures, such as those just mentioned, may remove
biological variability along with technical variability. An alternative
approach is to perform analyses <em>within</em> samples and then to
perform meta-analysis across them. One intriguing possibility is to
compare rank-based statistics across samples, which should be less
susceptible to batch effects. Another is to cluster across samples based
on overlapping sets of genes within intra-cluster samples. Such an
approach has been applied in the “meta-programs” defined in cancer by <a href="https://pubmed.ncbi.nlm.nih.gov/37258682/" class="external-link">Gavish and
colleagues</a>.</p>
<p>There is unlikely to be one turn-key approach to addressing the above
issues. You will likely need to explore the data – and that exploration
should be guided by some biological expectations that anchor your
choices.</p>
<div id="keypoints1" class="callout keypoints">
<div class="callout-square">
<i class="callout-icon" data-feather="key"></i>
</div>
<div class="callout-inner">
<h3 class="callout-title">Key Points</h3>
<div class="callout-content">
<ul><li>Differential expression testing pinpoints genes with significant
expression variations across regions or clusters.</li>
<li>Moran’s I statistic reveals spatial autocorrelation in gene
expression, critical for examining spatially dependent biological
activities.</li>
<li>Moran’s I algorithm effectively identifies genes expressed in
anatomically distinct regions, as validated from the correlation
analysis with the DE genes from the annotated regions.</li>
</ul></div>
</div>
</div>
<p>#<code>{r echo=FALSE,include=FALSE} #save_seurat_object(obj = sct_st, file_prefix = 'detesting') #</code></p>
<!--
Place links that you need to refer to multiple times across pages here. Delete
any links that you are not going to use.
 -->
</section></div> <!-- / div.lesson-content -->
    </main><!-- / main#main-content.main-content --><nav class="bottom-pagination mx-md-4" aria-label="Previous and Next Chapter"><div class="d-block d-sm-block d-md-none">
        <a class="chapter-link" href="deconvolve-cell-types-in-a-spot.html"><i aria-hidden="true" class="small-arrow" data-feather="arrow-left"></i>Previous</a>
        <a class="chapter-link float-end" href="final-exercise.html">Next<i aria-hidden="true" class="small-arrow" data-feather="arrow-right"></i></a>
      </div>
      <!-- content for large screens -->
      <div class="d-none d-sm-none d-md-block">
        <a class="chapter-link" href="deconvolve-cell-types-in-a-spot.html" rel="prev">
          <i aria-hidden="true" class="small-arrow" data-feather="arrow-left"></i>
          Previous: Deconvolution in
        </a>
        <a class="chapter-link float-end" href="final-exercise.html" rel="next">
          Next: Putting it all...
          <i aria-hidden="true" class="small-arrow" data-feather="arrow-right"></i>
        </a>
      </div>
    </nav></div> <!-- / div.primary-content.col-xs-12 -->
<!-- END:   inst/pkgdown/templates/content-instructor.html-->

      </div><!--/div.row-->
      		<footer class="row footer mx-md-3"><hr><div class="col-md-6">
        <p>This lesson is subject to the <a href="CODE_OF_CONDUCT.html">Code of Conduct</a></p>
        <p>

        <a href="https://github.com/smcclatchy/spatial-transcriptomics/edit/main/episodes/differential-expression-testing.Rmd" class="external-link">Edit on GitHub</a>

	
        | <a href="https://github.com/smcclatchy/spatial-transcriptomics/blob/main/CONTRIBUTING.md" class="external-link">Contributing</a>
        | <a href="https://github.com/smcclatchy/spatial-transcriptomics/" class="external-link">Source</a></p>
				<p><a href="https://github.com/smcclatchy/spatial-transcriptomics/blob/main/CITATION.cff" class="external-link">Cite</a> | <a href="mailto:susan.mcclatchy@jax.org">Contact</a> | <a href="https://carpentries.org/about/" class="external-link">About</a></p>
			</div>
			<div class="col-md-6">

        <p>Materials licensed under <a href="LICENSE.html">CC-BY 4.0</a> by the authors</p>

        <p>Template licensed under <a href="https://creativecommons.org/licenses/by-sa/4.0/" class="external-link">CC-BY 4.0</a> by <a href="https://carpentries.org/" class="external-link">The Carpentries</a></p>
        <p>Built with <a href="https://github.com/epiverse-trace/sandpaper/tree/patch-renv-github-bug" class="external-link">sandpaper (0.16.11)</a>, <a href="https://github.com/carpentries/pegboard/tree/5e0d711121a581777d83cf270e298c933f112d52" class="external-link">pegboard (0.7.9)</a>, and <a href="https://github.com/carpentries/varnish/tree/1.0.5" class="external-link">varnish (1.0.5)</a></p>
			</div>
		</footer></div> <!-- / div.container -->
	<div id="to-top">
		<a href="#top">
      <i class="search-icon" data-feather="arrow-up" role="img" aria-label="Back To Top"></i><br><!-- <span class="d-none d-sm-none d-md-none d-lg-none d-xl-block">Back</span> To Top --><span class="d-none d-sm-none d-md-none d-lg-none d-xl-block">Back</span> To Top
		</a>
	</div>
  <script type="application/ld+json">
    {
  "@context": "https://schema.org",
  "@type": "TrainingMaterial",
  "@id": "https://smcclatchy.github.io/spatial-transcriptomics/differential-expression-testing.html",
  "inLanguage": "en",
  "dct:conformsTo": "https://bioschemas.org/profiles/TrainingMaterial/1.0-RELEASE",
  "description": "A Carpentries Lesson teaching foundational data and coding skills to researchers worldwide",
  "keywords": "software, data, lesson, The Carpentries, transcriptome, Visium, cancer",
  "name": "Differential Expression Testing",
  "creativeWorkStatus": "active",
  "url": "https://smcclatchy.github.io/spatial-transcriptomics/differential-expression-testing.html",
  "identifier": "https://smcclatchy.github.io/spatial-transcriptomics/differential-expression-testing.html",
  "dateCreated": "2024-03-01",
  "dateModified": "2025-03-18",
  "datePublished": "2025-03-19"
}

  </script><script>
		feather.replace();
	</script></body></html><!-- END:   inst/pkgdown/templates/layout.html-->

