<!DOCTYPE html>
<!-- START: inst/pkgdown/templates/layout.html --><!-- Generated by pkgdown: do not edit by hand --><html lang="en" data-bs-theme="auto"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><title>Spatial Transcriptomics: Remove Low-quality Spots</title><meta name="viewport" content="width=device-width, initial-scale=1"><script src="assets/themetoggle.js"></script><link rel="stylesheet" type="text/css" href="assets/styles.css"><script src="assets/scripts.js" type="text/javascript"></script><!-- mathjax --><script type="text/x-mathjax-config">
    MathJax.Hub.Config({
      config: ["MMLorHTML.js"],
      jax: ["input/TeX","input/MathML","output/HTML-CSS","output/NativeMML", "output/PreviewHTML"],
      extensions: ["tex2jax.js","mml2jax.js","MathMenu.js","MathZoom.js", "fast-preview.js", "AssistiveMML.js", "a11y/accessibility-menu.js"],
      TeX: {
        extensions: ["AMSmath.js","AMSsymbols.js","noErrors.js","noUndefined.js"]
      },
      tex2jax: {
        inlineMath: [['\\(', '\\)']],
        displayMath: [ ['$$','$$'], ['\\[', '\\]'] ],
        processEscapes: true
      }
    });
    </script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><!-- Responsive Favicon for The Carpentries --><link rel="apple-touch-icon" sizes="180x180" href="favicons/incubator/apple-touch-icon.png"><link rel="icon" type="image/png" sizes="32x32" href="favicons/incubator/favicon-32x32.png"><link rel="icon" type="image/png" sizes="16x16" href="favicons/incubator/favicon-16x16.png"><link rel="manifest" href="favicons/incubator/site.webmanifest"><link rel="mask-icon" href="favicons/incubator/safari-pinned-tab.svg" color="#5bbad5"><meta name="msapplication-TileColor" content="#da532c"><meta name="theme-color" media="(prefers-color-scheme: light)" content="white"><meta name="theme-color" media="(prefers-color-scheme: dark)" content="black"></head><body>
    <header id="top" class="navbar navbar-expand-md top-nav incubator"><svg xmlns="http://www.w3.org/2000/svg" class="d-none"><symbol id="check2" viewbox="0 0 16 16"><path d="M13.854 3.646a.5.5 0 0 1 0 .708l-7 7a.5.5 0 0 1-.708 0l-3.5-3.5a.5.5 0 1 1 .708-.708L6.5 10.293l6.646-6.647a.5.5 0 0 1 .708 0z"></path></symbol><symbol id="circle-half" viewbox="0 0 16 16"><path d="M8 15A7 7 0 1 0 8 1v14zm0 1A8 8 0 1 1 8 0a8 8 0 0 1 0 16z"></path></symbol><symbol id="moon-stars-fill" viewbox="0 0 16 16"><path d="M6 .278a.768.768 0 0 1 .08.858 7.208 7.208 0 0 0-.878 3.46c0 4.021 3.278 7.277 7.318 7.277.527 0 1.04-.055 1.533-.16a.787.787 0 0 1 .81.316.733.733 0 0 1-.031.893A8.349 8.349 0 0 1 8.344 16C3.734 16 0 12.286 0 7.71 0 4.266 2.114 1.312 5.124.06A.752.752 0 0 1 6 .278z"></path><path d="M10.794 3.148a.217.217 0 0 1 .412 0l.387 1.162c.173.518.579.924 1.097 1.097l1.162.387a.217.217 0 0 1 0 .412l-1.162.387a1.734 1.734 0 0 0-1.097 1.097l-.387 1.162a.217.217 0 0 1-.412 0l-.387-1.162A1.734 1.734 0 0 0 9.31 6.593l-1.162-.387a.217.217 0 0 1 0-.412l1.162-.387a1.734 1.734 0 0 0 1.097-1.097l.387-1.162zM13.863.099a.145.145 0 0 1 .274 0l.258.774c.115.346.386.617.732.732l.774.258a.145.145 0 0 1 0 .274l-.774.258a1.156 1.156 0 0 0-.732.732l-.258.774a.145.145 0 0 1-.274 0l-.258-.774a1.156 1.156 0 0 0-.732-.732l-.774-.258a.145.145 0 0 1 0-.274l.774-.258c.346-.115.617-.386.732-.732L13.863.1z"></path></symbol><symbol id="sun-fill" viewbox="0 0 16 16"><path d="M8 12a4 4 0 1 0 0-8 4 4 0 0 0 0 8zM8 0a.5.5 0 0 1 .5.5v2a.5.5 0 0 1-1 0v-2A.5.5 0 0 1 8 0zm0 13a.5.5 0 0 1 .5.5v2a.5.5 0 0 1-1 0v-2A.5.5 0 0 1 8 13zm8-5a.5.5 0 0 1-.5.5h-2a.5.5 0 0 1 0-1h2a.5.5 0 0 1 .5.5zM3 8a.5.5 0 0 1-.5.5h-2a.5.5 0 0 1 0-1h2A.5.5 0 0 1 3 8zm10.657-5.657a.5.5 0 0 1 0 .707l-1.414 1.415a.5.5 0 1 1-.707-.708l1.414-1.414a.5.5 0 0 1 .707 0zm-9.193 9.193a.5.5 0 0 1 0 .707L3.05 13.657a.5.5 0 0 1-.707-.707l1.414-1.414a.5.5 0 0 1 .707 0zm9.193 2.121a.5.5 0 0 1-.707 0l-1.414-1.414a.5.5 0 0 1 .707-.707l1.414 1.414a.5.5 0 0 1 0 .707zM4.464 4.465a.5.5 0 0 1-.707 0L2.343 3.05a.5.5 0 1 1 .707-.707l1.414 1.414a.5.5 0 0 1 0 .708z"></path></symbol></svg><a class="visually-hidden-focusable skip-link" href="#main-content">Skip to main content</a>
  <div class="container-fluid top-nav-container">
    <div class="col-md-8">
      <div class="large-logo">
        <img id="incubator-logo" alt="Carpentries Incubator" src="assets/images/incubator-logo.svg"><span class="badge text-bg-info">
          <abbr title="This lesson is in the beta phase, which means that it is ready for teaching by instructors outside of the original author team.">
            <a href="https://cdh.carpentries.org/the-lesson-life-cycle.html#polishing-beta-stage" class="external-link alert-link">
              <i aria-hidden="true" class="icon" data-feather="alert-circle" style="border-radius: 5px"></i>
              Beta
            </a>
            <span class="visually-hidden">This lesson is in the beta phase, which means that it is ready for teaching by instructors outside of the original author team.</span>
          </abbr>
        </span>

      </div>
    </div>
    <div class="selector-container">
      <div id="theme-selector">
        <li class="nav-item dropdown" id="theme-button-list">
          <button class="btn btn-link nav-link px-0 px-lg-2 dropdown-toggle d-flex align-items-center" id="bd-theme" type="button" aria-expanded="false" data-bs-toggle="dropdown" data-bs-display="static" aria-label="Toggle theme (auto)">
            <svg class="bi my-1 theme-icon-active"><use href="#circle-half"></use></svg><i data-feather="chevron-down"></i>
          </button>
          <ul class="dropdown-menu dropdown-menu-end" aria-labelledby="bd-theme-text"><li>
              <button type="button" class="btn dropdown-item d-flex align-items-center" data-bs-theme-value="light" aria-pressed="false">
                <svg class="bi me-2 theme-icon"><use href="#sun-fill"></use></svg>
                Light
                <svg class="bi ms-auto d-none"><use href="#check2"></use></svg></button>
            </li>
            <li>
              <button type="button" class="btn dropdown-item d-flex align-items-center" data-bs-theme-value="dark" aria-pressed="false">
                <svg class="bi me-2 theme-icon"><use href="#moon-stars-fill"></use></svg>
                Dark
                <svg class="bi ms-auto d-none"><use href="#check2"></use></svg></button>
            </li>
            <li>
              <button type="button" class="btn dropdown-item d-flex align-items-center active" data-bs-theme-value="auto" aria-pressed="true">
                <svg class="bi me-2 theme-icon"><use href="#circle-half"></use></svg>
                Auto
                <svg class="bi ms-auto d-none"><use href="#check2"></use></svg></button>
            </li>
          </ul></li>
      </div>

      <div class="dropdown" id="instructor-dropdown">
        <button class="btn btn-secondary dropdown-toggle bordered-button" type="button" id="dropdownMenu1" data-bs-toggle="dropdown" aria-expanded="false">
          <i aria-hidden="true" class="icon" data-feather="eye"></i> Learner View <i data-feather="chevron-down"></i>
        </button>
        <ul class="dropdown-menu" aria-labelledby="dropdownMenu1"><li><button class="dropdown-item" type="button" onclick="window.location.href='instructor/remove-low-quality-spots.html';">Instructor View</button></li>
        </ul></div>
    </div>
  </div>
  <hr></header><nav class="navbar navbar-expand-xl bottom-nav incubator" aria-label="Main Navigation"><div class="container-fluid nav-container">
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbarSupportedContent" aria-controls="navbarSupportedContent" aria-expanded="false" aria-label="Toggle Navigation">
      <span class="navbar-toggler-icon"></span>
      <span class="menu-title">Menu</span>
    </button>
    <div class="nav-logo">
      <img class="small-logo" alt="Carpentries Incubator" src="assets/images/incubator-logo-sm.svg"></div>
    <div class="lesson-title-md">
      Spatial Transcriptomics
    </div>
    <div class="search-icon-sm">
      <!-- TODO: do not show until we have search
        <i role="img" aria-label="Search the All In One page" data-feather="search"></i>
      -->
    </div>
    <div class="desktop-nav">
      <ul class="navbar-nav me-auto mb-2 mb-lg-0"><li class="nav-item">
          <span class="lesson-title">
            Spatial Transcriptomics
          </span>
        </li>
        <li class="nav-item">
          <a class="nav-link" href="key-points.html">Key Points</a>
        </li>
        <li class="nav-item">
          <a class="nav-link" href="reference.html#glossary">Glossary</a>
        </li>
        <li class="nav-item">
          <a class="nav-link" href="profiles.html">Learner Profiles</a>
        </li>
        <li class="nav-item dropdown">
          <button class="nav-link dropdown-toggle" id="navbarDropdown" data-bs-toggle="dropdown" aria-expanded="false">
            More <i data-feather="chevron-down"></i>
          </button>
          <ul class="dropdown-menu" aria-labelledby="navbarDropdown"><li><a class="dropdown-item" href="reference.html">Reference</a></li>
          </ul></li>
      </ul></div>
    <!--
    <form class="d-flex col-md-2 search-form">
      <fieldset disabled>
      <input class="form-control me-2 searchbox" type="search" placeholder="" aria-label="">
        <button class="btn btn-outline-success tablet-search-button"  type="submit">
          <i class="search-icon" data-feather="search" role="img" aria-label="Search the All In One page"></i>
        </button>
      </fieldset>
    </form>
    -->
    <a id="search-button" class="btn btn-primary" href="aio.html" role="button" aria-label="Search the All In One page">Search the All In One page</a>
  </div><!--/div.container-fluid -->
</nav><div class="col-md-12 mobile-title">
  Spatial Transcriptomics
</div>

<aside class="col-md-12 lesson-progress"><div style="width: 25%" class="percentage">
    25%
  </div>
  <div class="progress incubator">
    <div class="progress-bar incubator" role="progressbar" style="width: 25%" aria-valuenow="25" aria-label="Lesson Progress" aria-valuemin="0" aria-valuemax="100">
    </div>
  </div>
</aside><div class="container">
      <div class="row">
        <!-- START: inst/pkgdown/templates/navbar.html -->
<div id="sidebar-col" class="col-lg-4">
  <div id="sidebar" class="sidebar">
      <nav aria-labelledby="flush-headingEleven"><button role="button" aria-label="close menu" alt="close menu" aria-expanded="true" aria-controls="sidebar" class="collapse-toggle" data-collapse="Collapse " data-episodes="Episodes ">
          <i class="search-icon" data-feather="x" role="img"></i>
        </button>
        <div class="sidebar-inner">
          <div class="row mobile-row" id="theme-row-mobile">
            <div class="col" id="theme-selector">
              <li class="nav-item dropdown" id="theme-button-list">
                <button class="btn btn-link nav-link px-0 px-lg-2 dropdown-toggle d-flex align-items-center" id="bd-theme" type="button" aria-expanded="false" data-bs-toggle="dropdown" data-bs-display="static" aria-label="Toggle theme (auto)">
                  <svg class="bi my-1 theme-icon-active"><use href="#circle-half"></use></svg><span class="d-lg-none ms-1" id="bd-theme-text">Toggle Theme</span>
                </button>
                <ul class="dropdown-menu dropdown-menu-right" aria-labelledby="bd-theme-text"><li>
                    <button type="button" class="btn dropdown-item d-flex align-items-center" data-bs-theme-value="light" aria-pressed="false">
                      <svg class="bi me-2 theme-icon"><use href="#sun-fill"></use></svg>
                      Light
                      <svg class="bi ms-auto d-none"><use href="#check2"></use></svg></button>
                  </li>
                  <li>
                    <button type="button" class="btn dropdown-item d-flex align-items-center" data-bs-theme-value="dark" aria-pressed="false">
                      <svg class="bi me-2 theme-icon"><use href="#moon-stars-fill"></use></svg>
                      Dark
                      <svg class="bi ms-auto d-none"><use href="#check2"></use></svg></button>
                  </li>
                  <li>
                    <button type="button" class="btn dropdown-item d-flex align-items-center active" data-bs-theme-value="auto" aria-pressed="true">
                      <svg class="bi me-2 theme-icon"><use href="#circle-half"></use></svg>
                      Auto
                      <svg class="bi ms-auto d-none"><use href="#check2"></use></svg></button>
                  </li>
                </ul></li>
            </div>
          </div>
          <div class="row mobile-row">
            <div class="col">
              <div class="sidenav-view-selector">
                <div class="accordion accordion-flush" id="accordionFlush9">
                  <div class="accordion-item">
                    <h2 class="accordion-header" id="flush-headingNine">
                      <button class="accordion-button collapsed" id="instructor" type="button" data-bs-toggle="collapse" data-bs-target="#flush-collapseNine" aria-expanded="false" aria-controls="flush-collapseNine">
                        <i id="eye" aria-hidden="true" class="icon" data-feather="eye"></i> Learner View
                      </button>
                    </h2>
                    <div id="flush-collapseNine" class="accordion-collapse collapse" aria-labelledby="flush-headingNine" data-bs-parent="#accordionFlush2">
                      <div class="accordion-body">
                        <a href="instructor/remove-low-quality-spots.html">Instructor View</a>
                      </div>
                    </div>
                  </div><!--/div.accordion-item-->
                </div><!--/div.accordion-flush-->
              </div><!--div.sidenav-view-selector -->
            </div><!--/div.col -->

            <hr></div><!--/div.mobile-row -->

          <div class="accordion accordion-flush" id="accordionFlush11">
            <div class="accordion-item">

              <button id="chapters" class="accordion-button show" type="button" data-bs-toggle="collapse" data-bs-target="#flush-collapseEleven" aria-expanded="false" aria-controls="flush-collapseEleven">
                <h2 class="accordion-header chapters" id="flush-headingEleven">
                  EPISODES
                </h2>
              </button>
              <div id="flush-collapseEleven" class="accordion-collapse show collapse" aria-labelledby="flush-headingEleven" data-bs-parent="#accordionFlush11">

                <div class="accordion-body">
                  <div class="accordion accordion-flush" id="accordionFlush1">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading1">
        <a href="index.html">Summary and Setup</a>
    </div><!--/div.accordion-header-->

  </div><!--/div.accordion-item-->
</div><!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush2">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading2">
        <a href="introduction.html">1. Spatially Resolved Transcriptomics in Life Sciences Research</a>
    </div><!--/div.accordion-header-->

  </div><!--/div.accordion-item-->
</div><!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush3">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading3">
        <a href="data-and-study-design.html">2. Data and Study Design</a>
    </div><!--/div.accordion-header-->

  </div><!--/div.accordion-item-->
</div><!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush4">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading4">
        <a href="data-preprocessing.html">3. Data Preprocessing</a>
    </div><!--/div.accordion-header-->

  </div><!--/div.accordion-item-->
</div><!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlushcurrent">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-headingcurrent">
      <button class="accordion-button" type="button" data-bs-toggle="collapse" data-bs-target="#flush-collapsecurrent" aria-expanded="true" aria-controls="flush-collapsecurrent">
        <span class="visually-hidden">Current Chapter</span>
        <span class="current-chapter">
        4. Remove Low-quality Spots
        </span>
      </button>
    </div><!--/div.accordion-header-->

    <div id="flush-collapsecurrent" class="accordion-collapse collapse show" aria-labelledby="flush-headingcurrent" data-bs-parent="#accordionFlushcurrent">
      <div class="accordion-body">
        <ul><li><a href="#introduction">Introduction</a></li>
<li><a href="#filtering-by-mitochondrial-gene-count">Filtering by Mitochondrial Gene Count</a></li>
<li><a href="#filter-by-umi-count-and-number-of-detected-genes">Filter by UMI Count and Number of Detected Genes</a></li>
        </ul></div><!--/div.accordion-body-->
    </div><!--/div.accordion-collapse-->

  </div><!--/div.accordion-item-->
</div><!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush6">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading6">
        <a href="apply-normalization-methods.html">5. Normalization in Spatial Transcriptomics</a>
    </div><!--/div.accordion-header-->

  </div><!--/div.accordion-item-->
</div><!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush7">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading7">
        <a href="feature-selection-dimensionality-reduction-clustering.html">6. Feature Selection, Dimensionality Reduction, and Spot Clustering</a>
    </div><!--/div.accordion-header-->

  </div><!--/div.accordion-item-->
</div><!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush8">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading8">
        <a href="deconvolve-cell-types-in-a-spot.html">7. Deconvolution in Spatial Transcriptomics</a>
    </div><!--/div.accordion-header-->

  </div><!--/div.accordion-item-->
</div><!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush9">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading9">
        <a href="differential-expression-testing.html">8. Differential Expression Testing</a>
    </div><!--/div.accordion-header-->

  </div><!--/div.accordion-item-->
</div><!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush10">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading10">
        <a href="final-exercise.html">9. Putting it all Together</a>
    </div><!--/div.accordion-header-->

  </div><!--/div.accordion-item-->
</div><!--/div.accordion-flush-->

                </div>
              </div>
            </div>

            <hr class="half-width"><div class="accordion accordion-flush lesson-resources" id="accordionFlush12">
              <div class="accordion-item">
                <h2 class="accordion-header" id="flush-headingTwelve">
                  <button class="accordion-button collapsed" id="lesson-resources" type="button" data-bs-toggle="collapse" data-bs-target="#flush-collapseTwelve" aria-expanded="false" aria-controls="flush-collapseTwelve">
                    RESOURCES
                  </button>
                </h2>
                <div id="flush-collapseTwelve" class="accordion-collapse collapse" aria-labelledby="flush-headingTwelve" data-bs-parent="#accordionFlush12">
                  <div class="accordion-body">
                    <ul><li>
                        <a href="key-points.html">Key Points</a>
                      </li>
                      <li>
                        <a href="reference.html#glossary">Glossary</a>
                      </li>
                      <li>
                        <a href="profiles.html">Learner Profiles</a>
                      </li>
                      <li><a href="reference.html">Reference</a></li>
                    </ul></div>
                </div>
              </div>
            </div>
            <hr class="half-width lesson-resources"><a href="aio.html">See all in one page</a>


            <hr class="d-none d-sm-block d-md-none"><div class="d-grid gap-1">

            </div>
          </div><!-- /div.accordion -->
        </div><!-- /div.sidebar-inner -->
      </nav></div><!-- /div.sidebar -->
  </div><!-- /div.sidebar-col -->
<!-- END:   inst/pkgdown/templates/navbar.html-->

        <!-- START: inst/pkgdown/templates/content-instructor.html -->
  <div class="col-xl-8 col-lg-12 primary-content">
    <nav class="lesson-content mx-md-4" aria-label="Previous and Next Chapter"><!-- content for small screens --><div class="d-block d-sm-block d-md-none">
        <a class="chapter-link" href="data-preprocessing.html"><i aria-hidden="true" class="small-arrow" data-feather="arrow-left"></i>Previous</a>
        <a class="chapter-link float-end" href="apply-normalization-methods.html">Next<i aria-hidden="true" class="small-arrow" data-feather="arrow-right"></i></a>
      </div>
      <!-- content for large screens -->
      <div class="d-none d-sm-none d-md-block">
        <a class="chapter-link" href="data-preprocessing.html" rel="prev">
          <i aria-hidden="true" class="small-arrow" data-feather="arrow-left"></i>
          Previous: Data Preprocessing
        </a>
        <a class="chapter-link float-end" href="apply-normalization-methods.html" rel="next">
          Next: Normalization in...
          <i aria-hidden="true" class="small-arrow" data-feather="arrow-right"></i>
        </a>
      </div>
      <hr></nav><main id="main-content" class="main-content"><div class="container lesson-content">
        <h1>Remove Low-quality Spots</h1>
        <p>Last updated on 2025-05-06 |

        <a href="https://github.com/smcclatchy/spatial-transcriptomics/edit/main/episodes/remove-low-quality-spots.Rmd" class="external-link">Edit this page <i aria-hidden="true" data-feather="edit"></i></a></p>



        <div class="text-end">
          <button role="button" aria-pressed="false" tabindex="0" id="expand-code" class="pull-right" data-expand="Expand All Solutions " data-collapse="Collapse All Solutions "> Expand All Solutions <i aria-hidden="true" data-feather="plus"></i></button>
        </div>

        

<div class="overview card">
<h2 class="card-header">Overview</h2>
<div class="row g-0">
<div class="col-md-4">
<div class="card-body">
<div class="inner">
<h3 class="card-title">Questions</h3>
<ul><li>How do I remove low-quality spots?</li>
<li>What kinds of problems produce low-quality spots?</li>
<li>What happens if I skip quality control and proceed with
analysis?</li>
</ul></div>
</div>
</div>
<div class="col-md-8">
<div class="card-body">
<div class="inner bordered">
<h3 class="card-title">Objectives</h3>
<ul><li>Understand how to look for low quality spots.</li>
<li>Decide whether to retain or remove low quality spots.</li>
</ul></div>
</div>
</div>
</div>
</div>
<section><h2 class="section-heading" id="introduction">Introduction<a class="anchor" aria-label="anchor" href="#introduction"></a></h2>
<hr class="half-width"><p>Spatial transcriptomics involves a complex process that may involve
some technical failures. If the processing of the entire slide fails, it
should be obvious due to a large number of gene appearing in spots
outside of the tissue or low UMIs across the whole tissue.</p>
<p>However, there can also be variation in spot quality in a slide that
has largely high-quality spots. These artifacts are much rarer than in
single-cell transcriptomics because the process of tissue sectioning is
less disruptive than tissue dissociation. Because of this, we recommend
light spot filtering.</p>
<p>There are three metrics that we will use to identify and remove
low-quality spots:</p>
<ol style="list-style-type: decimal"><li>Mitochondrial gene expression,</li>
<li>Total UMI counts,</li>
<li>Number of detected genes.</li>
</ol><p>During tissue processing, it is possible that some cells will be
lysed, spilling out the transcripts, but retaining the mitochondria.
These spots will appear with much higher mitochondrial gene expression.
High UMI counts or number of detected genes might also indicate spots
with bleed over of lysed content from neighboring cells.</p>
<p>There may be regions of the slide with artifacts introduced during
slide preparation. These include tearing and folding of the tissue. As
mentioned above, the 10x Space Ranger pipeline automatically segments
the tissue boundary. This generally performs well at a large scale.
However, at high resolution, it may fail to properly assign spots within
small tears or along the jagged edge to the background. Such spots might
be identified by low UMI counts or number of detected genes. Conversely,
folded tissue may have a higher density of cells, which could result in
high UMI counts or number of detected genes. Pathologist annotation of
H&amp;E image can flag artifactual regions that are then excluded from
downstream analysis. Image processing techniques may also be able to
automatically identify and exclude artifactual regions, particularly
folds. Using pathologist annotation or image processing to identify
tissue abnormalities is less common than using the simple data-driven
metrics considered here, and we do not discuss them further.</p>
<p>These metrics may be tissue-dependent. In some tissues, there may be
biological reasons for differential expression across the tissue. For
example, in a cancer sample, mitochondrial or total gene expression may
vary between stromal and tumor regions. It will be important for you to
familiarize yourself with the structure of the tissue that you are
analyzing in order to make rational judgments about filtering.</p>
</section><section><h2 class="section-heading" id="filtering-by-mitochondrial-gene-count">Filtering by Mitochondrial Gene Count<a class="anchor" aria-label="anchor" href="#filtering-by-mitochondrial-gene-count"></a></h2>
<hr class="half-width"><p>In single-cell RNA sequencing experiments, the tissue is digested and
the cells are dissociated. This mechanical disruption is stressful to
the cells and some of them are damaged in the process. Elevated levels
of mitochondrial genes often indicate cell death or damage because, when
a cell’s membrane is compromised, it loses most cytoplasmic content
while retaining mitochondrial RNA. Therefore, spots with high
mitochondrial RNA may represent damaged or dying cells, and their
exclusion helps focus the analysis on healthy, intact cells.</p>
<p>More details on this relationship can be found in the <a href="https://genomebiology.biomedcentral.com/articles/10.1186/s13059-016-0888-1" class="external-link">literature
on mitochondrial DNA and cell death</a>.</p>
<p>However, in spatial transcriptomics, the tissue is either frozen or
formalin-fixed and there is much less mechanical disruption of the
tissue. Because of this, we are skeptical of the value of filtering
spots based on mitochondrial gene counts.</p>
<p>For completeness, we show how to obtain the mitochondrial genes,
calculate the percentage of counts produced by these genes in each spot,
and add this to the Seurat object metadata.</p>
<p>We will search the gene symbols in the feature metadata to identify
mitochondrial genes. We do not need to find all genes in these
categories, so we will search for genes with symbols that start with
“MT”.</p>
<p>The Seurat object is designed to be flexible and may contains several
data types. For example, it may contain both gene counts and open
chromatin peaks. In this analysis, the Seurat object only contains gene
counts. The different types of data are called “Layers” in Seurat and
may be accessed using the <a href="https://satijalab.github.io/seurat-object/reference/Layers.html" class="external-link">Layers</a>
function.</p>
<div class="codewrapper sourceCode" id="cb1">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">Layers</span><span class="op">(</span><span class="va">filter_st</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>[1] "counts"</code></pre>
</div>
<p>This tells us that the “filter_st” object only contains one data
Layer called “counts”. We can access this using the <a href="https://satijalab.github.io/seurat-object/reference/Layers.html" class="external-link">LayerData</a>
function using “counts” as an argument.</p>
<div class="codewrapper sourceCode" id="cb3">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">counts</span> <span class="op">&lt;-</span> <span class="fu">LayerData</span><span class="op">(</span><span class="va">filter_st</span>, <span class="st">'counts'</span><span class="op">)</span></span>
<span><span class="fu">head</span><span class="op">(</span><span class="va">counts</span><span class="op">)</span><span class="op">[</span>,<span class="fl">1</span><span class="op">:</span><span class="fl">5</span><span class="op">]</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>6 x 5 sparse Matrix of class "dgCMatrix"
            AAACAAGTATCTCCCA-1 AAACAATCTACTAGCA-1 AAACACCAATAACTGC-1
MIR1302-2HG                  .                  .                  .
AL627309.1                   .                  .                  .
AL669831.5                   .                  .                  .
FAM87B                       .                  .                  .
LINC00115                    .                  .                  .
FAM41C                       .                  .                  .
            AAACAGAGCGACTCCT-1 AAACAGCTTTCAGAAG-1
MIR1302-2HG                  .                  .
AL627309.1                   .                  .
AL669831.5                   .                  .
FAM87B                       .                  .
LINC00115                    .                  .
FAM41C                       .                  .</code></pre>
</div>
<p>The output above may look odd to you since there are no numbers.
Notice that the text above the table says “sparse Matrix”. Many of the
counts in the file are likely to be zero. Due to the manner in which
numbers are stored in computer memory, a zero takes up as much space as
a number. If we had to store all of these zeros, it would consume a lot
of computer memory. A sparse matrix is a special data structure which
only stores the non-zero values. In the table above, each dot (.)
represents a position with zero counts.</p>

<p>If we look at another part of the “counts” matrix, we can see
numbers.</p>
<div class="codewrapper sourceCode" id="cb5">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">counts</span><span class="op">[</span><span class="fl">20000</span><span class="op">:</span><span class="fl">20005</span>,<span class="fl">1</span><span class="op">:</span><span class="fl">5</span><span class="op">]</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>6 x 5 sparse Matrix of class "dgCMatrix"
       AAACAAGTATCTCCCA-1 AAACAATCTACTAGCA-1 AAACACCAATAACTGC-1
DNAJB1                  .                  .                  .
TECR                    3                  1                  .
NDUFB7                  3                  1                  1
ZNF333                  .                  .                  .
ADGRE2                  .                  .                  .
OR7C1                   .                  .                  .
       AAACAGAGCGACTCCT-1 AAACAGCTTTCAGAAG-1
DNAJB1                  1                  .
TECR                    2                  1
NDUFB7                  2                  4
ZNF333                  .                  .
ADGRE2                  .                  .
OR7C1                   .                  .</code></pre>
</div>
<p>As you can see in the table above, the gene symbols are stored in the
rownames of “counts”. We will find find mitochondrial genes by searching
for gene symbols which start with “MT”.</p>
<div class="codewrapper sourceCode" id="cb7">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">mito_pattern</span> <span class="op">&lt;-</span> <span class="st">'^[Mm][Tt]-'</span></span>
<span><span class="va">mito_genes</span>   <span class="op">&lt;-</span> <span class="fu">rownames</span><span class="op">(</span><span class="va">counts</span><span class="op">)</span><span class="op">[</span><span class="fu">grep</span><span class="op">(</span><span class="va">mito_pattern</span>, <span class="fu">rownames</span><span class="op">(</span><span class="va">counts</span><span class="op">)</span><span class="op">)</span><span class="op">]</span></span>
<span><span class="va">mito_genes</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code> [1] "MT-ND1"  "MT-ND2"  "MT-CO1"  "MT-CO2"  "MT-ATP8" "MT-ATP6" "MT-CO3"
 [8] "MT-ND3"  "MT-ND4L" "MT-ND4"  "MT-ND5"  "MT-ND6"  "MT-CYB" </code></pre>
</div>
<p>We now have a set of mitochondrial genes. We will use these genes to
estimate the percentage of gene counts expressed by mitochondrial genes
in each cell and add this to the Seurat object. We will pass the
mitochondrial gene symbols into <a href="https://satijalab.org/seurat/reference/percentagefeatureset" class="external-link">PercentageFeatureSet</a>,
which will perform the calculation for us.</p>
<div class="codewrapper sourceCode" id="cb9">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">filter_st</span><span class="op">[[</span><span class="st">"percent.mt"</span><span class="op">]</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="fu">PercentageFeatureSet</span><span class="op">(</span><span class="va">filter_st</span>, pattern <span class="op">=</span> <span class="va">mito_pattern</span><span class="op">)</span></span></code></pre>
</div>
<p>This syntax adds a new column called “percent.mt” to the spot
metadata.</p>
<div class="codewrapper sourceCode" id="cb10">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">colnames</span><span class="op">(</span><span class="va">filter_st</span><span class="op">@</span><span class="va">meta.data</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>[1] "orig.ident"         "nCount_Spatial"     "nFeature_Spatial"
[4] "in_tissue"          "array_row"          "array_col"
[7] "pxl_row_in_fullres" "pxl_col_in_fullres" "percent.mt"        </code></pre>
</div>

<p>Let’s look at histograms of the ribosomal and mitochondrial gene
percentages.</p>
<div class="codewrapper sourceCode" id="cb12">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">hist</span><span class="op">(</span><span class="fu">FetchData</span><span class="op">(</span><span class="va">filter_st</span>, <span class="st">"percent.mt"</span><span class="op">)</span><span class="op">[</span>,<span class="fl">1</span><span class="op">]</span>,   main <span class="op">=</span> <span class="st">"% Mitochondrial Genes"</span>,</span>
<span>     xlab <span class="op">=</span> <span class="st">"%"</span><span class="op">)</span></span></code></pre>
</div>
<figure><img src="fig/remove-low-quality-spots-rendered-unnamed-chunk-8-1.png" alt="mito" style="display: block; margin: auto;" class="figure mx-auto d-block"></figure><p>In these plots, we are looking for spots which are outside of a
normal distribution. It is difficult to generalize how to select a
filtering threshold. Some tissue or cell types may have higher
mitochondrial gene expression. Further, heterogeneous tissues may have
subsets of cells with differing levels of mitochondrial gene
expression.</p>
<p>Let’s visually check whether the mitochondrial gene expression is
normally distributed.</p>
<div class="codewrapper sourceCode" id="cb13">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">mito_expr</span> <span class="op">&lt;-</span> <span class="fu">FetchData</span><span class="op">(</span><span class="va">filter_st</span>, <span class="st">"percent.mt"</span><span class="op">)</span><span class="op">[</span>,<span class="fl">1</span><span class="op">]</span></span>
<span><span class="fu">qqnorm</span><span class="op">(</span><span class="va">mito_expr</span>, las <span class="op">=</span> <span class="fl">1</span><span class="op">)</span></span>
<span><span class="fu">qqline</span><span class="op">(</span><span class="va">mito_expr</span><span class="op">)</span></span></code></pre>
</div>
<figure><img src="fig/remove-low-quality-spots-rendered-unnamed-chunk-9-1.png" alt="mito" style="display: block; margin: auto;" class="figure mx-auto d-block"></figure><p>In this case, there may be a reason to filter out spots with greater
than 35% mitochondrial counts.</p>
</section><section><h2 class="section-heading" id="filter-by-umi-count-and-number-of-detected-genes">Filter by UMI Count and Number of Detected Genes<a class="anchor" aria-label="anchor" href="#filter-by-umi-count-and-number-of-detected-genes"></a></h2>
<hr class="half-width"><p>In the previous lesson, we plotted number of UMIs and genes detected
spatially across the tissue. Let’s plot these values again, but this
time as a histogram.</p>

<div class="codewrapper sourceCode" id="cb14">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">layout</span><span class="op">(</span><span class="fu">matrix</span><span class="op">(</span><span class="fl">1</span><span class="op">:</span><span class="fl">2</span>, ncol <span class="op">=</span> <span class="fl">1</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu">hist</span><span class="op">(</span><span class="fu">FetchData</span><span class="op">(</span><span class="va">filter_st</span>, <span class="st">"nCount_Spatial"</span><span class="op">)</span><span class="op">[</span>,<span class="fl">1</span><span class="op">]</span>,</span>
<span>     main <span class="op">=</span> <span class="st">'UMIs per Spot'</span>, xlab <span class="op">=</span> <span class="st">'Counts'</span>, las <span class="op">=</span> <span class="fl">1</span><span class="op">)</span></span>
<span><span class="fu">hist</span><span class="op">(</span><span class="fu">FetchData</span><span class="op">(</span><span class="va">filter_st</span>, <span class="st">"nFeature_Spatial"</span><span class="op">)</span><span class="op">[</span>,<span class="fl">1</span><span class="op">]</span>, </span>
<span>     main <span class="op">=</span> <span class="st">'Genes per Spot'</span>, xlab <span class="op">=</span> <span class="st">'Genes'</span>, las <span class="op">=</span> <span class="fl">1</span><span class="op">)</span></span></code></pre>
</div>
<figure><img src="fig/remove-low-quality-spots-rendered-unnamed-chunk-10-1.png" alt="mito" style="display: block; margin: auto;" class="figure mx-auto d-block"></figure><p>Again, most of the spots fall within a reasonable distribution. The
right tail of the distribution is not very thick. We might filter spots
with over 14,000 UMIs or 3,000 genes. We will use these thresholds to
add a “keep” column to the Seurat object metadata.</p>
<p>First, we will create variables for each threshold. While we could
type the numbers directly into the logical comparison statements,
creating variables makes it clear what each number represents.</p>
<div class="codewrapper sourceCode" id="cb15">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">mito_thr</span>     <span class="op">&lt;-</span> <span class="fl">32</span></span>
<span><span class="va">counts_thr</span>   <span class="op">&lt;-</span> <span class="fl">14000</span></span>
<span><span class="va">features_thr</span> <span class="op">&lt;-</span> <span class="fl">5000</span></span></code></pre>
</div>
<p>Next, we will create a “keep” variable which will be TRUE for spots
that we want to keep.</p>
<div class="codewrapper sourceCode" id="cb16">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">keep</span> <span class="op">&lt;-</span> <span class="fu">FetchData</span><span class="op">(</span><span class="va">filter_st</span>, <span class="st">"percent.mt"</span><span class="op">)</span><span class="op">[</span>,<span class="fl">1</span><span class="op">]</span>       <span class="op">&lt;</span> <span class="va">mito_thr</span></span>
<span><span class="va">keep</span> <span class="op">&lt;-</span> <span class="fu">FetchData</span><span class="op">(</span><span class="va">filter_st</span>, <span class="st">"nCount_Spatial"</span><span class="op">)</span><span class="op">[</span>,<span class="fl">1</span><span class="op">]</span>   <span class="op">&lt;</span> <span class="va">counts_thr</span>   <span class="op">&amp;</span> <span class="va">keep</span></span>
<span><span class="va">keep</span> <span class="op">&lt;-</span> <span class="fu">FetchData</span><span class="op">(</span><span class="va">filter_st</span>, <span class="st">"nFeature_Spatial"</span><span class="op">)</span><span class="op">[</span>,<span class="fl">1</span><span class="op">]</span> <span class="op">&lt;</span> <span class="va">features_thr</span> <span class="op">&amp;</span> <span class="va">keep</span></span>
<span></span>
<span><span class="va">filter_st</span><span class="op">$</span><span class="va">keep</span> <span class="op">&lt;-</span> <span class="va">keep</span></span></code></pre>
</div>
<p>Now let’s plot the spots on the tissue and color them based on
whether we will keep them.</p>
<div class="codewrapper sourceCode" id="cb17">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">SpatialDimPlot</span><span class="op">(</span><span class="va">filter_st</span>, group.by <span class="op">=</span> <span class="st">"keep"</span><span class="op">)</span></span></code></pre>
</div>
<figure><img src="fig/remove-low-quality-spots-rendered-unnamed-chunk-13-1.png" style="display: block; margin: auto;" class="figure mx-auto d-block"></figure><p>When you examine the spots that have been flagged, it is important to
look for patterns. If a contiguous section of tissue contains spots that
will be removed, it is worth looking at the histology slide to see if
there are structures that correlate with the removed spots. If it is a
section of necrotic tissue, then, depending on your experimental
question, you may want to remove those spots. But you should always look
for patterns in the removed spots and convince yourself that they are
not biasing your results.</p>
<p>Note that we will only remove a few spots in this filtering step.</p>
<div class="codewrapper sourceCode" id="cb18">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">table</span><span class="op">(</span><span class="fu">FetchData</span><span class="op">(</span><span class="va">filter_st</span>, <span class="st">"keep"</span><span class="op">)</span><span class="op">[</span>,<span class="fl">1</span><span class="op">]</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>
FALSE  TRUE
    6  3633 </code></pre>
</div>
<p>We can remove the spots directly using the following syntax. In this
case, the “columns” of the Seurat object correspond to the spots.</p>
<div class="codewrapper sourceCode" id="cb20">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">filter_st</span> <span class="op">&lt;-</span> <span class="va">filter_st</span><span class="op">[</span>,<span class="va">keep</span><span class="op">]</span></span></code></pre>
</div>
<div id="challenge-1-change-the-total-counts-spot-filtering-threshold." class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<div id="challenge-1-change-the-total-counts-spot-filtering-threshold." class="callout-inner">
<h3 class="callout-title">Challenge 1: Change the total counts spot filtering threshold.</h3>
<div class="callout-content">
<ol style="list-style-type: decimal"><li>Make a copy of the seurat object.</li>
<li>Change the threshold for the number of UMIs per spot to keep spots
with <strong>more</strong> than 2000 counts. Note that we are filtering
on the lower side of the distribution.</li>
<li>Add new variable called “keep_counts” to the Seurat object.</li>
<li>Plot the spot overlaid on the tissue section, colored by whether you
are keeping them.</li>
</ol><p>Is there a pattern to the removed spots that seems to correlate with
the tissue structure?</p>
</div>
</div>
</div>
<div id="accordionSolution1" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution1" aria-expanded="false" aria-controls="collapseSolution1">
  <h4 class="accordion-header" id="headingSolution1"> Solution 1 </h4>
</button>
<div id="collapseSolution1" class="accordion-collapse collapse" data-bs-parent="#accordionSolution1" aria-labelledby="headingSolution1">
<div class="accordion-body">
<div class="codewrapper sourceCode" id="cb21">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">obj</span>             <span class="op">&lt;-</span> <span class="va">filter_st</span></span>
<span><span class="va">new_counts_thr</span>  <span class="op">&lt;-</span> <span class="fl">2000</span></span>
<span><span class="va">keep_counts</span>     <span class="op">&lt;-</span> <span class="fu">FetchData</span><span class="op">(</span><span class="va">obj</span>, <span class="st">"nCount_Spatial"</span><span class="op">)</span><span class="op">[</span>,<span class="fl">1</span><span class="op">]</span> <span class="op">&gt;</span> <span class="va">new_counts_thr</span></span>
<span><span class="va">obj</span><span class="op">$</span><span class="va">keep_counts</span> <span class="op">&lt;-</span> <span class="va">keep_counts</span></span>
<span><span class="fu">SpatialDimPlot</span><span class="op">(</span><span class="va">obj</span>, group.by <span class="op">=</span> <span class="st">"keep_counts"</span><span class="op">)</span></span></code></pre>
</div>
<figure><img src="fig/remove-low-quality-spots-rendered-unnamed-chunk-16-1.png" style="display: block; margin: auto;" class="figure mx-auto d-block"></figure><p>Note that the spots that we have flagged seem to correspond to
stripes in the tissue section. These may be regions of the brain which
have lower levels of gene expression, so we may want to revise or remove
this threshold. Overall, this exercise shows that it is important to use
judgement when filtering spots.</p>
</div>
</div>
</div>
</div>
<div id="challenge-2-change-the-mitochondrial-spot-filtering-thresholds." class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<div id="challenge-2-change-the-mitochondrial-spot-filtering-thresholds." class="callout-inner">
<h3 class="callout-title">Challenge 2: Change the mitochondrial spot filtering thresholds.</h3>
<div class="callout-content">
<ol style="list-style-type: decimal"><li>Make a copy of the seurat object.</li>
<li>Change the threshold for the percent mitochondrial expression per
spot to keep spots with <strong>less</strong> than 25% mitochondrial
expression. This might happen if you decide that too much mitochondrial
expression indicates some technical error.</li>
<li>Add new variable called “keep_counts” to the Seurat object.</li>
<li>Plot the spot overlaid on the tissue section, colored by whether you
are keeping them.</li>
</ol><p>Is there a pattern to the removed spots that seems to correlate with
the tissue structure?</p>
</div>
</div>
</div>
<div id="accordionSolution2" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution2" aria-expanded="false" aria-controls="collapseSolution2">
  <h4 class="accordion-header" id="headingSolution2"> Solution 2 </h4>
</button>
<div id="collapseSolution2" class="accordion-collapse collapse" data-bs-parent="#accordionSolution2" aria-labelledby="headingSolution2">
<div class="accordion-body">
<div class="codewrapper sourceCode" id="cb22">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">obj</span>          <span class="op">&lt;-</span> <span class="va">filter_st</span></span>
<span><span class="va">new_mito_thr</span> <span class="op">&lt;-</span> <span class="fl">20</span></span>
<span><span class="va">keep_counts</span>  <span class="op">&lt;-</span> <span class="fu">FetchData</span><span class="op">(</span><span class="va">obj</span>, <span class="st">"percent.mt"</span><span class="op">)</span><span class="op">[</span>,<span class="fl">1</span><span class="op">]</span> <span class="op">&lt;</span> <span class="va">new_mito_thr</span></span>
<span><span class="va">obj</span><span class="op">$</span><span class="va">keep_counts</span> <span class="op">&lt;-</span> <span class="va">keep_counts</span></span>
<span><span class="fu">SpatialDimPlot</span><span class="op">(</span><span class="va">obj</span>, group.by <span class="op">=</span> <span class="st">"keep_counts"</span><span class="op">)</span></span></code></pre>
</div>
<figure><img src="fig/remove-low-quality-spots-rendered-unnamed-chunk-17-1.png" style="display: block; margin: auto;" class="figure mx-auto d-block"></figure><p>Note that the spots that we have flagged are largely outside of the
lower left. Later in the lesson, we will find that this is the “white
matter”, and this region has different expression from the rest of the
tissue section. In fact, mitochondrial expression in general seems to be
higher in the upper right area.</p>
<div class="codewrapper sourceCode" id="cb23">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">SpatialFeaturePlot</span><span class="op">(</span><span class="va">obj</span>, features <span class="op">=</span> <span class="st">"percent.mt"</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu">labs</span><span class="op">(</span>title <span class="op">=</span> <span class="st">"Percentage MitochondrialExpression"</span><span class="op">)</span></span></code></pre>
</div>
<figure><img src="fig/remove-low-quality-spots-rendered-unnamed-chunk-18-1.png" style="display: block; margin: auto;" class="figure mx-auto d-block"></figure><div class="codewrapper sourceCode" id="cb24">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">rm</span><span class="op">(</span><span class="va">obj</span><span class="op">)</span></span></code></pre>
</div>
</div>
</div>
</div>
</div>
<div id="keypoints1" class="callout keypoints">
<div class="callout-square">
<i class="callout-icon" data-feather="key"></i>
</div>
<div class="callout-inner">
<h3 class="callout-title">Key Points</h3>
<div class="callout-content">
<ul><li>Spot filtering should be light.</li>
<li>Inspect the spots that you are filtering to confirm that you are not
discarding important tissue structures.</li>
</ul></div>
</div>
</div>
<!--
Place links that you need to refer to multiple times across pages here. Delete
any links that you are not going to use.
 -->
</section></div> <!-- / div.lesson-content -->
    </main><!-- / main#main-content.main-content --><nav class="bottom-pagination mx-md-4" aria-label="Previous and Next Chapter"><div class="d-block d-sm-block d-md-none">
        <a class="chapter-link" href="data-preprocessing.html"><i aria-hidden="true" class="small-arrow" data-feather="arrow-left"></i>Previous</a>
        <a class="chapter-link float-end" href="apply-normalization-methods.html">Next<i aria-hidden="true" class="small-arrow" data-feather="arrow-right"></i></a>
      </div>
      <!-- content for large screens -->
      <div class="d-none d-sm-none d-md-block">
        <a class="chapter-link" href="data-preprocessing.html" rel="prev">
          <i aria-hidden="true" class="small-arrow" data-feather="arrow-left"></i>
          Previous: Data Preprocessing
        </a>
        <a class="chapter-link float-end" href="apply-normalization-methods.html" rel="next">
          Next: Normalization in...
          <i aria-hidden="true" class="small-arrow" data-feather="arrow-right"></i>
        </a>
      </div>
    </nav></div> <!-- / div.primary-content.col-xs-12 -->
<!-- END:   inst/pkgdown/templates/content-instructor.html-->

      </div><!--/div.row-->
      		<footer class="row footer mx-md-3"><hr><div class="col-md-6">
        <p>This lesson is subject to the <a href="CODE_OF_CONDUCT.html">Code of Conduct</a></p>
        <p>

        <a href="https://github.com/smcclatchy/spatial-transcriptomics/edit/main/episodes/remove-low-quality-spots.Rmd" class="external-link">Edit on GitHub</a>

	
        | <a href="https://github.com/smcclatchy/spatial-transcriptomics/blob/main/CONTRIBUTING.md" class="external-link">Contributing</a>
        | <a href="https://github.com/smcclatchy/spatial-transcriptomics/" class="external-link">Source</a></p>
				<p><a href="https://github.com/smcclatchy/spatial-transcriptomics/blob/main/CITATION.cff" class="external-link">Cite</a> | <a href="mailto:susan.mcclatchy@jax.org">Contact</a> | <a href="https://carpentries.org/about/" class="external-link">About</a></p>
			</div>
			<div class="col-md-6">

        <p>Materials licensed under <a href="LICENSE.html">CC-BY 4.0</a> by the authors</p>

        <p>Template licensed under <a href="https://creativecommons.org/licenses/by-sa/4.0/" class="external-link">CC-BY 4.0</a> by <a href="https://carpentries.org/" class="external-link">The Carpentries</a></p>
        <p>Built with <a href="https://github.com/epiverse-trace/sandpaper/tree/patch-renv-github-bug" class="external-link">sandpaper (0.16.11)</a>, <a href="https://github.com/carpentries/pegboard/tree/5e0d711121a581777d83cf270e298c933f112d52" class="external-link">pegboard (0.7.9)</a>, and <a href="https://github.com/carpentries/varnish/tree/1.0.5" class="external-link">varnish (1.0.5)</a></p>
			</div>
		</footer></div> <!-- / div.container -->
	<div id="to-top">
		<a href="#top">
      <i class="search-icon" data-feather="arrow-up" role="img" aria-label="Back To Top"></i><br><!-- <span class="d-none d-sm-none d-md-none d-lg-none d-xl-block">Back</span> To Top --><span class="d-none d-sm-none d-md-none d-lg-none d-xl-block">Back</span> To Top
		</a>
	</div>
  <script type="application/ld+json">
    {
  "@context": "https://schema.org",
  "@type": "TrainingMaterial",
  "@id": "https://smcclatchy.github.io/spatial-transcriptomics/remove-low-quality-spots.html",
  "inLanguage": "en",
  "dct:conformsTo": "https://bioschemas.org/profiles/TrainingMaterial/1.0-RELEASE",
  "description": "A Carpentries Lesson teaching foundational data and coding skills to researchers worldwide",
  "keywords": "software, data, lesson, The Carpentries, transcriptome, Visium, cancer",
  "name": "Remove Low-quality Spots",
  "creativeWorkStatus": "active",
  "url": "https://smcclatchy.github.io/spatial-transcriptomics/remove-low-quality-spots.html",
  "identifier": "https://smcclatchy.github.io/spatial-transcriptomics/remove-low-quality-spots.html",
  "dateCreated": "2024-03-01",
  "dateModified": "2025-05-06",
  "datePublished": "2025-05-06"
}

  </script><script>
		feather.replace();
	</script></body></html><!-- END:   inst/pkgdown/templates/layout.html-->

