<!DOCTYPE html>
<!-- START: inst/pkgdown/templates/layout.html --><!-- Generated by pkgdown: do not edit by hand --><html lang="en" data-bs-theme="auto"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><title>Spatial Transcriptomics: Feature Selection, Dimensionality Reduction, and Spot Clustering</title><meta name="viewport" content="width=device-width, initial-scale=1"><script src="../assets/themetoggle.js"></script><link rel="stylesheet" type="text/css" href="../assets/styles.css"><script src="../assets/scripts.js" type="text/javascript"></script><!-- mathjax --><script type="text/x-mathjax-config">
    MathJax.Hub.Config({
      config: ["MMLorHTML.js"],
      jax: ["input/TeX","input/MathML","output/HTML-CSS","output/NativeMML", "output/PreviewHTML"],
      extensions: ["tex2jax.js","mml2jax.js","MathMenu.js","MathZoom.js", "fast-preview.js", "AssistiveMML.js", "a11y/accessibility-menu.js"],
      TeX: {
        extensions: ["AMSmath.js","AMSsymbols.js","noErrors.js","noUndefined.js"]
      },
      tex2jax: {
        inlineMath: [['\\(', '\\)']],
        displayMath: [ ['$$','$$'], ['\\[', '\\]'] ],
        processEscapes: true
      }
    });
    </script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><!-- Responsive Favicon for The Carpentries --><link rel="apple-touch-icon" sizes="180x180" href="../apple-touch-icon.png"><link rel="icon" type="image/png" sizes="32x32" href="../favicon-32x32.png"><link rel="icon" type="image/png" sizes="16x16" href="../favicon-16x16.png"><link rel="manifest" href="../site.webmanifest"><link rel="mask-icon" href="../safari-pinned-tab.svg" color="#5bbad5"><meta name="msapplication-TileColor" content="#da532c"><meta name="theme-color" media="(prefers-color-scheme: light)" content="white"><meta name="theme-color" media="(prefers-color-scheme: dark)" content="black"></head><body>
    <header id="top" class="navbar navbar-expand-md top-nav incubator"><svg xmlns="http://www.w3.org/2000/svg" class="d-none"><symbol id="check2" viewbox="0 0 16 16"><path d="M13.854 3.646a.5.5 0 0 1 0 .708l-7 7a.5.5 0 0 1-.708 0l-3.5-3.5a.5.5 0 1 1 .708-.708L6.5 10.293l6.646-6.647a.5.5 0 0 1 .708 0z"></path></symbol><symbol id="circle-half" viewbox="0 0 16 16"><path d="M8 15A7 7 0 1 0 8 1v14zm0 1A8 8 0 1 1 8 0a8 8 0 0 1 0 16z"></path></symbol><symbol id="moon-stars-fill" viewbox="0 0 16 16"><path d="M6 .278a.768.768 0 0 1 .08.858 7.208 7.208 0 0 0-.878 3.46c0 4.021 3.278 7.277 7.318 7.277.527 0 1.04-.055 1.533-.16a.787.787 0 0 1 .81.316.733.733 0 0 1-.031.893A8.349 8.349 0 0 1 8.344 16C3.734 16 0 12.286 0 7.71 0 4.266 2.114 1.312 5.124.06A.752.752 0 0 1 6 .278z"></path><path d="M10.794 3.148a.217.217 0 0 1 .412 0l.387 1.162c.173.518.579.924 1.097 1.097l1.162.387a.217.217 0 0 1 0 .412l-1.162.387a1.734 1.734 0 0 0-1.097 1.097l-.387 1.162a.217.217 0 0 1-.412 0l-.387-1.162A1.734 1.734 0 0 0 9.31 6.593l-1.162-.387a.217.217 0 0 1 0-.412l1.162-.387a1.734 1.734 0 0 0 1.097-1.097l.387-1.162zM13.863.099a.145.145 0 0 1 .274 0l.258.774c.115.346.386.617.732.732l.774.258a.145.145 0 0 1 0 .274l-.774.258a1.156 1.156 0 0 0-.732.732l-.258.774a.145.145 0 0 1-.274 0l-.258-.774a1.156 1.156 0 0 0-.732-.732l-.774-.258a.145.145 0 0 1 0-.274l.774-.258c.346-.115.617-.386.732-.732L13.863.1z"></path></symbol><symbol id="sun-fill" viewbox="0 0 16 16"><path d="M8 12a4 4 0 1 0 0-8 4 4 0 0 0 0 8zM8 0a.5.5 0 0 1 .5.5v2a.5.5 0 0 1-1 0v-2A.5.5 0 0 1 8 0zm0 13a.5.5 0 0 1 .5.5v2a.5.5 0 0 1-1 0v-2A.5.5 0 0 1 8 13zm8-5a.5.5 0 0 1-.5.5h-2a.5.5 0 0 1 0-1h2a.5.5 0 0 1 .5.5zM3 8a.5.5 0 0 1-.5.5h-2a.5.5 0 0 1 0-1h2A.5.5 0 0 1 3 8zm10.657-5.657a.5.5 0 0 1 0 .707l-1.414 1.415a.5.5 0 1 1-.707-.708l1.414-1.414a.5.5 0 0 1 .707 0zm-9.193 9.193a.5.5 0 0 1 0 .707L3.05 13.657a.5.5 0 0 1-.707-.707l1.414-1.414a.5.5 0 0 1 .707 0zm9.193 2.121a.5.5 0 0 1-.707 0l-1.414-1.414a.5.5 0 0 1 .707-.707l1.414 1.414a.5.5 0 0 1 0 .707zM4.464 4.465a.5.5 0 0 1-.707 0L2.343 3.05a.5.5 0 1 1 .707-.707l1.414 1.414a.5.5 0 0 1 0 .708z"></path></symbol></svg><a class="visually-hidden-focusable skip-link" href="#main-content">Skip to main content</a>
  <div class="container-fluid top-nav-container">
    <div class="col-md-8">
      <div class="large-logo">
        <img id="incubator-logo" alt="Carpentries Incubator" src="../assets/images/incubator-logo.svg"><span class="badge text-bg-warning">
          <abbr title="This lesson is in the alpha phase, which means that it has been taught once and lesson authors are iterating on feedback.">
            <a href="https://cdh.carpentries.org/the-lesson-life-cycle.html#field-testing-alpha-stage" class="external-link alert-link">
              <i aria-hidden="true" class="icon" data-feather="alert-triangle" style="border-radius: 5px"></i>
              Alpha
            </a>
            <span class="visually-hidden">This lesson is in the alpha phase, which means that it has been taught once and lesson authors are iterating on feedback.</span>
          </abbr>
        </span>

      </div>
    </div>
    <div class="selector-container">
      <div id="theme-selector">
        <li class="nav-item dropdown" id="theme-button-list">
          <button class="btn btn-link nav-link px-0 px-lg-2 dropdown-toggle d-flex align-items-center" id="bd-theme" type="button" aria-expanded="false" data-bs-toggle="dropdown" data-bs-display="static" aria-label="Toggle theme (auto)">
            <svg class="bi my-1 theme-icon-active"><use href="#circle-half"></use></svg><i data-feather="chevron-down"></i>
          </button>
          <ul class="dropdown-menu dropdown-menu-end" aria-labelledby="bd-theme-text"><li>
              <button type="button" class="btn dropdown-item d-flex align-items-center" data-bs-theme-value="light" aria-pressed="false">
                <svg class="bi me-2 theme-icon"><use href="#sun-fill"></use></svg>
                Light
                <svg class="bi ms-auto d-none"><use href="#check2"></use></svg></button>
            </li>
            <li>
              <button type="button" class="btn dropdown-item d-flex align-items-center" data-bs-theme-value="dark" aria-pressed="false">
                <svg class="bi me-2 theme-icon"><use href="#moon-stars-fill"></use></svg>
                Dark
                <svg class="bi ms-auto d-none"><use href="#check2"></use></svg></button>
            </li>
            <li>
              <button type="button" class="btn dropdown-item d-flex align-items-center active" data-bs-theme-value="auto" aria-pressed="true">
                <svg class="bi me-2 theme-icon"><use href="#circle-half"></use></svg>
                Auto
                <svg class="bi ms-auto d-none"><use href="#check2"></use></svg></button>
            </li>
          </ul></li>
      </div>

      <div class="dropdown" id="instructor-dropdown">
        <button class="btn btn-secondary dropdown-toggle bordered-button" type="button" id="dropdownMenu1" data-bs-toggle="dropdown" aria-expanded="false">
          <i aria-hidden="true" class="icon" data-feather="eye"></i> Instructor View <i data-feather="chevron-down"></i>
        </button>
        <ul class="dropdown-menu" aria-labelledby="dropdownMenu1"><li><button class="dropdown-item" type="button" onclick="window.location.href='../feature-selection-dimensionality-reduction-clustering.html';">Learner View</button></li>
        </ul></div>
    </div>
  </div>
  <hr></header><nav class="navbar navbar-expand-xl bottom-nav incubator" aria-label="Main Navigation"><div class="container-fluid nav-container">
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbarSupportedContent" aria-controls="navbarSupportedContent" aria-expanded="false" aria-label="Toggle Navigation">
      <span class="navbar-toggler-icon"></span>
      <span class="menu-title">Menu</span>
    </button>
    <div class="nav-logo">
      <img class="small-logo" alt="Carpentries Incubator" src="../assets/images/incubator-logo-sm.svg"></div>
    <div class="lesson-title-md">
      Spatial Transcriptomics
    </div>
    <div class="search-icon-sm">
      <!-- TODO: do not show until we have search
        <i role="img" aria-label="Search the All In One page" data-feather="search"></i>
      -->
    </div>
    <div class="desktop-nav">
      <ul class="navbar-nav me-auto mb-2 mb-lg-0"><li class="nav-item">
          <span class="lesson-title">
            Spatial Transcriptomics
          </span>
        </li>
        <li class="nav-item">
          <a class="nav-link" href="../instructor/key-points.html">Key Points</a>
        </li>
        <li class="nav-item">
          <a class="nav-link" href="../instructor/instructor-notes.html">Instructor Notes</a>
        </li>
        <li class="nav-item">
          <a class="nav-link" href="../instructor/images.html">Extract All Images</a>
        </li>
        <li class="nav-item dropdown">
          <button class="nav-link dropdown-toggle" id="navbarDropdown" data-bs-toggle="dropdown" aria-expanded="false">
            More <i data-feather="chevron-down"></i>
          </button>
          <ul class="dropdown-menu" aria-labelledby="navbarDropdown"><hr><li><a class="dropdown-item" href="reference.html">Reference</a></li>
          </ul></li>
      </ul></div>
    <!--
    <form class="d-flex col-md-2 search-form">
      <fieldset disabled>
      <input class="form-control me-2 searchbox" type="search" placeholder="" aria-label="">
        <button class="btn btn-outline-success tablet-search-button"  type="submit">
          <i class="search-icon" data-feather="search" role="img" aria-label="Search the All In One page"></i>
        </button>
      </fieldset>
    </form>
    -->
    <a id="search-button" class="btn btn-primary" href="../instructor/aio.html" role="button" aria-label="Search the All In One page">Search the All In One page</a>
  </div><!--/div.container-fluid -->
</nav><div class="col-md-12 mobile-title">
  Spatial Transcriptomics
</div>

<aside class="col-md-12 lesson-progress"><div style="width: 51%" class="percentage">
    51%
  </div>
  <div class="progress incubator">
    <div class="progress-bar incubator" role="progressbar" style="width: 51%" aria-valuenow="51" aria-label="Lesson Progress" aria-valuemin="0" aria-valuemax="100">
    </div>
  </div>
</aside><div class="container">
      <div class="row">
        <!-- START: inst/pkgdown/templates/navbar.html -->
<div id="sidebar-col" class="col-lg-4">
  <div id="sidebar" class="sidebar">
      <nav aria-labelledby="flush-headingEleven"><button role="button" aria-label="close menu" alt="close menu" aria-expanded="true" aria-controls="sidebar" class="collapse-toggle" data-collapse="Collapse " data-episodes="Episodes ">
          <i class="search-icon" data-feather="x" role="img"></i>
        </button>
        <div class="sidebar-inner">
          <div class="row mobile-row" id="theme-row-mobile">
            <div class="col" id="theme-selector">
              <li class="nav-item dropdown" id="theme-button-list">
                <button class="btn btn-link nav-link px-0 px-lg-2 dropdown-toggle d-flex align-items-center" id="bd-theme" type="button" aria-expanded="false" data-bs-toggle="dropdown" data-bs-display="static" aria-label="Toggle theme (auto)">
                  <svg class="bi my-1 theme-icon-active"><use href="#circle-half"></use></svg><span class="d-lg-none ms-1" id="bd-theme-text">Toggle Theme</span>
                </button>
                <ul class="dropdown-menu dropdown-menu-right" aria-labelledby="bd-theme-text"><li>
                    <button type="button" class="btn dropdown-item d-flex align-items-center" data-bs-theme-value="light" aria-pressed="false">
                      <svg class="bi me-2 theme-icon"><use href="#sun-fill"></use></svg>
                      Light
                      <svg class="bi ms-auto d-none"><use href="#check2"></use></svg></button>
                  </li>
                  <li>
                    <button type="button" class="btn dropdown-item d-flex align-items-center" data-bs-theme-value="dark" aria-pressed="false">
                      <svg class="bi me-2 theme-icon"><use href="#moon-stars-fill"></use></svg>
                      Dark
                      <svg class="bi ms-auto d-none"><use href="#check2"></use></svg></button>
                  </li>
                  <li>
                    <button type="button" class="btn dropdown-item d-flex align-items-center active" data-bs-theme-value="auto" aria-pressed="true">
                      <svg class="bi me-2 theme-icon"><use href="#circle-half"></use></svg>
                      Auto
                      <svg class="bi ms-auto d-none"><use href="#check2"></use></svg></button>
                  </li>
                </ul></li>
            </div>
          </div>
          <div class="row mobile-row">
            <div class="col">
              <div class="sidenav-view-selector">
                <div class="accordion accordion-flush" id="accordionFlush9">
                  <div class="accordion-item">
                    <h2 class="accordion-header" id="flush-headingNine">
                      <button class="accordion-button collapsed" id="instructor" type="button" data-bs-toggle="collapse" data-bs-target="#flush-collapseNine" aria-expanded="false" aria-controls="flush-collapseNine">
                        <i id="eye" aria-hidden="true" class="icon" data-feather="eye"></i> Instructor View
                      </button>
                    </h2>
                    <div id="flush-collapseNine" class="accordion-collapse collapse" aria-labelledby="flush-headingNine" data-bs-parent="#accordionFlush2">
                      <div class="accordion-body">
                        <a href="../feature-selection-dimensionality-reduction-clustering.html">Learner View</a>
                      </div>
                    </div>
                  </div><!--/div.accordion-item-->
                </div><!--/div.accordion-flush-->
              </div><!--div.sidenav-view-selector -->
            </div><!--/div.col -->

            <hr></div><!--/div.mobile-row -->

          <div class="accordion accordion-flush" id="accordionFlush11">
            <div class="accordion-item">

              <button id="chapters" class="accordion-button show" type="button" data-bs-toggle="collapse" data-bs-target="#flush-collapseEleven" aria-expanded="false" aria-controls="flush-collapseEleven">
                <h2 class="accordion-header chapters" id="flush-headingEleven">
                  EPISODES
                </h2>
              </button>
              <div id="flush-collapseEleven" class="accordion-collapse show collapse" aria-labelledby="flush-headingEleven" data-bs-parent="#accordionFlush11">

                <div class="accordion-body">
                  <div class="accordion accordion-flush" id="accordionFlush1">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading1">
        <a href="index.html">Summary and Schedule</a>
    </div><!--/div.accordion-header-->

  </div><!--/div.accordion-item-->
</div><!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush2">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading2">
        <a href="introduction.html">1. Spatially Resolved Transcriptomics in Life Sciences Research</a>
    </div><!--/div.accordion-header-->

  </div><!--/div.accordion-item-->
</div><!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush3">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading3">
        <a href="data-and-study-design.html">2. Data and Study Design</a>
    </div><!--/div.accordion-header-->

  </div><!--/div.accordion-item-->
</div><!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush4">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading4">
        <a href="data-preprocessing.html">3. Data Preprocessing</a>
    </div><!--/div.accordion-header-->

  </div><!--/div.accordion-item-->
</div><!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush5">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading5">
        <a href="remove-low-quality-spots.html">4. Remove Low-quality Spots</a>
    </div><!--/div.accordion-header-->

  </div><!--/div.accordion-item-->
</div><!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush6">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading6">
        <a href="apply-normalization-methods.html">5. Normalization in Spatial Transcriptomics</a>
    </div><!--/div.accordion-header-->

  </div><!--/div.accordion-item-->
</div><!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlushcurrent">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-headingcurrent">
      <button class="accordion-button" type="button" data-bs-toggle="collapse" data-bs-target="#flush-collapsecurrent" aria-expanded="true" aria-controls="flush-collapsecurrent">
        <span class="visually-hidden">Current Chapter</span>
        <span class="current-chapter">
        6. Feature Selection, Dimensionality Reduction, and Spot Clustering
        </span>
      </button>
    </div><!--/div.accordion-header-->

    <div id="flush-collapsecurrent" class="accordion-collapse collapse show" aria-labelledby="flush-headingcurrent" data-bs-parent="#accordionFlushcurrent">
      <div class="accordion-body">
        <ul><li><a href="#understanding-the-morphology-of-your-tissue">Understanding the Morphology of your Tissue</a></li>
<li><a href="#understanding-feature-selection-in-spatial-transcriptomics">Understanding Feature Selection in Spatial Transcriptomics</a></li>
        </ul></div><!--/div.accordion-body-->
    </div><!--/div.accordion-collapse-->

  </div><!--/div.accordion-item-->
</div><!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush8">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading8">
        <a href="deconvolve-cell-types-in-a-spot.html">7. Deconvolution in Spatial Transcriptomics</a>
    </div><!--/div.accordion-header-->

  </div><!--/div.accordion-item-->
</div><!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush9">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading9">
        <a href="differential-expression-testing.html">8. Differential Expression Testing</a>
    </div><!--/div.accordion-header-->

  </div><!--/div.accordion-item-->
</div><!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush10">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading10">
        <a href="final-exercise.html">9. Putting it all Together</a>
    </div><!--/div.accordion-header-->

  </div><!--/div.accordion-item-->
</div><!--/div.accordion-flush-->

                </div>
              </div>
            </div>

            <hr class="half-width"><div class="accordion accordion-flush resources" id="accordionFlush12">
              <div class="accordion-item">
                <h2 class="accordion-header" id="flush-headingTwelve">
                  <button class="accordion-button collapsed" id="resources" type="button" data-bs-toggle="collapse" data-bs-target="#flush-collapseTwelve" aria-expanded="false" aria-controls="flush-collapseTwelve">
                    RESOURCES
                  </button>
                </h2>
                <div id="flush-collapseTwelve" class="accordion-collapse collapse" aria-labelledby="flush-headingTwelve" data-bs-parent="#accordionFlush12">
                  <div class="accordion-body">
                    <ul><li>
                        <a href="../instructor/key-points.html">Key Points</a>
                      </li>
                      <li>
                        <a href="../instructor/instructor-notes.html">Instructor Notes</a>
                      </li>
                      <li>
                        <a href="../instructor/images.html">Extract All Images</a>
                      </li>
                      <hr><li><a class="dropdown-item" href="reference.html">Reference</a></li>
                    </ul></div>
                </div>
              </div>
            </div>
            <hr class="half-width resources"><a href="../instructor/aio.html">See all in one page</a>


            <hr class="d-none d-sm-block d-md-none"><div class="d-grid gap-1">

            </div>
          </div><!-- /div.accordion -->
        </div><!-- /div.sidebar-inner -->
      </nav></div><!-- /div.sidebar -->
  </div><!-- /div.sidebar-col -->
<!-- END:   inst/pkgdown/templates/navbar.html-->

        <!-- START: inst/pkgdown/templates/content-instructor.html -->
  <div class="col-xl-8 col-lg-12 primary-content">
    <nav class="lesson-content mx-md-4" aria-label="Previous and Next Chapter"><!-- content for small screens --><div class="d-block d-sm-block d-md-none">
        <a class="chapter-link" href="../instructor/apply-normalization-methods.html"><i aria-hidden="true" class="small-arrow" data-feather="arrow-left"></i>Previous</a>
        <a class="chapter-link float-end" href="../instructor/deconvolve-cell-types-in-a-spot.html">Next<i aria-hidden="true" class="small-arrow" data-feather="arrow-right"></i></a>
      </div>
      <!-- content for large screens -->
      <div class="d-none d-sm-none d-md-block">
        <a class="chapter-link" href="../instructor/apply-normalization-methods.html" rel="prev">
          <i aria-hidden="true" class="small-arrow" data-feather="arrow-left"></i>
          Previous: Normalization in
        </a>
        <a class="chapter-link float-end" href="../instructor/deconvolve-cell-types-in-a-spot.html" rel="next">
          Next: Deconvolution in...
          <i aria-hidden="true" class="small-arrow" data-feather="arrow-right"></i>
        </a>
      </div>
      <hr></nav><main id="main-content" class="main-content"><div class="container lesson-content">
        <h1>Feature Selection, Dimensionality Reduction, and Spot Clustering</h1>
        <p>Last updated on 2024-08-09 |

        <a href="https://github.com/smcclatchy/spatial-transcriptomics/edit/main/episodes/feature-selection-dimensionality-reduction-clustering.Rmd" class="external-link">Edit this page <i aria-hidden="true" data-feather="edit"></i></a></p>



        <p>Estimated time: <i aria-hidden="true" data-feather="clock"></i> 85 minutes</p>

        <div class="text-end">
          <button role="button" aria-pressed="false" tabindex="0" id="expand-code" class="pull-right" data-expand="Expand All Solutions " data-collapse="Collapse All Solutions "> Expand All Solutions <i aria-hidden="true" data-feather="plus"></i></button>
        </div>

        

<div class="overview card">
<h2 class="card-header">Overview</h2>
<div class="row g-0">
<div class="col-md-4">
<div class="card-body">
<div class="inner">
<h3 class="card-title">Questions</h3>
<ul><li>Why is feature selection important in spatial transcriptomics?</li>
<li>What are the implications of using different proportions of highly
variable genes (HVGs) in data analysis?</li>
<li>Why is feature selection in spatial transcriptomics not typically
necessary with normalization techniques like SCTransform?</li>
<li>How do PCA and UMAP differ in their approach to dimensionality
reduction in spatial transcriptomics?</li>
<li>What advantages do linear methods like PCA offer before applying
nonlinear methods like UMAP?</li>
<li>How do these dimensionality reduction techniques impact downstream
analysis such as clustering and visualization?</li>
</ul></div>
</div>
</div>
<div class="col-md-8">
<div class="card-body">
<div class="inner bordered">
<h3 class="card-title">Objectives</h3>
<ul><li>Identify appropriate feature selection methods for different
normalization techniques in spatial transcriptomics.</li>
<li>Evaluate the effects of varying the proportion of highly variable
genes on the resolution of clustering and PCA outcomes.</li>
<li>Understand the rationale behind the dependency of feature selection
on specific normalization methods like NormalizeData.</li>
<li>Differentiate between linear and nonlinear dimensionality reduction
methods and their applications in spatial transcriptomics.</li>
<li>Implement PCA to preprocess data before applying UMAP to enhance
interpretability and structure recognition.</li>
<li>Assess the effectiveness of each method in revealing spatial and
molecular patterns within the data.</li>
</ul></div>
</div>
</div>
</div>
</div>
<section><h2 class="section-heading" id="understanding-the-morphology-of-your-tissue">Understanding the Morphology of your Tissue<a class="anchor" aria-label="anchor" href="#understanding-the-morphology-of-your-tissue"></a>
<a class="anchor" aria-label="anchor" href="#understanding-the-morphology-of-your-tissue"></a></h2>
<hr class="half-width"><p>In any analysis, it is important that you understand the structure
and cell types in the tissue which you are analyzing. While it would be
ideal to have an unbiased analysis that uses normalizations and
clustering methods to automatically assign cell types and define tissue
structure, in practice we adjust these parameters based on the tissue
structure that we expect to find.</p>
</section><section><h2 class="section-heading" id="understanding-feature-selection-in-spatial-transcriptomics">Understanding Feature Selection in Spatial Transcriptomics<a class="anchor" aria-label="anchor" href="#understanding-feature-selection-in-spatial-transcriptomics"></a>
<a class="anchor" aria-label="anchor" href="#understanding-feature-selection-in-spatial-transcriptomics"></a></h2>
<hr class="half-width"><p>Feature selection in spatial transcriptomics is essential for
reducing the dimensionality of high-dimensional datasets, enhancing
model performance, and improving interpretability. This process is
crucial because it helps in minimizing computational demands, reducing
noise, and speeding up downstream analyses like clustering and principal
components analysis (PCA). By focusing on a subset of genes that show
significant variability or are biologically relevant, researchers can
achieve more robust and generalizable models, draw clearer conclusions,
and facilitate hypothesis testing.</p>
<div class="section level3">
<h3 id="choosing-feature-selection-methods">Choosing Feature Selection Methods<a class="anchor" aria-label="anchor" href="#choosing-feature-selection-methods"></a></h3>
<div class="section level4">
<h4 id="importance-of-high-variable-gene-selection">Importance of High Variable Gene Selection<a class="anchor" aria-label="anchor" href="#importance-of-high-variable-gene-selection"></a></h4>
<p>Feature selection methods such as variance stabilizing transformation
(VST) and mean-variance plotting are crucial for refining the dataset to
include genes that exhibit meaningful variability across different
spatial regions. These methods help focus on genes that are most
informative for downstream analyses like clustering and dimensionality
reduction.</p>
</div>
</div>
<div class="section level3">
<h3 id="feature-selection-with-normalizedata">Feature Selection with <code>NormalizeData</code>
<a class="anchor" aria-label="anchor" href="#feature-selection-with-normalizedata"></a></h3>
<p>When using normalization methods like <code>NormalizeData</code>,
which focuses on scaling gene expression data without variance
stabilization, applying feature selection becomes essential. This method
requires the selection of highly variable genes to enhance the analysis,
particularly in clustering and PCA. Typically, you will select in the
range of 2,000 to 5,000 highly variable genes.</p>
</div>
<div class="section level3">
<h3 id="feature-selection-with-sctransform">Feature Selection with <code>SCTransform</code>
<a class="anchor" aria-label="anchor" href="#feature-selection-with-sctransform"></a></h3>
<p><code>SCTransform</code>, a normalization method, adjusts gene
expression data to stabilize the variance, and it also provides default
feature selection. This method ensures that the genes retained are
already adjusted for technical variability, highlighting those with
biological significance.</p>
<p>The <code>SCTransform</code> selects 3,000 variable features by
default. We will use those variable features to calculate <em>principal
components</em> (PCs) of the gene expression data. As always, we must
first scale (or standardize) the normalized counts data.</p>
<div id="challenge-1-number-of-variable-genes" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<div id="challenge-1-number-of-variable-genes" class="callout-inner">
<h3 class="callout-title">Challenge 1: Number of Variable Genes<a class="anchor" aria-label="anchor" href="#challenge-1-number-of-variable-genes"></a>
</h3>
<div class="callout-content">
<p>How do you think that your results would be affected by selecting too
few highly variable genes? What about too many highly variable
genes?</p>
</div>
</div>
</div>
<div id="accordionSolution1" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution1" aria-expanded="false" aria-controls="collapseSolution1">
  <h4 class="accordion-header" id="headingSolution1"> Show me the solution </h4>
</button>
<div id="collapseSolution1" class="accordion-collapse collapse" aria-labelledby="headingSolution1" data-bs-parent="#accordionSolution1">
<div class="accordion-body">
<p>Too few variable genes may underestimate the variance between spots
and tissue sections and would reduce our ability to discern tissue
structure. Too many variable genes would increase computational time and
might add noise to the analysis.</p>
</div>
</div>
</div>
</div>
</div>
<div class="section level3">
<h3 id="dimensionality-reduction-using-principal-components">Dimensionality Reduction using Principal Components<a class="anchor" aria-label="anchor" href="#dimensionality-reduction-using-principal-components"></a></h3>
<p>Dimensionality reduction is a crucial step in managing
high-dimensional spatial transcriptomics data, enhancing analytical
clarity, and reducing computational load. Linear methods like PCA and
nonlinear methods like UMAP each play distinct roles in processing and
interpreting complex datasets.</p>
<p>PCA is a linear technique that reduces dimensionality by transforming
data into a set of uncorrelated variables called <em>principal
components</em> (PCs). This method efficiently captures the main
variance in the data, which is vital for preliminary data exploration
and noise reduction.</p>
<p>In order to cluster the spots by similarity, we use PCs to reduce the
number of dimensions in the data. Using a smaller number of PCs allows
us to capture the variability in the dataset while using a smaller
number of dimensions. The PCs are then used to cluster the spots by
similarity.</p>
<div class="codewrapper sourceCode" id="cb1">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">filter_st</span> <span class="op">&lt;-</span> <span class="fu">ScaleData</span><span class="op">(</span><span class="va">filter_st</span><span class="op">)</span> <span class="op">%&gt;%</span></span>
<span>               <span class="fu">RunPCA</span><span class="op">(</span>npcs <span class="op">=</span> <span class="fl">75</span>, verbose <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>Centering and scaling data matrix</code></pre>
</div>
<p>In the command above, we specifically asked Seurat to calculate 75
principal components. There are thousands of genes and we know that we
don’t need to calculate all PCs. By default, the <a href="https://satijalab.org/seurat/reference/runpca" class="external-link"><code>RunPCA</code></a>
method calculates 50 PCs. Let’s make an elbow plot of the number of PCs
versus the standard deviation explained by each PC. Traditionally, there
is a bend in the curve which indicates that adding more PCs doesn’t
account for more of the variance.</p>
<div class="codewrapper sourceCode" id="cb3">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">ElbowPlot</span><span class="op">(</span><span class="va">filter_st</span>, ndims <span class="op">=</span> <span class="fl">75</span><span class="op">)</span></span></code></pre>
</div>
<figure><img src="../fig/feature-selection-dimensionality-reduction-clustering-rendered-unnamed-chunk-3-1.png" style="display: block; margin: auto;" class="figure mx-auto d-block"></figure><p>From the plot above, we would normally select somewhere between 10
and 20 PCs because there seems to be little benefit in adding more PCs.
Adding more PCs does not seem to add more explanatory variance.</p>
<p>But in spatial transcriptomics, the elbow plot often does not tell
the whole story. We want to select a number of PCs such that we are able
to discern the structure and cell type composition of our tissue. So it
is better to try a range of number of PCs.</p>
<p>For now, let’s use all 75 PCs.</p>
<div class="codewrapper sourceCode" id="cb4">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">n_pcs</span> <span class="op">=</span> <span class="fl">75</span></span></code></pre>
</div>
<p>In the next step, we will cluster the spots based on expression
similarity using the principal components that we just generated. We
will use Seurat’s <a href="https://satijalab.org/seurat/reference/findneighbors" class="external-link"><code>FindNeighbors</code></a>
and <a href="https://satijalab.org/seurat/reference/findclusters" class="external-link"><code>FindClusters</code></a>
functions. <code>FindNeighbors</code> find the K nearest neighbors of
each spot in the dataset. The default values is to use the 20 nearest
neighbors in the dataset and we will use this value in this lesson.
However, as with the number of features used to create the PCs, this is
another parameter that is worth varying before proceeding with your
analysis.</p>
<div class="codewrapper sourceCode" id="cb5">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">filter_st</span> <span class="op">&lt;-</span> <span class="fu">FindNeighbors</span><span class="op">(</span><span class="va">filter_st</span>, </span>
<span>                           reduction <span class="op">=</span> <span class="st">"pca"</span>, </span>
<span>                           dims      <span class="op">=</span> <span class="fl">1</span><span class="op">:</span><span class="va">n_pcs</span><span class="op">)</span> <span class="op">%&gt;%</span> </span>
<span>               <span class="fu">FindClusters</span><span class="op">(</span>resolution <span class="op">=</span> <span class="fl">1</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>Computing nearest neighbor graph</code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>Computing SNN</code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>Modularity Optimizer version 1.3.0 by Ludo Waltman and Nees Jan van Eck

Number of nodes: 3633
Number of edges: 187716

Running Louvain algorithm...
Maximum modularity in 10 random starts: 0.7178
Number of communities: 9
Elapsed time: 0 seconds</code></pre>
</div>
<p>The clustering has added a new column to the spot metadata in the
Seurat object called <code>seurat_clusters</code>. Let’s look the
metadata to see this.</p>
<div class="codewrapper sourceCode" id="cb9">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">head</span><span class="op">(</span><span class="va">filter_st</span><span class="op">[[</span><span class="op">]</span><span class="op">]</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>                      orig.ident nCount_Spatial nFeature_Spatial in_tissue
AAACAAGTATCTCCCA-1 SeuratProject           8458             3586         1
AAACAATCTACTAGCA-1 SeuratProject           1667             1150         1
AAACACCAATAACTGC-1 SeuratProject           3769             1960         1
AAACAGAGCGACTCCT-1 SeuratProject           5433             2424         1
AAACAGCTTTCAGAAG-1 SeuratProject           4278             2264         1
AAACAGGGTCTATATT-1 SeuratProject           4004             2178         1
                   array_row array_col pxl_row_in_fullres pxl_col_in_fullres
AAACAAGTATCTCCCA-1        50       102               8468               9791
AAACAATCTACTAGCA-1         3        43               2807               5769
AAACACCAATAACTGC-1        59        19               9505               4068
AAACAGAGCGACTCCT-1        14        94               4151               9271
AAACAGCTTTCAGAAG-1        43         9               7583               3393
AAACAGGGTCTATATT-1        47        13               8064               3665
                   percent.mt keep keep_counts layer_guess cell_count
AAACAAGTATCTCCCA-1   16.63514 TRUE        TRUE      Layer3          6
AAACAATCTACTAGCA-1   12.23755 TRUE       FALSE      Layer1         16
AAACACCAATAACTGC-1   11.40886 TRUE        TRUE          WM          5
AAACAGAGCGACTCCT-1   24.22234 TRUE        TRUE      Layer3          2
AAACAGCTTTCAGAAG-1   15.21739 TRUE        TRUE      Layer5          4
AAACAGGGTCTATATT-1   15.50949 TRUE        TRUE      Layer6          6
                   nCount_SCT nFeature_SCT SCT_snn_res.1 seurat_clusters
AAACAAGTATCTCCCA-1       5159         3183             1               1
AAACAATCTACTAGCA-1       3400         1264             4               4
AAACACCAATAACTGC-1       3828         1941             6               6
AAACAGAGCGACTCCT-1       4682         2393             5               5
AAACAGCTTTCAGAAG-1       4228         2239             0               0
AAACAGGGTCTATATT-1       3991         2148             3               3</code></pre>
</div>
<p>We previously discussed the layer annotations provided by Maynard and
colleagues. We added them to our Seurat object and plotted them. Let’s
look at them again to compare them to our clusters.</p>
<div class="codewrapper sourceCode" id="cb11">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">SpatialDimPlotColorSafe</span><span class="op">(</span><span class="va">filter_st</span><span class="op">[</span>, <span class="op">!</span><span class="fu">is.na</span><span class="op">(</span><span class="va">filter_st</span><span class="op">[[</span><span class="op">]</span><span class="op">]</span><span class="op">$</span><span class="va">layer_guess</span><span class="op">)</span><span class="op">]</span>, <span class="st">"layer_guess"</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu">labs</span><span class="op">(</span>fill<span class="op">=</span><span class="st">"Layer"</span><span class="op">)</span> </span></code></pre>
</div>
<figure><img src="../fig/feature-selection-dimensionality-reduction-clustering-rendered-unnamed-chunk-7-1.png" style="display: block; margin: auto;" class="figure mx-auto d-block"></figure><p>The authors describe six layers arranged from the upper right to the
lower left, and a white matter (WM) later. At this stage of the
analysis, we have nine clusters, but they do not show the clear
separation of the ground truth in the source publication.</p>
<p>We will now plot the spots in the tissue, colored by the clusters
which we have identified in <code>seurat_clusters</code> to evaluate the
quality of the cluster identities by looking for the clarity of the
stripes forming each layer.</p>
<div class="codewrapper sourceCode" id="cb12">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">SpatialDimPlotColorSafe</span><span class="op">(</span><span class="va">filter_st</span>, <span class="st">"seurat_clusters"</span><span class="op">)</span> <span class="op">+</span> <span class="fu">labs</span><span class="op">(</span>fill<span class="op">=</span><span class="st">"Cluster"</span><span class="op">)</span> </span></code></pre>
</div>
<figure><img src="../fig/feature-selection-dimensionality-reduction-clustering-rendered-unnamed-chunk-8-1.png" style="display: block; margin: auto;" class="figure mx-auto d-block"></figure><p>How many layers do we have compared to the publication? What do you
think about the quality of the layers in this plot? Are there clear
layers in the tissue?</p>
<p>Another way for us to look at the clusters it to plot them in a
Uniform Manifold Approximation and Projection (UMAP). UMAP is a
non-linear dimension reduction technique that can be used to visualize
data. This is often used in single-cell RNASeq to identify different
cell types. Here, each spot is potentially composed of more than one
cell type, so the clustering may not be as clear.</p>
<div class="codewrapper sourceCode" id="cb13">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">num_clusters</span> <span class="op">&lt;-</span> <span class="fu">length</span><span class="op">(</span><span class="fu">unique</span><span class="op">(</span><span class="va">filter_st</span><span class="op">[[</span><span class="op">]</span><span class="op">]</span><span class="op">$</span><span class="va">seurat_clusters</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">color_pal</span>    <span class="op">&lt;-</span> <span class="fu">setNames</span><span class="op">(</span><span class="fu">carto_pal</span><span class="op">(</span><span class="va">num_clusters</span>, <span class="st">"Safe"</span><span class="op">)</span>, <span class="fl">0</span><span class="op">:</span><span class="op">(</span><span class="va">num_clusters</span> <span class="op">-</span> <span class="fl">1</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="va">filter_st</span> <span class="op">&lt;-</span> <span class="fu">RunUMAP</span><span class="op">(</span><span class="va">filter_st</span>, </span>
<span>                     reduction <span class="op">=</span> <span class="st">'pca'</span>, </span>
<span>                     dims      <span class="op">=</span> <span class="fl">1</span><span class="op">:</span><span class="va">n_pcs</span>, </span>
<span>                     verbose   <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span>
<span></span>
<span><span class="fu">UMAPPlot</span><span class="op">(</span><span class="va">filter_st</span>, </span>
<span>         label      <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>         cols       <span class="op">=</span> <span class="va">color_pal</span>,</span>
<span>         pt.size    <span class="op">=</span> <span class="fl">2</span>,</span>
<span>         label.size <span class="op">=</span> <span class="fl">6</span><span class="op">)</span></span></code></pre>
</div>
<figure><img src="../fig/feature-selection-dimensionality-reduction-clustering-rendered-unnamed-chunk-9-1.png" style="display: block; margin: auto;" class="figure mx-auto d-block"></figure><p>We have made some decisions above which might affect the quality of
our spot clusters, including the number of nearest neighbors, the number
of variable features, the number of PCs, and the cluster resolution. As
we have mentioned, it is critical to have some understanding of the
structure of the tissue that you are analyzing. In the absence of
“ground truth”, you will need to try several different parameters and
look at how the clusters group in your tissue. Below, we have included
code which plots the tissue with different numbers of nearest neighbors,
principal components, and cluster resolutions. We will not run this code
because it takes a long time to run, but we show the output below this
code block.</p>
<div class="codewrapper sourceCode" id="cb14">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Set several cluster resolution values.</span></span>
<span><span class="va">resol</span> <span class="op">&lt;-</span> <span class="fu">c</span><span class="op">(</span><span class="fl">0.5</span>, <span class="fl">1</span>, <span class="fl">2</span><span class="op">)</span></span>
<span><span class="co"># Set several numbers of principal components.</span></span>
<span><span class="va">npcs</span>  <span class="op">&lt;-</span> <span class="fu">c</span><span class="op">(</span><span class="fl">25</span>, <span class="fl">50</span>, <span class="fl">75</span><span class="op">)</span></span>
<span></span>
<span><span class="va">num_clusters</span> <span class="op">&lt;-</span> <span class="fl">12</span></span>
<span><span class="va">color_pal</span>    <span class="op">&lt;-</span> <span class="fu">setNames</span><span class="op">(</span><span class="fu">carto_pal</span><span class="op">(</span><span class="va">num_clusters</span>, <span class="st">"Safe"</span><span class="op">)</span>, <span class="fl">0</span><span class="op">:</span><span class="op">(</span><span class="va">num_clusters</span> <span class="op">-</span> <span class="fl">1</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="va">umap_plots</span> <span class="op">&lt;-</span> <span class="fu">vector</span><span class="op">(</span><span class="st">'list'</span>, <span class="fu">length</span><span class="op">(</span><span class="va">resol</span><span class="op">)</span> <span class="op">*</span> <span class="fu">length</span><span class="op">(</span><span class="va">npcs</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">plots</span>      <span class="op">&lt;-</span> <span class="fu">vector</span><span class="op">(</span><span class="st">'list'</span>, <span class="fu">length</span><span class="op">(</span><span class="va">resol</span><span class="op">)</span> <span class="op">*</span> <span class="fu">length</span><span class="op">(</span><span class="va">npcs</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="kw">for</span><span class="op">(</span><span class="va">i</span> <span class="kw">in</span> <span class="fu">seq_along</span><span class="op">(</span><span class="va">resol</span><span class="op">)</span><span class="op">)</span> <span class="op">{</span></span>
<span>  <span class="kw">for</span><span class="op">(</span><span class="va">j</span> <span class="kw">in</span> <span class="fu">seq_along</span><span class="op">(</span><span class="va">npcs</span><span class="op">)</span><span class="op">)</span> <span class="op">{</span></span>
<span></span>
<span>    <span class="va">index</span>     <span class="op">&lt;-</span> <span class="op">(</span><span class="va">i</span><span class="op">-</span><span class="fl">1</span><span class="op">)</span> <span class="op">*</span> <span class="fu">length</span><span class="op">(</span><span class="va">npcs</span><span class="op">)</span> <span class="op">+</span> <span class="va">j</span></span>
<span>    <span class="fu">print</span><span class="op">(</span><span class="fu">paste</span><span class="op">(</span><span class="st">'Index ='</span>, <span class="va">index</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span>    <span class="va">filter_st</span> <span class="op">&lt;-</span> <span class="fu">FindNeighbors</span><span class="op">(</span><span class="va">filter_st</span>, </span>
<span>                               reduction <span class="op">=</span> <span class="st">"pca"</span>, </span>
<span>                               dims      <span class="op">=</span> <span class="fl">1</span><span class="op">:</span><span class="va">npcs</span><span class="op">[</span><span class="va">j</span><span class="op">]</span><span class="op">)</span> <span class="op">%&gt;%</span></span>
<span>                   <span class="fu">FindClusters</span><span class="op">(</span>resolution <span class="op">=</span> <span class="va">resol</span><span class="op">[</span><span class="va">i</span><span class="op">]</span><span class="op">)</span> <span class="op">%&gt;%</span></span>
<span>                   <span class="fu">RunUMAP</span><span class="op">(</span>reduction <span class="op">=</span> <span class="st">'pca'</span>, </span>
<span>                           dims      <span class="op">=</span> <span class="fl">1</span><span class="op">:</span><span class="va">npcs</span><span class="op">[</span><span class="va">j</span><span class="op">]</span>,</span>
<span>                           verbose   <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span>
<span></span>
<span>    <span class="va">umap_plots</span><span class="op">[[</span><span class="va">index</span><span class="op">]</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="fu">UMAPPlot</span><span class="op">(</span><span class="va">filter_st</span>, label <span class="op">=</span> <span class="cn">TRUE</span>, cols <span class="op">=</span> <span class="va">color_pal</span>, </span>
<span>                                    pt.size <span class="op">=</span> <span class="fl">2</span>, label.size <span class="op">=</span> <span class="fl">4</span><span class="op">)</span> <span class="op">+</span></span>
<span>                             <span class="fu">ggtitle</span><span class="op">(</span><span class="fu">paste</span><span class="op">(</span><span class="st">"res ="</span>, <span class="va">resol</span><span class="op">[</span><span class="va">i</span><span class="op">]</span>, <span class="st">": pc ="</span>, <span class="va">npcs</span><span class="op">[</span><span class="va">j</span><span class="op">]</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>                             <span class="fu">theme</span><span class="op">(</span>legend.position <span class="op">=</span> <span class="st">"none"</span><span class="op">)</span></span>
<span></span>
<span>    <span class="va">plots</span><span class="op">[[</span><span class="va">index</span><span class="op">]</span><span class="op">]</span>      <span class="op">&lt;-</span> <span class="fu">SpatialDimPlot</span><span class="op">(</span><span class="va">filter_st</span>, </span>
<span>                                          group.by <span class="op">=</span> <span class="st">"seurat_clusters"</span>, </span>
<span>                                          cols <span class="op">=</span> <span class="va">color_pal</span><span class="op">)</span> <span class="op">+</span></span>
<span>                          <span class="fu">ggtitle</span><span class="op">(</span><span class="fu">paste</span><span class="op">(</span><span class="st">"res ="</span>, <span class="va">resol</span><span class="op">[</span><span class="va">i</span><span class="op">]</span>, <span class="st">": pc ="</span>, <span class="va">npcs</span><span class="op">[</span><span class="va">j</span><span class="op">]</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span> </span>
<span>                          <span class="fu">theme</span><span class="op">(</span>legend.position <span class="op">=</span> <span class="st">"none"</span><span class="op">)</span></span>
<span></span>
<span>  <span class="op">}</span> <span class="co"># for(j)</span></span>
<span></span>
<span><span class="op">}</span> <span class="co"># for(i)</span></span>
<span></span>
<span><span class="fu">png</span><span class="op">(</span><span class="fu">file.path</span><span class="op">(</span><span class="st">'episodes'</span>, <span class="st">'fig'</span>, <span class="st">'tissue_cluster_resol.png'</span><span class="op">)</span>,</span>
<span>    width <span class="op">=</span> <span class="fl">1000</span>, height <span class="op">=</span> <span class="fl">1000</span>, res <span class="op">=</span> <span class="fl">128</span><span class="op">)</span></span>
<span><span class="fu">print</span><span class="op">(</span><span class="fu">gridExtra</span><span class="fu">::</span><span class="fu">grid.arrange</span><span class="op">(</span>grobs <span class="op">=</span> <span class="va">plots</span>, nrow <span class="op">=</span> <span class="fl">3</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu">dev.off</span><span class="op">(</span><span class="op">)</span></span>
<span></span>
<span><span class="fu">png</span><span class="op">(</span><span class="fu">file.path</span><span class="op">(</span><span class="st">'episodes'</span>, <span class="st">'fig'</span>, <span class="st">'umap_cluster_resol.png'</span><span class="op">)</span>,</span>
<span>    width <span class="op">=</span> <span class="fl">1000</span>, height <span class="op">=</span> <span class="fl">1000</span>, res <span class="op">=</span> <span class="fl">128</span><span class="op">)</span></span>
<span><span class="fu">print</span><span class="op">(</span><span class="fu">gridExtra</span><span class="fu">::</span><span class="fu">grid.arrange</span><span class="op">(</span>grobs <span class="op">=</span> <span class="va">umap_plots</span>, nrow <span class="op">=</span> <span class="fl">3</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu">dev.off</span><span class="op">(</span><span class="op">)</span></span></code></pre>
</div>
<p>The plot below shows the clusters as colors overlayed on the tissue.
Each row shows a different cluster resolution and each column shows the
number of PCs.</p>
<figure><img src="../fig/tissue_cluster_resol.png" alt="Tissue Clustering showing the tissue layers at different resolutions and number of PCs" class="figure mx-auto d-block"><div class="figcaption">Tissue Clustering with Different Numbers of PCs
and Clustering Resolutions</div>
</figure><p>The plot below shows the UMAP clustering for the same set of
parameters in the same order. The cluster colors are the same in the
plots above and below. Increasing the cluster resolution increases the
number of clusters.</p>
<figure><img src="../fig/umap_cluster_resol.png" alt="UMAP Clustering with Different Numbers of PCs and Clustering Resolutions" class="figure mx-auto d-block"><div class="figcaption">UMAP Clustering with Different Numbers of PCs
and Clustering Resolutions</div>
</figure><div id="challenge-1-select-cluster-resolution-and-number-of-pcs" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<div id="challenge-1-select-cluster-resolution-and-number-of-pcs" class="callout-inner">
<h3 class="callout-title">Challenge 1: Select cluster resolution and number of PCs<a class="anchor" aria-label="anchor" href="#challenge-1-select-cluster-resolution-and-number-of-pcs"></a>
</h3>
<div class="callout-content">
<p>Look at the two plots above which show the tissue and UMAP clusters
at different cluster resolutions and number of PCs. Think about which
settings seem to produce clustering that matches the expectations from
the publication. Turn to the person next to you and discuss your
opinions about which settings to use.</p>
</div>
</div>
</div>
<div id="accordionSolution2" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution2" aria-expanded="false" aria-controls="collapseSolution2">
  <h4 class="accordion-header" id="headingSolution2"> Solution 1 </h4>
</button>
<div id="collapseSolution2" class="accordion-collapse collapse" aria-labelledby="headingSolution2" data-bs-parent="#accordionSolution2">
<div class="accordion-body">
<p>All of the plots in the tissue clustering show layers which broadly
conform to the publication’s “ground truth.” In the top row at a
resolution of 0.5, we see about four layers with varying levels of
clarity, depending on the number of PCs. As we move down in the tissue
clustering plot, the layers become clearer at a resolution of 1, but
then become less clear at a resolution of 2 in the bottom row. When we
look at the corresponding row in the UMAP plots, we see that the white
matter is clearly separated in all of the plots. The number of clusters
increases with increasing cluster resolution, with the middle row
(cluster resolution = 1), having close to seven layers, like the
publication ground truth. The middle plot in the middle row has nine
clusters, which is also close to the publication.</p>
</div>
</div>
</div>
</div>
<p>We selected a cluster resolution of 0.8 and 50 PCs for the following
work.</p>
<div class="codewrapper sourceCode" id="cb15">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">filter_st</span> <span class="op">&lt;-</span> <span class="fu">FindNeighbors</span><span class="op">(</span><span class="va">filter_st</span>, </span>
<span>                           reduction <span class="op">=</span> <span class="st">"pca"</span>, </span>
<span>                           dims      <span class="op">=</span> <span class="fl">1</span><span class="op">:</span><span class="fl">50</span><span class="op">)</span> <span class="op">%&gt;%</span> </span>
<span>               <span class="fu">FindClusters</span><span class="op">(</span>resolution <span class="op">=</span> <span class="fl">0.8</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>Modularity Optimizer version 1.3.0 by Ludo Waltman and Nees Jan van Eck

Number of nodes: 3633
Number of edges: 171945

Running Louvain algorithm...
Maximum modularity in 10 random starts: 0.7408
Number of communities: 9
Elapsed time: 0 seconds</code></pre>
</div>
<div class="codewrapper sourceCode" id="cb17">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">SpatialDimPlotColorSafe</span><span class="op">(</span><span class="va">filter_st</span>, <span class="st">"seurat_clusters"</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu">ggtitle</span><span class="op">(</span>label <span class="op">=</span> <span class="st">"Tissue Clusters: 50 PCs, resol = 0.8"</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu">labs</span><span class="op">(</span>fill <span class="op">=</span> <span class="st">"Cluster"</span><span class="op">)</span></span></code></pre>
</div>
<figure><img src="../fig/feature-selection-dimensionality-reduction-clustering-rendered-unnamed-chunk-12-1.png" style="display: block; margin: auto;" class="figure mx-auto d-block"></figure><div id="keypoints1" class="callout keypoints">
<div class="callout-square">
<i class="callout-icon" data-feather="key"></i>
</div>
<div class="callout-inner">
<h3 class="callout-title">Key Points<a class="anchor" aria-label="anchor" href="#keypoints1"></a>
</h3>
<div class="callout-content">
<ul><li>Feature selection is a crucial step in spatial transcriptomics
analysis, particularly for non-variance-stabilizing normalization
methods like <code>NormalizeData</code>.</li>
<li>Techniques such as VST and mean-variance plotting enable researchers
to focus on genes that provide the most biological insight.</li>
<li>Different proportions of highly variable genes and feature selection
methods can significantly influence the analytical outcomes, emphasizing
the need for tailored approaches based on the specific characteristics
of each dataset.</li>
<li>Linear dimensionality reduction methods like PCA are crucial for
initial data simplification and noise reduction.</li>
<li>Nonlinear methods like UMAP are valuable for detailed exploration of
data structures post-linear preprocessing.</li>
<li>The sequential application of PCA and UMAP can provide a
comprehensive view of the spatial transcriptomics data, leveraging the
strengths of both linear and nonlinear approaches.</li>
</ul></div>
</div>
</div>
<!--
Place links that you need to refer to multiple times across pages here. Delete
any links that you are not going to use.
 -->
</div>
</section></div> <!-- / div.lesson-content -->
    </main><!-- / main#main-content.main-content --><nav class="bottom-pagination mx-md-4" aria-label="Previous and Next Chapter"><div class="d-block d-sm-block d-md-none">
        <a class="chapter-link" href="../instructor/apply-normalization-methods.html"><i aria-hidden="true" class="small-arrow" data-feather="arrow-left"></i>Previous</a>
        <a class="chapter-link float-end" href="../instructor/deconvolve-cell-types-in-a-spot.html">Next<i aria-hidden="true" class="small-arrow" data-feather="arrow-right"></i></a>
      </div>
      <!-- content for large screens -->
      <div class="d-none d-sm-none d-md-block">
        <a class="chapter-link" href="../instructor/apply-normalization-methods.html" rel="prev">
          <i aria-hidden="true" class="small-arrow" data-feather="arrow-left"></i>
          Previous: Normalization in
        </a>
        <a class="chapter-link float-end" href="../instructor/deconvolve-cell-types-in-a-spot.html" rel="next">
          Next: Deconvolution in...
          <i aria-hidden="true" class="small-arrow" data-feather="arrow-right"></i>
        </a>
      </div>
    </nav></div> <!-- / div.primary-content.col-xs-12 -->
<!-- END:   inst/pkgdown/templates/content-instructor.html-->

      </div><!--/div.row-->
      		<footer class="row footer mx-md-3"><hr><div class="col-md-6">
        <p>This lesson is subject to the <a href="CODE_OF_CONDUCT.html">Code of Conduct</a></p>
        <p>

        <a href="https://github.com/smcclatchy/spatial-transcriptomics/edit/main/episodes/feature-selection-dimensionality-reduction-clustering.Rmd" class="external-link">Edit on GitHub</a>

	
        | <a href="https://github.com/smcclatchy/spatial-transcriptomics/blob/main/CONTRIBUTING.md" class="external-link">Contributing</a>
        | <a href="https://github.com/smcclatchy/spatial-transcriptomics/" class="external-link">Source</a></p>
				<p><a href="https://github.com/smcclatchy/spatial-transcriptomics/blob/main/CITATION.cff" class="external-link">Cite</a> | <a href="mailto:susan.mcclatchy@jax.org">Contact</a> | <a href="https://carpentries.org/about/" class="external-link">About</a></p>
			</div>
			<div class="col-md-6">

        <p>Materials licensed under <a href="LICENSE.html">CC-BY 4.0</a> by the authors</p>

        <p>Template licensed under <a href="https://creativecommons.org/licenses/by-sa/4.0/" class="external-link">CC-BY 4.0</a> by <a href="https://carpentries.org/" class="external-link">The Carpentries</a></p>
        <p>Built with <a href="https://github.com/epiverse-trace/sandpaper/tree/patch-renv-github-bug" class="external-link">sandpaper (0.16.5)</a>, <a href="https://github.com/carpentries/pegboard/tree/ad2542f7f7b9c3f90cc871b355ff81ec14a09ca9" class="external-link">pegboard (0.7.6)</a>, and <a href="https://github.com/carpentries/varnish/tree/1.0.3" class="external-link">varnish (1.0.3)</a></p>
			</div>
		</footer></div> <!-- / div.container -->
	<div id="to-top">
		<a href="#top">
      <i class="search-icon" data-feather="arrow-up" role="img" aria-label="Back To Top"></i><br><!-- <span class="d-none d-sm-none d-md-none d-lg-none d-xl-block">Back</span> To Top --><span class="d-none d-sm-none d-md-none d-lg-none d-xl-block">Back</span> To Top
		</a>
	</div>
  <script type="application/ld+json">
    {
  "@context": "https://schema.org",
  "@type": "TrainingMaterial",
  "@id": "https://smcclatchy.github.io/spatial-transcriptomics/instructor/feature-selection-dimensionality-reduction-clustering.html",
  "inLanguage": "en",
  "dct:conformsTo": "https://bioschemas.org/profiles/TrainingMaterial/1.0-RELEASE",
  "description": "A Carpentries Lesson teaching foundational data and coding skills to researchers worldwide",
  "keywords": "software, data, lesson, The Carpentries, transcriptome, Visium, cancer",
  "name": "Feature Selection, Dimensionality Reduction, and Spot Clustering",
  "creativeWorkStatus": "active",
  "url": "https://smcclatchy.github.io/spatial-transcriptomics/instructor/feature-selection-dimensionality-reduction-clustering.html",
  "identifier": "https://smcclatchy.github.io/spatial-transcriptomics/instructor/feature-selection-dimensionality-reduction-clustering.html",
  "dateCreated": "2024-03-01",
  "dateModified": "2024-08-09",
  "datePublished": "2024-08-13"
}

  </script><script>
		feather.replace();
	</script></body></html><!-- END:   inst/pkgdown/templates/layout.html-->

